{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# table of contents:\n",
    "\n",
    "[1. Import hail, other libraries and data](#1)\n",
    "\n",
    "[2. Explore MatrixTable, collect field descriptions](#2)\n",
    "\n",
    "   [2.1 Removing the star alleles](#2.1)\n",
    "    \n",
    "[3. Do simple aggregations, filtering and plots](#3)\n",
    "\n",
    "[4. Work with single sample and single variant](#4)\n",
    "\n",
    "   [4.1 Single sample](#4.1)\n",
    "\n",
    "   [4.2 Single variant](#4.2)\n",
    "\n",
    "[5. Explore Clinvar sinificance of detected variants](#5)\n",
    "\n",
    "[5.1 Are there known pathogenic variants in > 1 sample?](#5.1)\n",
    "\n",
    "[6. Annotate with CADD (in progress)](#6)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='1'></a> \n",
    "## 1. Import hail, other libraries and data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "always run this code to widen notebook:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 456,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:100% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:100% !important; }</style>\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Running on Apache Spark version 2.4.1\n",
      "SparkUI available at http://349d1de1bab4:4040\n",
      "Welcome to\n",
      "     __  __     <>__\n",
      "    / /_/ /__  __/ /\n",
      "   / __  / _ `/ / /\n",
      "  /_/ /_/\\_,_/_/_/   version 0.2.26-2dcc3d963867\n",
      "LOGGING: writing to /hail/hail-20191114-0831-0.2.26-2dcc3d963867.log\n"
     ]
    }
   ],
   "source": [
    "import hail as hl\n",
    "hl.init() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 455,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bk-root\">\n",
       "        <a href=\"https://bokeh.pydata.org\" target=\"_blank\" class=\"bk-logo bk-logo-small bk-logo-notebook\"></a>\n",
       "        <span id=\"6287\">Loading BokehJS ...</span>\n",
       "    </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  var force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "  var JS_MIME_TYPE = 'application/javascript';\n",
       "  var HTML_MIME_TYPE = 'text/html';\n",
       "  var EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  var CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    var script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    var cell = handle.cell;\n",
       "\n",
       "    var id = cell.output_area._bokeh_element_id;\n",
       "    var server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      var cmd = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            var id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      var cmd = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    var output_area = handle.output_area;\n",
       "    var output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!output.data.hasOwnProperty(EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    var toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      var bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      var script_attrs = bk_div.children[0].attributes;\n",
       "      for (var i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      var toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      var props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    var events = require('base/js/events');\n",
       "    var OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  \n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  var NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    var el = document.getElementById(\"6287\");\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error() {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (var i = 0; i < css_urls.length; i++) {\n",
       "      var url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error;\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (var i = 0; i < js_urls.length; i++) {\n",
       "      var url = js_urls[i];\n",
       "      var element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error;\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };var element = document.getElementById(\"6287\");\n",
       "  if (element == null) {\n",
       "    console.error(\"Bokeh: ERROR: autoload.js configured with elementid '6287' but no matching script tag was found. \")\n",
       "    return false;\n",
       "  }\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  var js_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-1.2.0.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.2.0.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.2.0.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-gl-1.2.0.min.js\"];\n",
       "  var css_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-1.2.0.min.css\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.2.0.min.css\", \"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.2.0.min.css\"];\n",
       "\n",
       "  var inline_js = [\n",
       "    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "    \n",
       "    function(Bokeh) {\n",
       "      \n",
       "    },\n",
       "    function(Bokeh) {} // ensure no trailing comma for IE\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    \n",
       "    if ((root.Bokeh !== undefined) || (force === true)) {\n",
       "      for (var i = 0; i < inline_js.length; i++) {\n",
       "        inline_js[i].call(root, root.Bokeh);\n",
       "      }if (force === true) {\n",
       "        display_loaded();\n",
       "      }} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      var cell = $(document.getElementById(\"6287\")).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  var force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n  \n\n  \n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  var NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    var el = document.getElementById(\"6287\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error() {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (var i = 0; i < css_urls.length; i++) {\n      var url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (var i = 0; i < js_urls.length; i++) {\n      var url = js_urls[i];\n      var element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };var element = document.getElementById(\"6287\");\n  if (element == null) {\n    console.error(\"Bokeh: ERROR: autoload.js configured with elementid '6287' but no matching script tag was found. \")\n    return false;\n  }\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  var js_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-1.2.0.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.2.0.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.2.0.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-gl-1.2.0.min.js\"];\n  var css_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-1.2.0.min.css\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.2.0.min.css\", \"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.2.0.min.css\"];\n\n  var inline_js = [\n    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\n    \n    function(Bokeh) {\n      \n    },\n    function(Bokeh) {} // ensure no trailing comma for IE\n  ];\n\n  function run_inline_js() {\n    \n    if ((root.Bokeh !== undefined) || (force === true)) {\n      for (var i = 0; i < inline_js.length; i++) {\n        inline_js[i].call(root, root.Bokeh);\n      }if (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      var cell = $(document.getElementById(\"6287\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from hail.plot import show\n",
    "from pprint import pprint\n",
    "hl.plot.output_notebook()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-11-12 15:44:48 Hail: INFO: Ordering unsorted dataset with network shuffle\n",
      "2019-11-12 15:45:18 Hail: INFO: wrote matrix table with 726380 rows and 151 columns in 116 partitions to data/sample.mt\n"
     ]
    }
   ],
   "source": [
    "hl.import_vcf('data/annotated.test.vcf', reference_genome='GRCh38').write('data/sample.mt', overwrite=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 356,
   "metadata": {},
   "outputs": [],
   "source": [
    "mt = hl.read_matrix_table('data/sample.mt') # mt stands for MatrixTable"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='2'></a> \n",
    "## 2. Explore MatrixTable, collect field descriptions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [],
   "source": [
    "header = hl.get_vcf_metadata('data/annotated.test.vcf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 381,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'filter': {'PASS': {'Description': 'All filters passed'}},\n",
      " 'format': {'AD': {'Description': 'Allelic depths for the ref and alt alleles '\n",
      "                                  'in the order listed',\n",
      "                   'Number': '.',\n",
      "                   'Type': 'Integer'},\n",
      "            'DP': {'Description': 'Read depth',\n",
      "                   'Number': '1',\n",
      "                   'Type': 'Integer'},\n",
      "            'GQ': {'Description': 'Genotype quality',\n",
      "                   'Number': '1',\n",
      "                   'Type': 'Integer'},\n",
      "            'GT': {'Description': 'Genotype', 'Number': '1', 'Type': 'String'},\n",
      "            'PGT': {'Description': 'Physical phasing haplotype information, '\n",
      "                                   'describing how the alternate alleles are '\n",
      "                                   'phased in relation to one another',\n",
      "                    'Number': '1',\n",
      "                    'Type': 'String'},\n",
      "            'PID': {'Description': 'Physical phasing ID information, where '\n",
      "                                   'each unique ID within a given sample (but '\n",
      "                                   'not across samples) connects records '\n",
      "                                   'within a phasing group',\n",
      "                    'Number': '1',\n",
      "                    'Type': 'String'},\n",
      "            'PL': {'Description': 'The phred-scaled genotype likelihoods '\n",
      "                                  'rounded to the closest integer',\n",
      "                   'Number': 'G',\n",
      "                   'Type': 'Integer'}},\n",
      " 'info': {'AC': {'Description': 'Allele count in genotypes, for each ALT '\n",
      "                                'allele, in the same order as listed',\n",
      "                 'Number': 'A',\n",
      "                 'Type': 'Integer'},\n",
      "          'AF': {'Description': 'Allele frequency, for each ALT allele, in the '\n",
      "                                'same order as listed',\n",
      "                 'Number': 'A',\n",
      "                 'Type': 'Float'},\n",
      "          'ALTS': {'Description': 'ALT allels observed on this site. '\n",
      "                                  'Determined by counting ALT alleles observed '\n",
      "                                  'in samples in input VCF.',\n",
      "                   'Number': '.',\n",
      "                   'Type': 'String'},\n",
      "          'ALTS_indices': {'Description': 'Index of a corresponding ALT allele '\n",
      "                                          'from ALTS field (1-based).',\n",
      "                           'Number': 'A',\n",
      "                           'Type': 'String'},\n",
      "          'ALTS_pos': {'Description': 'Position of ALT alleles observed on '\n",
      "                                      'this site. Determined by counting ALT '\n",
      "                                      'alleles observed in samples in input '\n",
      "                                      'VCF.',\n",
      "                       'Number': '1',\n",
      "                       'Type': 'Integer'},\n",
      "          'AN': {'Description': 'Total number of alleles in called genotypes',\n",
      "                 'Number': '1',\n",
      "                 'Type': 'Integer'},\n",
      "          'ANN': {'Description': \"Functional annotations: 'Allele | Annotation \"\n",
      "                                 '| Annotation_Impact | Gene_Name | Gene_ID | '\n",
      "                                 'Feature_Type | Feature_ID | '\n",
      "                                 'Transcript_BioType | Rank | HGVS.c | HGVS.p '\n",
      "                                 '| cDNA.pos / cDNA.length | CDS.pos / '\n",
      "                                 'CDS.length | AA.pos / AA.length | Distance | '\n",
      "                                 \"ERRORS / WARNINGS / INFO'\",\n",
      "                  'Number': '.',\n",
      "                  'Type': 'String'},\n",
      "          'BaseQRankSum': {'Description': 'Z-score from Wilcoxon rank sum test '\n",
      "                                          'of Alt Vs. Ref base qualities',\n",
      "                           'Number': '1',\n",
      "                           'Type': 'Float'},\n",
      "          'ClippingRankSum': {'Description': 'Z-score From Wilcoxon rank sum '\n",
      "                                             'test of Alt vs. Ref number of '\n",
      "                                             'hard clipped bases',\n",
      "                              'Number': '1',\n",
      "                              'Type': 'Float'},\n",
      "          'DB': {'Description': 'dbSNP membership',\n",
      "                 'Number': '0',\n",
      "                 'Type': 'Flag'},\n",
      "          'DP': {'Description': 'Combined depth across samples',\n",
      "                 'Number': '1',\n",
      "                 'Type': 'Integer'},\n",
      "          'ExcessHet': {'Description': 'Phred-scaled p-value for exact test of '\n",
      "                                       'excess heterozygosity',\n",
      "                        'Number': '1',\n",
      "                        'Type': 'Float'},\n",
      "          'FS': {'Description': \"Phred-scaled p-value using Fisher's exact \"\n",
      "                                'test to detect strand bias',\n",
      "                 'Number': '1',\n",
      "                 'Type': 'Float'},\n",
      "          'ISEQ_AGGREGATED_CLINVAR_SIGNIFICANCE': {'Description': 'The overall '\n",
      "                                                                  'interpretation '\n",
      "                                                                  'of clinical '\n",
      "                                                                  'significance '\n",
      "                                                                  'of a '\n",
      "                                                                  'variation '\n",
      "                                                                  'based on '\n",
      "                                                                  'aggregating '\n",
      "                                                                  'data from '\n",
      "                                                                  'submitters. '\n",
      "                                                                  'Format '\n",
      "                                                                  'clinical_significance '\n",
      "                                                                  ': (...) : '\n",
      "                                                                  'clinical_significance',\n",
      "                                                   'Number': 'A',\n",
      "                                                   'Type': 'String'},\n",
      "          'ISEQ_AGGREGATED_CLINVAR_SIGNIFICANCE_INCLUDED': {'Description': 'Clinical '\n",
      "                                                                           'significance '\n",
      "                                                                           'for '\n",
      "                                                                           'a '\n",
      "                                                                           'haplotype '\n",
      "                                                                           'or '\n",
      "                                                                           'genotype '\n",
      "                                                                           'that '\n",
      "                                                                           'includes '\n",
      "                                                                           'the '\n",
      "                                                                           'variant '\n",
      "                                                                           'in '\n",
      "                                                                           'ClinVar. '\n",
      "                                                                           'Format: '\n",
      "                                                                           'VariationID '\n",
      "                                                                           '| '\n",
      "                                                                           'clinical '\n",
      "                                                                           'significance '\n",
      "                                                                           '^ '\n",
      "                                                                           '(...) '\n",
      "                                                                           '^ '\n",
      "                                                                           'clinical '\n",
      "                                                                           'significance '\n",
      "                                                                           ': '\n",
      "                                                                           '(...) '\n",
      "                                                                           ': '\n",
      "                                                                           'VariationID '\n",
      "                                                                           '| '\n",
      "                                                                           'clinical '\n",
      "                                                                           'significance '\n",
      "                                                                           '^ '\n",
      "                                                                           '(...) '\n",
      "                                                                           '^ '\n",
      "                                                                           'clinical '\n",
      "                                                                           'significance',\n",
      "                                                            'Number': 'A',\n",
      "                                                            'Type': 'String'},\n",
      "          'ISEQ_CLINVAR_ALLELE_ID': {'Description': 'ClinVar Allele ID',\n",
      "                                     'Number': 'A',\n",
      "                                     'Type': 'String'},\n",
      "          'ISEQ_CLINVAR_DISEASES': {'Description': 'Names of diseases '\n",
      "                                                   'associated with genes that '\n",
      "                                                   'overlap the allele '\n",
      "                                                   '(source: ClinVar). Names '\n",
      "                                                   'of diseases associated '\n",
      "                                                   \"with a given gene are '^' \"\n",
      "                                                   '- delimited. Grups of '\n",
      "                                                   'diseases associated with '\n",
      "                                                   \"genes are ':' - delimited. \"\n",
      "                                                   'The order of those groups '\n",
      "                                                   'is the same as the order '\n",
      "                                                   'of genes in '\n",
      "                                                   'ISEQ_GENES_NAMES field.',\n",
      "                                    'Number': 'A',\n",
      "                                    'Type': 'String'},\n",
      "          'ISEQ_CLINVAR_REVIEW_STATUS': {'Description': 'ClinVar review status '\n",
      "                                                        'for the Variation ID. '\n",
      "                                                        'Format: reviev_status '\n",
      "                                                        ': (...) : '\n",
      "                                                        'review_status',\n",
      "                                         'Number': 'A',\n",
      "                                         'Type': 'String'},\n",
      "          'ISEQ_CLINVAR_SIGNIFICANCE': {'Description': 'List of clinical '\n",
      "                                                       'significances of a '\n",
      "                                                       'variation gathered '\n",
      "                                                       'from ClinVar '\n",
      "                                                       'submissions. Taken '\n",
      "                                                       'from '\n",
      "                                                       'variant_summary.txt. '\n",
      "                                                       'Format '\n",
      "                                                       'clinical_significance '\n",
      "                                                       ': (...) : '\n",
      "                                                       'clinical_significance',\n",
      "                                        'Number': 'A',\n",
      "                                        'Type': 'String'},\n",
      "          'ISEQ_CLINVAR_SIGNIFICANCE_INCLUDED': {'Description': 'Clinical '\n",
      "                                                                'significance '\n",
      "                                                                'for a '\n",
      "                                                                'haplotype or '\n",
      "                                                                'genotype that '\n",
      "                                                                'includes the '\n",
      "                                                                'variant in '\n",
      "                                                                'ClinVar, '\n",
      "                                                                'based on '\n",
      "                                                                'variant_summary.txt. '\n",
      "                                                                'Format: '\n",
      "                                                                'VariationID | '\n",
      "                                                                'clinical '\n",
      "                                                                'significance '\n",
      "                                                                '^ (...) ^ '\n",
      "                                                                'clinical '\n",
      "                                                                'significance '\n",
      "                                                                ': (...) : '\n",
      "                                                                'VariationID | '\n",
      "                                                                'clinical '\n",
      "                                                                'significance '\n",
      "                                                                '^ (...) ^ '\n",
      "                                                                'clinical '\n",
      "                                                                'significance',\n",
      "                                                 'Number': 'A',\n",
      "                                                 'Type': 'String'},\n",
      "          'ISEQ_CLINVAR_VARIATION_ID': {'Description': 'ClinVar Variation IDs. '\n",
      "                                                       'Format: variation_ID : '\n",
      "                                                       '(...) : variation_ID',\n",
      "                                        'Number': 'A',\n",
      "                                        'Type': 'String'},\n",
      "          'ISEQ_DBSNP_RS': {'Description': 'dbSNP ID (i.e. rs number)',\n",
      "                            'Number': 'A',\n",
      "                            'Type': 'String'},\n",
      "          'ISEQ_ESP6500_AA_AF': {'Description': 'Alternate allele frequency - '\n",
      "                                                'ESP6500, African American '\n",
      "                                                'population (AA)',\n",
      "                                 'Number': 'A',\n",
      "                                 'Type': 'Float'},\n",
      "          'ISEQ_ESP6500_AF': {'Description': 'Alternate allele frequency - '\n",
      "                                             'ESP6500',\n",
      "                              'Number': 'A',\n",
      "                              'Type': 'Float'},\n",
      "          'ISEQ_ESP6500_AN': {'Description': 'Total allele count for - ESP6500',\n",
      "                              'Number': '1',\n",
      "                              'Type': 'Integer'},\n",
      "          'ISEQ_ESP6500_EA_AF': {'Description': 'Alternate allele frequency - '\n",
      "                                                'ESP6500, European American '\n",
      "                                                'population (EA)',\n",
      "                                 'Number': 'A',\n",
      "                                 'Type': 'Float'},\n",
      "          'ISEQ_EXAC_AC': {'Description': 'Alternate allele count for samples '\n",
      "                                          '(AC_Adj) - ExAC',\n",
      "                           'Number': 'A',\n",
      "                           'Type': 'Float'},\n",
      "          'ISEQ_EXAC_AF': {'Description': 'Alternate allele frequency in '\n",
      "                                          'samples (AC_Adj / AN_Adj) - ExAC',\n",
      "                           'Number': 'A',\n",
      "                           'Type': 'Float'},\n",
      "          'ISEQ_EXAC_AFR_AF': {'Description': 'Alternate allele frequency in '\n",
      "                                              'samples - ExAC, African / '\n",
      "                                              'African American population '\n",
      "                                              '(ARF)',\n",
      "                               'Number': 'A',\n",
      "                               'Type': 'Float'},\n",
      "          'ISEQ_EXAC_AMR_AF': {'Description': 'Alternate allele frequency in '\n",
      "                                              'samples - ExAC, Latino '\n",
      "                                              'population (AMR)',\n",
      "                               'Number': 'A',\n",
      "                               'Type': 'Float'},\n",
      "          'ISEQ_EXAC_AN': {'Description': 'Total number of alleles in samples '\n",
      "                                          '(AN_Adj) - ExAC',\n",
      "                           'Number': '1',\n",
      "                           'Type': 'Integer'},\n",
      "          'ISEQ_EXAC_EAS_AF': {'Description': 'Alternate allele frequency in '\n",
      "                                              'samples - ExAC, East Asian '\n",
      "                                              'population (EAS)',\n",
      "                               'Number': 'A',\n",
      "                               'Type': 'Float'},\n",
      "          'ISEQ_EXAC_FILTER_STATUS': {'Description': 'Filter status. Format: '\n",
      "                                                     'filter_status : (...) : '\n",
      "                                                     'filter_status. Possible '\n",
      "                                                     'values: PASS (All '\n",
      "                                                     'filters passed), '\n",
      "                                                     'InbreedingCoeff_Filter '\n",
      "                                                     '(InbreedingCoeff <= '\n",
      "                                                     '-0.8), LowQual (Low '\n",
      "                                                     'quality), NewCut_Filter '\n",
      "                                                     '(VQSLOD > -2.632 && '\n",
      "                                                     'InbreedingCoeff >-0.8), '\n",
      "                                                     'VQSRTrancheINDEL95.00to96.00 '\n",
      "                                                     '(Truth sensitivity '\n",
      "                                                     'tranche level for INDEL '\n",
      "                                                     'model at VQS Lod: 0.9503 '\n",
      "                                                     '<= x < 1.2168), '\n",
      "                                                     'VQSRTrancheINDEL96.00to97.00 '\n",
      "                                                     '(Truth sensitivity '\n",
      "                                                     'tranche level for INDEL '\n",
      "                                                     'model at VQS Lod: 0.7622 '\n",
      "                                                     '<= x < 0.9503), '\n",
      "                                                     'VQSRTrancheINDEL97.00to99.00 '\n",
      "                                                     '(Truth sensitivity '\n",
      "                                                     'tranche level for INDEL '\n",
      "                                                     'model at VQS Lod: 0.0426 '\n",
      "                                                     '<= x < 0.7622), '\n",
      "                                                     'VQSRTrancheINDEL99.00to99.50 '\n",
      "                                                     '(Truth sensitivity '\n",
      "                                                     'tranche level for INDEL '\n",
      "                                                     'model at VQS Lod: '\n",
      "                                                     '-0.8363 <= x < 0.0426), '\n",
      "                                                     'VQSRTrancheINDEL99.50to99.90 '\n",
      "                                                     '(Truth sensitivity '\n",
      "                                                     'tranche level for INDEL '\n",
      "                                                     'model at VQS Lod: '\n",
      "                                                     '-8.5421 <= x < -0.8363), '\n",
      "                                                     'VQSRTrancheINDEL99.90to99.95 '\n",
      "                                                     '(Truth sensitivity '\n",
      "                                                     'tranche level for INDEL '\n",
      "                                                     'model at VQS Lod: '\n",
      "                                                     '-18.4482 <= x < '\n",
      "                                                     '-8.5421), '\n",
      "                                                     'VQSRTrancheINDEL99.95to100.00+ '\n",
      "                                                     '(Truth sensitivity '\n",
      "                                                     'tranche level for INDEL '\n",
      "                                                     'model at VQS Lod < '\n",
      "                                                     '-37254.4742), '\n",
      "                                                     'VQSRTrancheINDEL99.95to100.00 '\n",
      "                                                     '(Truth sensitivity '\n",
      "                                                     'tranche level for INDEL '\n",
      "                                                     'model at VQS Lod: '\n",
      "                                                     '-37254.4742 <= x < '\n",
      "                                                     '-18.4482), '\n",
      "                                                     'VQSRTrancheSNP99.60to99.80 '\n",
      "                                                     '(Truth sensitivity '\n",
      "                                                     'tranche level for SNP '\n",
      "                                                     'model at VQS Lod: '\n",
      "                                                     '-4.9627 <= x < -1.8251), '\n",
      "                                                     'VQSRTrancheSNP99.80to99.90 '\n",
      "                                                     '(Truth sensitivity '\n",
      "                                                     'tranche level for SNP '\n",
      "                                                     'model at VQS Lod: '\n",
      "                                                     '-31.4709 <= x < '\n",
      "                                                     '-4.9627), '\n",
      "                                                     'VQSRTrancheSNP99.90to99.95 '\n",
      "                                                     '(Truth sensitivity '\n",
      "                                                     'tranche level for SNP '\n",
      "                                                     'model at VQS Lod: '\n",
      "                                                     '-170.3725 <= x < '\n",
      "                                                     '-31.4709), '\n",
      "                                                     'VQSRTrancheSNP99.95to100.00+ '\n",
      "                                                     '(Truth sensitivity '\n",
      "                                                     'tranche level for SNP '\n",
      "                                                     'model at VQS Lod < '\n",
      "                                                     '-39645.8352), '\n",
      "                                                     'VQSRTrancheSNP99.95to100.00 '\n",
      "                                                     '(Truth sensitivity '\n",
      "                                                     'tranche level for SNP '\n",
      "                                                     'model at VQS Lod: '\n",
      "                                                     '-39645.8352 <= x < '\n",
      "                                                     '-170.3725), '\n",
      "                                                     'AC_Adj0_Filter (No '\n",
      "                                                     'sample passed the '\n",
      "                                                     'adjusted threshold (GQ '\n",
      "                                                     '>= 20 and DP >= 10)) - '\n",
      "                                                     'ExAC',\n",
      "                                      'Number': 'A',\n",
      "                                      'Type': 'String'},\n",
      "          'ISEQ_EXAC_FIN_AF': {'Description': 'Alternate allele frequency in '\n",
      "                                              'samples - ExAC, Finnish '\n",
      "                                              'population (FIN)',\n",
      "                               'Number': 'A',\n",
      "                               'Type': 'Float'},\n",
      "          'ISEQ_EXAC_NFE_AF': {'Description': 'Alternate allele frequency in '\n",
      "                                              'samples - ExAC, Non-Finnish '\n",
      "                                              'European population (EUR)',\n",
      "                               'Number': 'A',\n",
      "                               'Type': 'Float'},\n",
      "          'ISEQ_EXAC_OTH_AF': {'Description': 'Alternate allele frequency in '\n",
      "                                              'samples - ExAC, Other '\n",
      "                                              'population (OTH)',\n",
      "                               'Number': 'A',\n",
      "                               'Type': 'Float'},\n",
      "          'ISEQ_EXAC_POPMAX': {'Description': 'Population with maximum allele '\n",
      "                                              'frequency - ExAC',\n",
      "                               'Number': 'A',\n",
      "                               'Type': 'String'},\n",
      "          'ISEQ_EXAC_POPMAX_AF': {'Description': 'Maximum allele frequency '\n",
      "                                                 'across populations - ExAC',\n",
      "                                  'Number': 'A',\n",
      "                                  'Type': 'Float'},\n",
      "          'ISEQ_EXAC_SAS_AF': {'Description': 'Alternate allele frequency in '\n",
      "                                              'samples - ExAC, South Asian '\n",
      "                                              'population (SAS)',\n",
      "                               'Number': 'A',\n",
      "                               'Type': 'Float'},\n",
      "          'ISEQ_GENES_NAMES': {'Description': 'Names of genes the variant has '\n",
      "                                              'at least MODERATE impact on.',\n",
      "                               'Number': 'A',\n",
      "                               'Type': 'String'},\n",
      "          'ISEQ_GERP': {'Description': 'GERP++ score',\n",
      "                        'Number': 'A',\n",
      "                        'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_AC': {'Description': 'Alternate allele count for '\n",
      "                                                   'samples - gnomAD exomes',\n",
      "                                    'Number': 'A',\n",
      "                                    'Type': 'Integer'},\n",
      "          'ISEQ_GNOMAD_EXOMES_AF': {'Description': 'Alternate allele frequency '\n",
      "                                                   'in samples - gnomAD exomes',\n",
      "                                    'Number': 'A',\n",
      "                                    'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_AF_afr': {'Description': 'Alternate allele '\n",
      "                                                       'frequency in samples '\n",
      "                                                       'of African-American '\n",
      "                                                       'ancestry (afr) - '\n",
      "                                                       'gnomAD exomes',\n",
      "                                        'Number': 'A',\n",
      "                                        'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_AF_amr': {'Description': 'Alternate allele '\n",
      "                                                       'frequency in samples '\n",
      "                                                       'of Latino ancestry '\n",
      "                                                       '(amr) - gnomAD exomes',\n",
      "                                        'Number': 'A',\n",
      "                                        'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_AF_asj': {'Description': 'Alternate allele '\n",
      "                                                       'frequency in samples '\n",
      "                                                       'of Ashkenazi Jewish '\n",
      "                                                       'ancestry (asj) - '\n",
      "                                                       'gnomAD exomes',\n",
      "                                        'Number': 'A',\n",
      "                                        'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_AF_eas': {'Description': 'Alternate allele '\n",
      "                                                       'frequency in samples '\n",
      "                                                       'of East Asian ancestry '\n",
      "                                                       '(eas) - gnomAD exomes',\n",
      "                                        'Number': 'A',\n",
      "                                        'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_AF_eas_subpopulations': {'Description': 'Alternate '\n",
      "                                                                      'allele '\n",
      "                                                                      'frequencies '\n",
      "                                                                      'in '\n",
      "                                                                      'samples '\n",
      "                                                                      'of '\n",
      "                                                                      'Korean '\n",
      "                                                                      'ancestry '\n",
      "                                                                      '(kor), '\n",
      "                                                                      'Japanese '\n",
      "                                                                      'ancestry '\n",
      "                                                                      '(jap), '\n",
      "                                                                      'non-Korean, '\n",
      "                                                                      'non-Japanese '\n",
      "                                                                      'East '\n",
      "                                                                      'Asian '\n",
      "                                                                      'ancestry '\n",
      "                                                                      '(oea) '\n",
      "                                                                      'in all '\n",
      "                                                                      'samples, '\n",
      "                                                                      'in the '\n",
      "                                                                      'controls '\n",
      "                                                                      'subset, '\n",
      "                                                                      'in the '\n",
      "                                                                      'non_neuro '\n",
      "                                                                      'subset '\n",
      "                                                                      'and in '\n",
      "                                                                      'the '\n",
      "                                                                      'non_cancer '\n",
      "                                                                      'subset: '\n",
      "                                                                      \"'Allele \"\n",
      "                                                                      '| '\n",
      "                                                                      'subpopulation_name '\n",
      "                                                                      ': '\n",
      "                                                                      'AF_eas_kor '\n",
      "                                                                      ': '\n",
      "                                                                      'controls_AF_eas_kor '\n",
      "                                                                      ': '\n",
      "                                                                      'non_neuro_AF_eas_kor '\n",
      "                                                                      ': '\n",
      "                                                                      'non_cancer_AF_eas_kor '\n",
      "                                                                      ': | '\n",
      "                                                                      'subpopulation_name '\n",
      "                                                                      ': '\n",
      "                                                                      'AF_eas_jpn '\n",
      "                                                                      ': '\n",
      "                                                                      'controls_AF_eas_jpn '\n",
      "                                                                      ': '\n",
      "                                                                      'non_neuro_AF_eas_jpn '\n",
      "                                                                      ': '\n",
      "                                                                      'non_cancer_AF_eas_jpn '\n",
      "                                                                      ': | '\n",
      "                                                                      'subpopulation_name '\n",
      "                                                                      ': '\n",
      "                                                                      'AF_eas_oea '\n",
      "                                                                      ': '\n",
      "                                                                      'controls_AF_eas_oea '\n",
      "                                                                      ': '\n",
      "                                                                      'non_neuro_AF_eas_oea '\n",
      "                                                                      ': '\n",
      "                                                                      \"non_cancer_AF_eas_oea' \"\n",
      "                                                                      '- '\n",
      "                                                                      'gnomAD '\n",
      "                                                                      'exomes',\n",
      "                                                       'Number': 'A',\n",
      "                                                       'Type': 'String'},\n",
      "          'ISEQ_GNOMAD_EXOMES_AF_fin': {'Description': 'Alternate allele '\n",
      "                                                       'frequency in samples '\n",
      "                                                       'of Finnish ancestry '\n",
      "                                                       '(fin) - gnomAD exomes',\n",
      "                                        'Number': 'A',\n",
      "                                        'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_AF_nfe': {'Description': 'Alternate allele '\n",
      "                                                       'frequency in samples '\n",
      "                                                       'of non-Finnish '\n",
      "                                                       'European ancestry '\n",
      "                                                       '(nfe) - gnomAD exomes',\n",
      "                                        'Number': 'A',\n",
      "                                        'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_AF_nfe_subpopulations': {'Description': 'Alternate '\n",
      "                                                                      'allele '\n",
      "                                                                      'frequencies '\n",
      "                                                                      'in '\n",
      "                                                                      'samples '\n",
      "                                                                      'of  '\n",
      "                                                                      'Bulgarian '\n",
      "                                                                      'ancestry '\n",
      "                                                                      '(bgr), '\n",
      "                                                                      'Estonian '\n",
      "                                                                      'ancestry '\n",
      "                                                                      '(est), '\n",
      "                                                                      'North-Western '\n",
      "                                                                      'European '\n",
      "                                                                      'ancestry '\n",
      "                                                                      '(nwe), '\n",
      "                                                                      'Southern '\n",
      "                                                                      'European '\n",
      "                                                                      'ancestry '\n",
      "                                                                      '(seu), '\n",
      "                                                                      'Swedish '\n",
      "                                                                      'ancestry '\n",
      "                                                                      '(swe), '\n",
      "                                                                      'non-Finnish '\n",
      "                                                                      'but '\n",
      "                                                                      'otherwise '\n",
      "                                                                      'indeterminate '\n",
      "                                                                      'European '\n",
      "                                                                      'ancestry '\n",
      "                                                                      'ancestry '\n",
      "                                                                      '(onf) '\n",
      "                                                                      'in all '\n",
      "                                                                      'samples, '\n",
      "                                                                      'in the '\n",
      "                                                                      'controls '\n",
      "                                                                      'subset, '\n",
      "                                                                      'in the '\n",
      "                                                                      'non_neuro '\n",
      "                                                                      'subset: '\n",
      "                                                                      \"'Allele \"\n",
      "                                                                      '| '\n",
      "                                                                      'subpopulation_name '\n",
      "                                                                      ': '\n",
      "                                                                      'AF_nfe_bgr '\n",
      "                                                                      ': '\n",
      "                                                                      'controls_AF_nfe_bgr '\n",
      "                                                                      ': '\n",
      "                                                                      'non_neuro_AF_nfe_bgr '\n",
      "                                                                      ': '\n",
      "                                                                      'non_cancer_AF_nfe_bgr '\n",
      "                                                                      ': | '\n",
      "                                                                      'subpopulation_name '\n",
      "                                                                      ': '\n",
      "                                                                      'AF_nfe_est '\n",
      "                                                                      ': '\n",
      "                                                                      'controls_AF_nfe_est '\n",
      "                                                                      ': '\n",
      "                                                                      'non_neuro_AF_nfe_est '\n",
      "                                                                      ': '\n",
      "                                                                      'non_cancer_AF_nfe_est '\n",
      "                                                                      ': | '\n",
      "                                                                      'subpopulation_name '\n",
      "                                                                      ': '\n",
      "                                                                      'AF_nfe_nwe '\n",
      "                                                                      ': '\n",
      "                                                                      'controls_AF_nfe_nwe '\n",
      "                                                                      ': '\n",
      "                                                                      'non_neuro_AF_nfe_nwe '\n",
      "                                                                      ': '\n",
      "                                                                      'non_cancer_AF_nfe_nwe '\n",
      "                                                                      ': | '\n",
      "                                                                      'subpopulation_name '\n",
      "                                                                      ': '\n",
      "                                                                      'AF_nfe_seu '\n",
      "                                                                      ': '\n",
      "                                                                      'controls_AF_nfe_seu '\n",
      "                                                                      ': '\n",
      "                                                                      'non_neuro_AF_nfe_seu '\n",
      "                                                                      ': '\n",
      "                                                                      'non_cancer_AF_nfe_seu '\n",
      "                                                                      ': | '\n",
      "                                                                      'subpopulation_name '\n",
      "                                                                      ': '\n",
      "                                                                      'AF_nfe_swe '\n",
      "                                                                      ': '\n",
      "                                                                      'controls_AF_nfe_swe '\n",
      "                                                                      ': '\n",
      "                                                                      'non_neuro_AF_nfe_swe '\n",
      "                                                                      ': '\n",
      "                                                                      'non_cancer_AF_nfe_swe '\n",
      "                                                                      ': | '\n",
      "                                                                      'subpopulation_name '\n",
      "                                                                      ': '\n",
      "                                                                      'AF_nfe_onf '\n",
      "                                                                      ': '\n",
      "                                                                      'controls_AF_nfe_onf '\n",
      "                                                                      ': '\n",
      "                                                                      'non_neuro_AF_nfe_onf '\n",
      "                                                                      ': '\n",
      "                                                                      \"non_cancer_AF_nfe_onf' \"\n",
      "                                                                      '- '\n",
      "                                                                      'gnomAD '\n",
      "                                                                      'exomes',\n",
      "                                                       'Number': 'A',\n",
      "                                                       'Type': 'String'},\n",
      "          'ISEQ_GNOMAD_EXOMES_AF_oth': {'Description': 'Alternate allele '\n",
      "                                                       'frequency in samples '\n",
      "                                                       'of uncertain ancestry '\n",
      "                                                       '(oth) - gnomAD exomes',\n",
      "                                        'Number': 'A',\n",
      "                                        'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_AF_sas': {'Description': 'Alternate allele '\n",
      "                                                       'frequency in samples '\n",
      "                                                       'of South Asian '\n",
      "                                                       'ancestry (sas) - '\n",
      "                                                       'gnomAD exomes',\n",
      "                                        'Number': 'A',\n",
      "                                        'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_AN': {'Description': 'Total number of alleles in '\n",
      "                                                   'samples - gnomAD exomes',\n",
      "                                    'Number': '1',\n",
      "                                    'Type': 'Integer'},\n",
      "          'ISEQ_GNOMAD_EXOMES_FILTER_STATUS': {'Description': 'Filter status. '\n",
      "                                                              'Format: '\n",
      "                                                              'filter_status : '\n",
      "                                                              '(...) : '\n",
      "                                                              'filter_status. '\n",
      "                                                              'Possible '\n",
      "                                                              'values: PASS '\n",
      "                                                              '(All filters '\n",
      "                                                              'passed), '\n",
      "                                                              'InbreedingCoeff '\n",
      "                                                              '(InbreedingCoeff '\n",
      "                                                              '< -0.3), RF '\n",
      "                                                              '(Failed random '\n",
      "                                                              'forest '\n",
      "                                                              'filtering '\n",
      "                                                              'thresholds of '\n",
      "                                                              '0.055272738028512555, '\n",
      "                                                              '0.20641025579497013 '\n",
      "                                                              '(probabilities '\n",
      "                                                              'of being a true '\n",
      "                                                              'positive '\n",
      "                                                              'variant) for '\n",
      "                                                              'SNPs, indels), '\n",
      "                                                              'AC0 (Allele '\n",
      "                                                              'count is zero '\n",
      "                                                              'after filtering '\n",
      "                                                              'out '\n",
      "                                                              'low-confidence '\n",
      "                                                              'genotypes (GQ < '\n",
      "                                                              '20; DP < 10; '\n",
      "                                                              'and AB < 0.2 '\n",
      "                                                              'for het calls) '\n",
      "                                                              '- gnomAD exomes',\n",
      "                                               'Number': 'A',\n",
      "                                               'Type': 'String'},\n",
      "          'ISEQ_GNOMAD_EXOMES_controls_AF': {'Description': 'Alternate allele '\n",
      "                                                            'frequency in '\n",
      "                                                            'samples in the '\n",
      "                                                            'controls subset - '\n",
      "                                                            'gnomAD exomes',\n",
      "                                             'Number': 'A',\n",
      "                                             'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_controls_AF_afr': {'Description': 'Alternate '\n",
      "                                                                'allele '\n",
      "                                                                'frequency in '\n",
      "                                                                'samples of '\n",
      "                                                                'African-American '\n",
      "                                                                'ancestry in '\n",
      "                                                                'the controls '\n",
      "                                                                'subset (afr) '\n",
      "                                                                '- gnomAD '\n",
      "                                                                'exomes',\n",
      "                                                 'Number': 'A',\n",
      "                                                 'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_controls_AF_amr': {'Description': 'Alternate '\n",
      "                                                                'allele '\n",
      "                                                                'frequency in '\n",
      "                                                                'samples of '\n",
      "                                                                'Latino '\n",
      "                                                                'ancestry in '\n",
      "                                                                'the controls '\n",
      "                                                                'subset (amr) '\n",
      "                                                                '- gnomAD '\n",
      "                                                                'exomes',\n",
      "                                                 'Number': 'A',\n",
      "                                                 'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_controls_AF_asj': {'Description': 'Alternate '\n",
      "                                                                'allele '\n",
      "                                                                'frequency in '\n",
      "                                                                'samples of '\n",
      "                                                                'Ashkenazi '\n",
      "                                                                'Jewish '\n",
      "                                                                'ancestry in '\n",
      "                                                                'the controls '\n",
      "                                                                'subset (asj) '\n",
      "                                                                '- gnomAD '\n",
      "                                                                'exomes',\n",
      "                                                 'Number': 'A',\n",
      "                                                 'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_controls_AF_eas': {'Description': 'Alternate '\n",
      "                                                                'allele '\n",
      "                                                                'frequency in '\n",
      "                                                                'samples of '\n",
      "                                                                'East Asian '\n",
      "                                                                'ancestry in '\n",
      "                                                                'the controls '\n",
      "                                                                'subset (eas) '\n",
      "                                                                '- gnomAD '\n",
      "                                                                'exomes',\n",
      "                                                 'Number': 'A',\n",
      "                                                 'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_controls_AF_fin': {'Description': 'Alternate '\n",
      "                                                                'allele '\n",
      "                                                                'frequency in '\n",
      "                                                                'samples of '\n",
      "                                                                'Finnish '\n",
      "                                                                'ancestry in '\n",
      "                                                                'the controls '\n",
      "                                                                'subset (fin) '\n",
      "                                                                '- gnomAD '\n",
      "                                                                'exomes',\n",
      "                                                 'Number': 'A',\n",
      "                                                 'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_controls_AF_nfe': {'Description': 'Alternate '\n",
      "                                                                'allele '\n",
      "                                                                'frequency in '\n",
      "                                                                'samples of '\n",
      "                                                                'non-Finnish '\n",
      "                                                                'European '\n",
      "                                                                'ancestry in '\n",
      "                                                                'the controls '\n",
      "                                                                'subset (nfe) '\n",
      "                                                                '- gnomAD '\n",
      "                                                                'exomes',\n",
      "                                                 'Number': 'A',\n",
      "                                                 'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_controls_AF_oth': {'Description': 'Alternate '\n",
      "                                                                'allele '\n",
      "                                                                'frequency in '\n",
      "                                                                'samples of '\n",
      "                                                                'uncertain '\n",
      "                                                                'ancestry in '\n",
      "                                                                'the controls '\n",
      "                                                                'subset (oth) '\n",
      "                                                                '- gnomAD '\n",
      "                                                                'exomes',\n",
      "                                                 'Number': 'A',\n",
      "                                                 'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_controls_AF_popmax': {'Description': 'Maximum '\n",
      "                                                                   'allele '\n",
      "                                                                   'frequency '\n",
      "                                                                   'across '\n",
      "                                                                   'populations '\n",
      "                                                                   '(excluding '\n",
      "                                                                   'samples of '\n",
      "                                                                   'Ashkenazi, '\n",
      "                                                                   'Finnish, '\n",
      "                                                                   'and '\n",
      "                                                                   'indeterminate '\n",
      "                                                                   'ancestry) '\n",
      "                                                                   'in the '\n",
      "                                                                   'controls '\n",
      "                                                                   'subset - '\n",
      "                                                                   'gnomAD '\n",
      "                                                                   'exomes',\n",
      "                                                    'Number': 'A',\n",
      "                                                    'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_controls_AF_sas': {'Description': 'Alternate '\n",
      "                                                                'allele '\n",
      "                                                                'frequency in '\n",
      "                                                                'samples of '\n",
      "                                                                'South Asian '\n",
      "                                                                'ancestry in '\n",
      "                                                                'the controls '\n",
      "                                                                'subset (sas) '\n",
      "                                                                '- gnomAD '\n",
      "                                                                'exomes',\n",
      "                                                 'Number': 'A',\n",
      "                                                 'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_controls_nhomalt': {'Description': 'Count of '\n",
      "                                                                 'homozygous '\n",
      "                                                                 'individuals '\n",
      "                                                                 'in samples '\n",
      "                                                                 'in the '\n",
      "                                                                 'controls '\n",
      "                                                                 'subset - '\n",
      "                                                                 'gnomAD '\n",
      "                                                                 'exomes',\n",
      "                                                  'Number': 'A',\n",
      "                                                  'Type': 'Integer'},\n",
      "          'ISEQ_GNOMAD_EXOMES_controls_nhomalt_popmax': {'Description': 'Count '\n",
      "                                                                        'of '\n",
      "                                                                        'homozygous '\n",
      "                                                                        'individuals '\n",
      "                                                                        'in '\n",
      "                                                                        'the '\n",
      "                                                                        'population '\n",
      "                                                                        'with '\n",
      "                                                                        'the '\n",
      "                                                                        'maximum '\n",
      "                                                                        'allele '\n",
      "                                                                        'frequency '\n",
      "                                                                        'in '\n",
      "                                                                        'the '\n",
      "                                                                        'controls '\n",
      "                                                                        'subset '\n",
      "                                                                        '- '\n",
      "                                                                        'gnomAD '\n",
      "                                                                        'exomes',\n",
      "                                                         'Number': 'A',\n",
      "                                                         'Type': 'Integer'},\n",
      "          'ISEQ_GNOMAD_EXOMES_controls_popmax': {'Description': 'Population '\n",
      "                                                                'with maximum '\n",
      "                                                                'AF in the '\n",
      "                                                                'controls '\n",
      "                                                                'subset - '\n",
      "                                                                'gnomAD exomes',\n",
      "                                                 'Number': 'A',\n",
      "                                                 'Type': 'String'},\n",
      "          'ISEQ_GNOMAD_EXOMES_decoy': {'Description': 'Variant falls within a '\n",
      "                                                      'reference decoy region '\n",
      "                                                      '- gnomAD exomes',\n",
      "                                       'Number': '0',\n",
      "                                       'Type': 'Flag'},\n",
      "          'ISEQ_GNOMAD_EXOMES_lcr': {'Description': 'Variant falls within a '\n",
      "                                                    'low complexity region - '\n",
      "                                                    'gnomAD exomes',\n",
      "                                     'Number': '0',\n",
      "                                     'Type': 'Flag'},\n",
      "          'ISEQ_GNOMAD_EXOMES_nhomalt': {'Description': 'Count of homozygous '\n",
      "                                                        'individuals in '\n",
      "                                                        'samples - gnomAD '\n",
      "                                                        'exomes',\n",
      "                                         'Number': 'A',\n",
      "                                         'Type': 'Integer'},\n",
      "          'ISEQ_GNOMAD_EXOMES_nhomalt_popmax': {'Description': 'Count of '\n",
      "                                                               'homozygous '\n",
      "                                                               'individuals in '\n",
      "                                                               'the population '\n",
      "                                                               'with the '\n",
      "                                                               'maximum allele '\n",
      "                                                               'frequency - '\n",
      "                                                               'gnomAD exomes',\n",
      "                                                'Number': 'A',\n",
      "                                                'Type': 'Integer'},\n",
      "          'ISEQ_GNOMAD_EXOMES_non_cancer_AF': {'Description': 'Alternate '\n",
      "                                                              'allele '\n",
      "                                                              'frequency in '\n",
      "                                                              'samples in the '\n",
      "                                                              'non_cancer '\n",
      "                                                              'subset - gnomAD '\n",
      "                                                              'exomes',\n",
      "                                               'Number': 'A',\n",
      "                                               'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_non_cancer_AF_afr': {'Description': 'Alternate '\n",
      "                                                                  'allele '\n",
      "                                                                  'frequency '\n",
      "                                                                  'in samples '\n",
      "                                                                  'of '\n",
      "                                                                  'African-American '\n",
      "                                                                  'ancestry in '\n",
      "                                                                  'the '\n",
      "                                                                  'non_cancer '\n",
      "                                                                  'subset '\n",
      "                                                                  '(afr) - '\n",
      "                                                                  'gnomAD '\n",
      "                                                                  'exomes',\n",
      "                                                   'Number': 'A',\n",
      "                                                   'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_non_cancer_AF_amr': {'Description': 'Alternate '\n",
      "                                                                  'allele '\n",
      "                                                                  'frequency '\n",
      "                                                                  'in samples '\n",
      "                                                                  'of Latino '\n",
      "                                                                  'ancestry in '\n",
      "                                                                  'the '\n",
      "                                                                  'non_cancer '\n",
      "                                                                  'subset '\n",
      "                                                                  '(amr) - '\n",
      "                                                                  'gnomAD '\n",
      "                                                                  'exomes',\n",
      "                                                   'Number': 'A',\n",
      "                                                   'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_non_cancer_AF_asj': {'Description': 'Alternate '\n",
      "                                                                  'allele '\n",
      "                                                                  'frequency '\n",
      "                                                                  'in samples '\n",
      "                                                                  'of '\n",
      "                                                                  'Ashkenazi '\n",
      "                                                                  'Jewish '\n",
      "                                                                  'ancestry in '\n",
      "                                                                  'the '\n",
      "                                                                  'non_cancer '\n",
      "                                                                  'subset '\n",
      "                                                                  '(asj) - '\n",
      "                                                                  'gnomAD '\n",
      "                                                                  'exomes',\n",
      "                                                   'Number': 'A',\n",
      "                                                   'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_non_cancer_AF_eas': {'Description': 'Alternate '\n",
      "                                                                  'allele '\n",
      "                                                                  'frequency '\n",
      "                                                                  'in samples '\n",
      "                                                                  'of East '\n",
      "                                                                  'Asian '\n",
      "                                                                  'ancestry in '\n",
      "                                                                  'the '\n",
      "                                                                  'non_cancer '\n",
      "                                                                  'subset '\n",
      "                                                                  '(eas) - '\n",
      "                                                                  'gnomAD '\n",
      "                                                                  'exomes',\n",
      "                                                   'Number': 'A',\n",
      "                                                   'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_non_cancer_AF_fin': {'Description': 'Alternate '\n",
      "                                                                  'allele '\n",
      "                                                                  'frequency '\n",
      "                                                                  'in samples '\n",
      "                                                                  'of Finnish '\n",
      "                                                                  'ancestry in '\n",
      "                                                                  'the '\n",
      "                                                                  'non_cancer '\n",
      "                                                                  'subset '\n",
      "                                                                  '(fin) - '\n",
      "                                                                  'gnomAD '\n",
      "                                                                  'exomes',\n",
      "                                                   'Number': 'A',\n",
      "                                                   'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_non_cancer_AF_nfe': {'Description': 'Alternate '\n",
      "                                                                  'allele '\n",
      "                                                                  'frequency '\n",
      "                                                                  'in samples '\n",
      "                                                                  'of '\n",
      "                                                                  'non-Finnish '\n",
      "                                                                  'European '\n",
      "                                                                  'ancestry in '\n",
      "                                                                  'the '\n",
      "                                                                  'non_cancer '\n",
      "                                                                  'subset '\n",
      "                                                                  '(nfe) - '\n",
      "                                                                  'gnomAD '\n",
      "                                                                  'exomes',\n",
      "                                                   'Number': 'A',\n",
      "                                                   'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_non_cancer_AF_oth': {'Description': 'Alternate '\n",
      "                                                                  'allele '\n",
      "                                                                  'frequency '\n",
      "                                                                  'in samples '\n",
      "                                                                  'of '\n",
      "                                                                  'uncertain '\n",
      "                                                                  'ancestry in '\n",
      "                                                                  'the '\n",
      "                                                                  'non_cancer '\n",
      "                                                                  'subset '\n",
      "                                                                  '(oth) - '\n",
      "                                                                  'gnomAD '\n",
      "                                                                  'exomes',\n",
      "                                                   'Number': 'A',\n",
      "                                                   'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_non_cancer_AF_popmax': {'Description': 'Maximum '\n",
      "                                                                     'allele '\n",
      "                                                                     'frequency '\n",
      "                                                                     'across '\n",
      "                                                                     'populations '\n",
      "                                                                     '(excluding '\n",
      "                                                                     'samples '\n",
      "                                                                     'of '\n",
      "                                                                     'Ashkenazi, '\n",
      "                                                                     'Finnish, '\n",
      "                                                                     'and '\n",
      "                                                                     'indeterminate '\n",
      "                                                                     'ancestry) '\n",
      "                                                                     'in the '\n",
      "                                                                     'non_cancer '\n",
      "                                                                     'subset - '\n",
      "                                                                     'gnomAD '\n",
      "                                                                     'exomes',\n",
      "                                                      'Number': 'A',\n",
      "                                                      'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_non_cancer_AF_sas': {'Description': 'Alternate '\n",
      "                                                                  'allele '\n",
      "                                                                  'frequency '\n",
      "                                                                  'in samples '\n",
      "                                                                  'of South '\n",
      "                                                                  'Asian '\n",
      "                                                                  'ancestry in '\n",
      "                                                                  'the '\n",
      "                                                                  'non_cancer '\n",
      "                                                                  'subset '\n",
      "                                                                  '(sas) - '\n",
      "                                                                  'gnomAD '\n",
      "                                                                  'exomes',\n",
      "                                                   'Number': 'A',\n",
      "                                                   'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_non_cancer_nhomalt': {'Description': 'Count of '\n",
      "                                                                   'homozygous '\n",
      "                                                                   'individuals '\n",
      "                                                                   'in samples '\n",
      "                                                                   'in the '\n",
      "                                                                   'non_cancer '\n",
      "                                                                   'subset - '\n",
      "                                                                   'gnomAD '\n",
      "                                                                   'exomes',\n",
      "                                                    'Number': 'A',\n",
      "                                                    'Type': 'Integer'},\n",
      "          'ISEQ_GNOMAD_EXOMES_non_cancer_nhomalt_popmax': {'Description': 'Count '\n",
      "                                                                          'of '\n",
      "                                                                          'homozygous '\n",
      "                                                                          'individuals '\n",
      "                                                                          'in '\n",
      "                                                                          'the '\n",
      "                                                                          'population '\n",
      "                                                                          'with '\n",
      "                                                                          'the '\n",
      "                                                                          'maximum '\n",
      "                                                                          'allele '\n",
      "                                                                          'frequency '\n",
      "                                                                          'in '\n",
      "                                                                          'the '\n",
      "                                                                          'non_cancer '\n",
      "                                                                          'subset '\n",
      "                                                                          '- '\n",
      "                                                                          'gnomAD '\n",
      "                                                                          'exomes',\n",
      "                                                           'Number': 'A',\n",
      "                                                           'Type': 'Integer'},\n",
      "          'ISEQ_GNOMAD_EXOMES_non_cancer_popmax': {'Description': 'Population '\n",
      "                                                                  'with '\n",
      "                                                                  'maximum AF '\n",
      "                                                                  'in the '\n",
      "                                                                  'non_cancer '\n",
      "                                                                  'subset - '\n",
      "                                                                  'gnomAD '\n",
      "                                                                  'exomes',\n",
      "                                                   'Number': 'A',\n",
      "                                                   'Type': 'String'},\n",
      "          'ISEQ_GNOMAD_EXOMES_non_neuro_AF': {'Description': 'Alternate allele '\n",
      "                                                             'frequency in '\n",
      "                                                             'samples in the '\n",
      "                                                             'non_neuro subset '\n",
      "                                                             '- gnomAD exomes',\n",
      "                                              'Number': 'A',\n",
      "                                              'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_non_neuro_AF_afr': {'Description': 'Alternate '\n",
      "                                                                 'allele '\n",
      "                                                                 'frequency in '\n",
      "                                                                 'samples of '\n",
      "                                                                 'African-American '\n",
      "                                                                 'ancestry in '\n",
      "                                                                 'the '\n",
      "                                                                 'non_neuro '\n",
      "                                                                 'subset (afr) '\n",
      "                                                                 '- gnomAD '\n",
      "                                                                 'exomes',\n",
      "                                                  'Number': 'A',\n",
      "                                                  'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_non_neuro_AF_amr': {'Description': 'Alternate '\n",
      "                                                                 'allele '\n",
      "                                                                 'frequency in '\n",
      "                                                                 'samples of '\n",
      "                                                                 'Latino '\n",
      "                                                                 'ancestry in '\n",
      "                                                                 'the '\n",
      "                                                                 'non_neuro '\n",
      "                                                                 'subset (amr) '\n",
      "                                                                 '- gnomAD '\n",
      "                                                                 'exomes',\n",
      "                                                  'Number': 'A',\n",
      "                                                  'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_non_neuro_AF_asj': {'Description': 'Alternate '\n",
      "                                                                 'allele '\n",
      "                                                                 'frequency in '\n",
      "                                                                 'samples of '\n",
      "                                                                 'Ashkenazi '\n",
      "                                                                 'Jewish '\n",
      "                                                                 'ancestry in '\n",
      "                                                                 'the '\n",
      "                                                                 'non_neuro '\n",
      "                                                                 'subset (asj) '\n",
      "                                                                 '- gnomAD '\n",
      "                                                                 'exomes',\n",
      "                                                  'Number': 'A',\n",
      "                                                  'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_non_neuro_AF_eas': {'Description': 'Alternate '\n",
      "                                                                 'allele '\n",
      "                                                                 'frequency in '\n",
      "                                                                 'samples of '\n",
      "                                                                 'East Asian '\n",
      "                                                                 'ancestry in '\n",
      "                                                                 'the '\n",
      "                                                                 'non_neuro '\n",
      "                                                                 'subset (eas) '\n",
      "                                                                 '- gnomAD '\n",
      "                                                                 'exomes',\n",
      "                                                  'Number': 'A',\n",
      "                                                  'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_non_neuro_AF_fin': {'Description': 'Alternate '\n",
      "                                                                 'allele '\n",
      "                                                                 'frequency in '\n",
      "                                                                 'samples of '\n",
      "                                                                 'Finnish '\n",
      "                                                                 'ancestry in '\n",
      "                                                                 'the '\n",
      "                                                                 'non_neuro '\n",
      "                                                                 'subset (fin) '\n",
      "                                                                 '- gnomAD '\n",
      "                                                                 'exomes',\n",
      "                                                  'Number': 'A',\n",
      "                                                  'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_non_neuro_AF_nfe': {'Description': 'Alternate '\n",
      "                                                                 'allele '\n",
      "                                                                 'frequency in '\n",
      "                                                                 'samples of '\n",
      "                                                                 'non-Finnish '\n",
      "                                                                 'European '\n",
      "                                                                 'ancestry in '\n",
      "                                                                 'the '\n",
      "                                                                 'non_neuro '\n",
      "                                                                 'subset (nfe) '\n",
      "                                                                 '- gnomAD '\n",
      "                                                                 'exomes',\n",
      "                                                  'Number': 'A',\n",
      "                                                  'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_non_neuro_AF_oth': {'Description': 'Alternate '\n",
      "                                                                 'allele '\n",
      "                                                                 'frequency in '\n",
      "                                                                 'samples of '\n",
      "                                                                 'uncertain '\n",
      "                                                                 'ancestry in '\n",
      "                                                                 'the '\n",
      "                                                                 'non_neuro '\n",
      "                                                                 'subset (oth) '\n",
      "                                                                 '- gnomAD '\n",
      "                                                                 'exomes',\n",
      "                                                  'Number': 'A',\n",
      "                                                  'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_non_neuro_AF_popmax': {'Description': 'Maximum '\n",
      "                                                                    'allele '\n",
      "                                                                    'frequency '\n",
      "                                                                    'across '\n",
      "                                                                    'populations '\n",
      "                                                                    '(excluding '\n",
      "                                                                    'samples '\n",
      "                                                                    'of '\n",
      "                                                                    'Ashkenazi, '\n",
      "                                                                    'Finnish, '\n",
      "                                                                    'and '\n",
      "                                                                    'indeterminate '\n",
      "                                                                    'ancestry) '\n",
      "                                                                    'in the '\n",
      "                                                                    'non_neuro '\n",
      "                                                                    'subset - '\n",
      "                                                                    'gnomAD '\n",
      "                                                                    'exomes',\n",
      "                                                     'Number': 'A',\n",
      "                                                     'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_non_neuro_AF_sas': {'Description': 'Alternate '\n",
      "                                                                 'allele '\n",
      "                                                                 'frequency in '\n",
      "                                                                 'samples of '\n",
      "                                                                 'South Asian '\n",
      "                                                                 'ancestry in '\n",
      "                                                                 'the '\n",
      "                                                                 'non_neuro '\n",
      "                                                                 'subset (sas) '\n",
      "                                                                 '- gnomAD '\n",
      "                                                                 'exomes',\n",
      "                                                  'Number': 'A',\n",
      "                                                  'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_non_neuro_nhomalt': {'Description': 'Count of '\n",
      "                                                                  'homozygous '\n",
      "                                                                  'individuals '\n",
      "                                                                  'in samples '\n",
      "                                                                  'in the '\n",
      "                                                                  'non_neuro '\n",
      "                                                                  'subset - '\n",
      "                                                                  'gnomAD '\n",
      "                                                                  'exomes',\n",
      "                                                   'Number': 'A',\n",
      "                                                   'Type': 'Integer'},\n",
      "          'ISEQ_GNOMAD_EXOMES_non_neuro_nhomalt_popmax': {'Description': 'Count '\n",
      "                                                                         'of '\n",
      "                                                                         'homozygous '\n",
      "                                                                         'individuals '\n",
      "                                                                         'in '\n",
      "                                                                         'the '\n",
      "                                                                         'population '\n",
      "                                                                         'with '\n",
      "                                                                         'the '\n",
      "                                                                         'maximum '\n",
      "                                                                         'allele '\n",
      "                                                                         'frequency '\n",
      "                                                                         'in '\n",
      "                                                                         'the '\n",
      "                                                                         'non_neuro '\n",
      "                                                                         'subset '\n",
      "                                                                         '- '\n",
      "                                                                         'gnomAD '\n",
      "                                                                         'exomes',\n",
      "                                                          'Number': 'A',\n",
      "                                                          'Type': 'Integer'},\n",
      "          'ISEQ_GNOMAD_EXOMES_non_neuro_popmax': {'Description': 'Population '\n",
      "                                                                 'with maximum '\n",
      "                                                                 'AF in the '\n",
      "                                                                 'non_neuro '\n",
      "                                                                 'subset - '\n",
      "                                                                 'gnomAD '\n",
      "                                                                 'exomes',\n",
      "                                                  'Number': 'A',\n",
      "                                                  'Type': 'String'},\n",
      "          'ISEQ_GNOMAD_EXOMES_nonpar': {'Description': 'Variant (on sex '\n",
      "                                                       'chromosome) falls '\n",
      "                                                       'outside a '\n",
      "                                                       'pseudoautosomal region '\n",
      "                                                       '- gnomAD exomes',\n",
      "                                        'Number': '0',\n",
      "                                        'Type': 'Flag'},\n",
      "          'ISEQ_GNOMAD_EXOMES_popmax': {'Description': 'Population with '\n",
      "                                                       'maximum AF - gnomAD '\n",
      "                                                       'exomes',\n",
      "                                        'Number': 'A',\n",
      "                                        'Type': 'String'},\n",
      "          'ISEQ_GNOMAD_EXOMES_popmax_AF': {'Description': 'Alternate allele '\n",
      "                                                          'frequency in '\n",
      "                                                          'samples - gnomAD '\n",
      "                                                          'exomes',\n",
      "                                           'Number': 'A',\n",
      "                                           'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_EXOMES_segdup': {'Description': 'Variant falls within a '\n",
      "                                                       'segmental duplication '\n",
      "                                                       'region - gnomAD exomes',\n",
      "                                        'Number': '0',\n",
      "                                        'Type': 'Flag'},\n",
      "          'ISEQ_GNOMAD_GENOMES_AC': {'Description': 'Alternate allele count '\n",
      "                                                    'for samples - gnomAD '\n",
      "                                                    'genomes',\n",
      "                                     'Number': 'A',\n",
      "                                     'Type': 'Integer'},\n",
      "          'ISEQ_GNOMAD_GENOMES_AF': {'Description': 'Alternate allele '\n",
      "                                                    'frequency in samples - '\n",
      "                                                    'gnomAD genomes',\n",
      "                                     'Number': 'A',\n",
      "                                     'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_AF_afr': {'Description': 'Alternate allele '\n",
      "                                                        'frequency in samples '\n",
      "                                                        'of African-American '\n",
      "                                                        'ancestry (afr) - '\n",
      "                                                        'gnomAD genomes',\n",
      "                                         'Number': 'A',\n",
      "                                         'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_AF_amr': {'Description': 'Alternate allele '\n",
      "                                                        'frequency in samples '\n",
      "                                                        'of Latino ancestry '\n",
      "                                                        '(amr) - gnomAD '\n",
      "                                                        'genomes',\n",
      "                                         'Number': 'A',\n",
      "                                         'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_AF_asj': {'Description': 'Alternate allele '\n",
      "                                                        'frequency in samples '\n",
      "                                                        'of Ashkenazi Jewish '\n",
      "                                                        'ancestry (asj) - '\n",
      "                                                        'gnomAD genomes',\n",
      "                                         'Number': 'A',\n",
      "                                         'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_AF_eas': {'Description': 'Alternate allele '\n",
      "                                                        'frequency in samples '\n",
      "                                                        'of East Asian '\n",
      "                                                        'ancestry (eas) - '\n",
      "                                                        'gnomAD genomes',\n",
      "                                         'Number': 'A',\n",
      "                                         'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_AF_fin': {'Description': 'Alternate allele '\n",
      "                                                        'frequency in samples '\n",
      "                                                        'of Finnish ancestry '\n",
      "                                                        '(fin) - gnomAD '\n",
      "                                                        'genomes',\n",
      "                                         'Number': 'A',\n",
      "                                         'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_AF_nfe': {'Description': 'Alternate allele '\n",
      "                                                        'frequency in samples '\n",
      "                                                        'of non-Finnish '\n",
      "                                                        'European ancestry '\n",
      "                                                        '(nfe) - gnomAD '\n",
      "                                                        'genomes',\n",
      "                                         'Number': 'A',\n",
      "                                         'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_AF_nfe_subpopulations': {'Description': 'Alternate '\n",
      "                                                                       'allele '\n",
      "                                                                       'frequencies '\n",
      "                                                                       'in '\n",
      "                                                                       'samples '\n",
      "                                                                       'of '\n",
      "                                                                       'Estonian '\n",
      "                                                                       'ancestry '\n",
      "                                                                       '(est), '\n",
      "                                                                       'North-Western '\n",
      "                                                                       'European '\n",
      "                                                                       'ancestry '\n",
      "                                                                       '(nwe), '\n",
      "                                                                       'Southern '\n",
      "                                                                       'European '\n",
      "                                                                       'ancestry '\n",
      "                                                                       '(seu), '\n",
      "                                                                       'non-Finnish '\n",
      "                                                                       'but '\n",
      "                                                                       'otherwise '\n",
      "                                                                       'indeterminate '\n",
      "                                                                       'European '\n",
      "                                                                       'ancestry '\n",
      "                                                                       'ancestry '\n",
      "                                                                       '(onf) '\n",
      "                                                                       'in all '\n",
      "                                                                       'samples, '\n",
      "                                                                       'in the '\n",
      "                                                                       'controls '\n",
      "                                                                       'subset '\n",
      "                                                                       'and in '\n",
      "                                                                       'the '\n",
      "                                                                       'non_neuro '\n",
      "                                                                       'subset: '\n",
      "                                                                       \"'Allele \"\n",
      "                                                                       '| '\n",
      "                                                                       'subpopulation_name '\n",
      "                                                                       ': '\n",
      "                                                                       'AF_nfe_est '\n",
      "                                                                       ': '\n",
      "                                                                       'controls_AF_nfe_est '\n",
      "                                                                       ': '\n",
      "                                                                       'non_neuro_AF_nfe_est '\n",
      "                                                                       '| '\n",
      "                                                                       'subpopulation_name '\n",
      "                                                                       ': '\n",
      "                                                                       'AF_nfe_nwe '\n",
      "                                                                       ': '\n",
      "                                                                       'controls_AF_nfe_nwe: '\n",
      "                                                                       'non_neuro_AF_nfe_nwe '\n",
      "                                                                       '| '\n",
      "                                                                       'subpopulation_name '\n",
      "                                                                       ': '\n",
      "                                                                       'AF_nfe_seu '\n",
      "                                                                       ': '\n",
      "                                                                       'controls_AF_nfe_seu '\n",
      "                                                                       ': '\n",
      "                                                                       'non_neuro_AF_nfe_seu '\n",
      "                                                                       '| '\n",
      "                                                                       'subpopulation_name '\n",
      "                                                                       ': '\n",
      "                                                                       'AF_nfe_onf '\n",
      "                                                                       ': '\n",
      "                                                                       'controls_AF_nfe_onf '\n",
      "                                                                       ': '\n",
      "                                                                       \"non_neuro_AF_nfe_onf' \"\n",
      "                                                                       '- '\n",
      "                                                                       'gnomAD '\n",
      "                                                                       'genomes',\n",
      "                                                        'Number': 'A',\n",
      "                                                        'Type': 'String'},\n",
      "          'ISEQ_GNOMAD_GENOMES_AF_oth': {'Description': 'Alternate allele '\n",
      "                                                        'frequency in samples '\n",
      "                                                        'of uncertain ancestry '\n",
      "                                                        '- gnomAD genomes',\n",
      "                                         'Number': 'A',\n",
      "                                         'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_AN': {'Description': 'Total number of alleles '\n",
      "                                                    'in samples -  gnomAD '\n",
      "                                                    'genomes',\n",
      "                                     'Number': '1',\n",
      "                                     'Type': 'Integer'},\n",
      "          'ISEQ_GNOMAD_GENOMES_FILTER_STATUS': {'Description': 'Filter status. '\n",
      "                                                               'Format: '\n",
      "                                                               'filter_status '\n",
      "                                                               ': (...) : '\n",
      "                                                               'filter_status. '\n",
      "                                                               'Possible '\n",
      "                                                               'values: PASS '\n",
      "                                                               '(All filters '\n",
      "                                                               'passed), '\n",
      "                                                               'InbreedingCoeff '\n",
      "                                                               '(InbreedingCoeff '\n",
      "                                                               '< -0.3), RF '\n",
      "                                                               '(failed random '\n",
      "                                                               'forest '\n",
      "                                                               'filtering '\n",
      "                                                               'thresholds of '\n",
      "                                                               '0.2634762834546574, '\n",
      "                                                               '0.22213813189901457 '\n",
      "                                                               '(probabilities '\n",
      "                                                               'of being a '\n",
      "                                                               'true positive '\n",
      "                                                               'variant) for '\n",
      "                                                               'SNPs, indels), '\n",
      "                                                               'AC0 (Allele '\n",
      "                                                               'count is zero '\n",
      "                                                               'after '\n",
      "                                                               'filtering out '\n",
      "                                                               'low-confidence '\n",
      "                                                               'genotypes (GQ '\n",
      "                                                               '< 20; DP < 10; '\n",
      "                                                               'and AB < 0.2 '\n",
      "                                                               'for het calls) '\n",
      "                                                               '- gnomAD '\n",
      "                                                               'genomes',\n",
      "                                                'Number': 'A',\n",
      "                                                'Type': 'String'},\n",
      "          'ISEQ_GNOMAD_GENOMES_V3_AC': {'Description': 'Alternate allele count '\n",
      "                                                       'for samples - gnomAD '\n",
      "                                                       'v3',\n",
      "                                        'Number': 'A',\n",
      "                                        'Type': 'Integer'},\n",
      "          'ISEQ_GNOMAD_GENOMES_V3_AF': {'Description': 'Alternate allele '\n",
      "                                                       'frequency in samples - '\n",
      "                                                       'gnomAD v3',\n",
      "                                        'Number': 'A',\n",
      "                                        'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_V3_AF_afr': {'Description': 'Alternate allele '\n",
      "                                                           'frequency in '\n",
      "                                                           'samples of '\n",
      "                                                           'African-American '\n",
      "                                                           'ancestry (afr) - '\n",
      "                                                           'gnomAD v3',\n",
      "                                            'Number': 'A',\n",
      "                                            'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_V3_AF_ami': {'Description': 'Alternate allele '\n",
      "                                                           'frequency in '\n",
      "                                                           'samples of Amish '\n",
      "                                                           'ancestry (ami) - '\n",
      "                                                           'gnomAD v3',\n",
      "                                            'Number': 'A',\n",
      "                                            'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_V3_AF_amr': {'Description': 'Alternate allele '\n",
      "                                                           'frequency in '\n",
      "                                                           'samples of Latino '\n",
      "                                                           'ancestry (amr) - '\n",
      "                                                           'gnomAD v3',\n",
      "                                            'Number': 'A',\n",
      "                                            'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_V3_AF_asj': {'Description': 'Alternate allele '\n",
      "                                                           'frequency in '\n",
      "                                                           'samples of '\n",
      "                                                           'Ashkenazi Jewish '\n",
      "                                                           'ancestry (asj) - '\n",
      "                                                           'gnomAD v3',\n",
      "                                            'Number': 'A',\n",
      "                                            'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_V3_AF_eas': {'Description': 'Alternate allele '\n",
      "                                                           'frequency in '\n",
      "                                                           'samples of East '\n",
      "                                                           'Asian ancestry '\n",
      "                                                           '(eas) - gnomAD v3',\n",
      "                                            'Number': 'A',\n",
      "                                            'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_V3_AF_fin': {'Description': 'Alternate allele '\n",
      "                                                           'frequency in '\n",
      "                                                           'samples of Finnish '\n",
      "                                                           'ancestry (fin) - '\n",
      "                                                           'gnomAD v3',\n",
      "                                            'Number': 'A',\n",
      "                                            'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_V3_AF_nfe': {'Description': 'Alternate allele '\n",
      "                                                           'frequency in '\n",
      "                                                           'samples of '\n",
      "                                                           'non-Finnish '\n",
      "                                                           'European ancestry '\n",
      "                                                           '(nfe) - gnomAD v3',\n",
      "                                            'Number': 'A',\n",
      "                                            'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_V3_AF_oth': {'Description': 'Alternate allele '\n",
      "                                                           'frequency in '\n",
      "                                                           'samples of '\n",
      "                                                           'uncertain ancestry '\n",
      "                                                           '- gnomAD v3',\n",
      "                                            'Number': 'A',\n",
      "                                            'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_V3_AF_sas': {'Description': 'Alternate allele '\n",
      "                                                           'frequency in '\n",
      "                                                           'samples of South '\n",
      "                                                           'Asian ancestry '\n",
      "                                                           '(sas) - gnomAD v3',\n",
      "                                            'Number': 'A',\n",
      "                                            'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_V3_AN': {'Description': 'Total number of '\n",
      "                                                       'alleles in samples -  '\n",
      "                                                       'gnomAD genomes',\n",
      "                                        'Number': '1',\n",
      "                                        'Type': 'Integer'},\n",
      "          'ISEQ_GNOMAD_GENOMES_V3_FILTER_STATUS': {'Description': 'Filter '\n",
      "                                                                  'status. '\n",
      "                                                                  'Format: '\n",
      "                                                                  'filter_status '\n",
      "                                                                  ': (...) : '\n",
      "                                                                  'filter_status. '\n",
      "                                                                  'Possible '\n",
      "                                                                  'values: '\n",
      "                                                                  'PASS (All '\n",
      "                                                                  'filters '\n",
      "                                                                  'passed), '\n",
      "                                                                  'InbreedingCoeff '\n",
      "                                                                  '(InbreedingCoeff '\n",
      "                                                                  '< -0.3), RF '\n",
      "                                                                  '(failed '\n",
      "                                                                  'random '\n",
      "                                                                  'forest '\n",
      "                                                                  'filtering '\n",
      "                                                                  'thresholds '\n",
      "                                                                  'of '\n",
      "                                                                  '0.2634762834546574, '\n",
      "                                                                  '0.22213813189901457 '\n",
      "                                                                  '(probabilities '\n",
      "                                                                  'of being a '\n",
      "                                                                  'true '\n",
      "                                                                  'positive '\n",
      "                                                                  'variant) '\n",
      "                                                                  'for SNPs, '\n",
      "                                                                  'indels), '\n",
      "                                                                  'AC0 (Allele '\n",
      "                                                                  'count is '\n",
      "                                                                  'zero after '\n",
      "                                                                  'filtering '\n",
      "                                                                  'out '\n",
      "                                                                  'low-confidence '\n",
      "                                                                  'genotypes '\n",
      "                                                                  '(GQ < 20; '\n",
      "                                                                  'DP < 10; '\n",
      "                                                                  'and AB < '\n",
      "                                                                  '0.2 for het '\n",
      "                                                                  'calls) - '\n",
      "                                                                  'gnomAD v3',\n",
      "                                                   'Number': 'A',\n",
      "                                                   'Type': 'String'},\n",
      "          'ISEQ_GNOMAD_GENOMES_V3_lcr': {'Description': 'Variant falls within '\n",
      "                                                        'a low complexity '\n",
      "                                                        'region - gnomAD v3',\n",
      "                                         'Number': '0',\n",
      "                                         'Type': 'Flag'},\n",
      "          'ISEQ_GNOMAD_GENOMES_V3_nhomalt': {'Description': 'Count of '\n",
      "                                                            'homozygous '\n",
      "                                                            'individuals in '\n",
      "                                                            'samples - gnomAD '\n",
      "                                                            'v3',\n",
      "                                             'Number': 'A',\n",
      "                                             'Type': 'Integer'},\n",
      "          'ISEQ_GNOMAD_GENOMES_V3_nhomalt_popmax': {'Description': 'Count of '\n",
      "                                                                   'homozygous '\n",
      "                                                                   'individuals '\n",
      "                                                                   'in the '\n",
      "                                                                   'population '\n",
      "                                                                   'with the '\n",
      "                                                                   'maximum '\n",
      "                                                                   'allele '\n",
      "                                                                   'frequency '\n",
      "                                                                   '- gnomAD '\n",
      "                                                                   'v3',\n",
      "                                                    'Number': 'A',\n",
      "                                                    'Type': 'String'},\n",
      "          'ISEQ_GNOMAD_GENOMES_V3_nonpar': {'Description': 'Variant (on sex '\n",
      "                                                           'chromosome) falls '\n",
      "                                                           'outside a '\n",
      "                                                           'pseudoautosomal '\n",
      "                                                           'region - gnomAD v3',\n",
      "                                            'Number': '0',\n",
      "                                            'Type': 'Flag'},\n",
      "          'ISEQ_GNOMAD_GENOMES_V3_popmax': {'Description': 'Population with '\n",
      "                                                           'maximum AF -  '\n",
      "                                                           'gnomAD genomes',\n",
      "                                            'Number': 'A',\n",
      "                                            'Type': 'String'},\n",
      "          'ISEQ_GNOMAD_GENOMES_V3_popmax_AF': {'Description': 'Maximum allele '\n",
      "                                                              'frequency '\n",
      "                                                              'across '\n",
      "                                                              'populations '\n",
      "                                                              '(excluding '\n",
      "                                                              'samples of '\n",
      "                                                              'Ashkenazi, '\n",
      "                                                              'Finnish, and '\n",
      "                                                              'indeterminate '\n",
      "                                                              'ancestry) - '\n",
      "                                                              'gnomAD v3',\n",
      "                                               'Number': 'A',\n",
      "                                               'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_controls_AF': {'Description': 'Alternate allele '\n",
      "                                                             'frequency in '\n",
      "                                                             'samples in the '\n",
      "                                                             'controls subset '\n",
      "                                                             '-  gnomAD '\n",
      "                                                             'genomes',\n",
      "                                              'Number': 'A',\n",
      "                                              'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_controls_AF_afr': {'Description': 'Alternate '\n",
      "                                                                 'allele '\n",
      "                                                                 'frequency in '\n",
      "                                                                 'samples of '\n",
      "                                                                 'African-American '\n",
      "                                                                 'ancestry in '\n",
      "                                                                 'the controls '\n",
      "                                                                 'subset (afr) '\n",
      "                                                                 '- gnomAD '\n",
      "                                                                 'genomes',\n",
      "                                                  'Number': 'A',\n",
      "                                                  'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_controls_AF_amr': {'Description': 'Alternate '\n",
      "                                                                 'allele '\n",
      "                                                                 'frequency in '\n",
      "                                                                 'samples of '\n",
      "                                                                 'Latino '\n",
      "                                                                 'ancestry in '\n",
      "                                                                 'the controls '\n",
      "                                                                 'subset (amr) '\n",
      "                                                                 '- gnomAD '\n",
      "                                                                 'genomes',\n",
      "                                                  'Number': 'A',\n",
      "                                                  'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_controls_AF_asj': {'Description': 'Alternate '\n",
      "                                                                 'allele '\n",
      "                                                                 'frequency in '\n",
      "                                                                 'samples of '\n",
      "                                                                 'Ashkenazi '\n",
      "                                                                 'Jewish '\n",
      "                                                                 'ancestry in '\n",
      "                                                                 'the controls '\n",
      "                                                                 'subset (asj) '\n",
      "                                                                 '- gnomAD '\n",
      "                                                                 'genomes',\n",
      "                                                  'Number': 'A',\n",
      "                                                  'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_controls_AF_eas': {'Description': 'Alternate '\n",
      "                                                                 'allele '\n",
      "                                                                 'frequency in '\n",
      "                                                                 'samples of '\n",
      "                                                                 'East Asian '\n",
      "                                                                 'ancestry in '\n",
      "                                                                 'the controls '\n",
      "                                                                 'subset (eas) '\n",
      "                                                                 '- gnomAD '\n",
      "                                                                 'genomes',\n",
      "                                                  'Number': 'A',\n",
      "                                                  'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_controls_AF_fin': {'Description': 'Alternate '\n",
      "                                                                 'allele '\n",
      "                                                                 'frequency in '\n",
      "                                                                 'samples of '\n",
      "                                                                 'Finnish '\n",
      "                                                                 'ancestry in '\n",
      "                                                                 'the controls '\n",
      "                                                                 'subset (fin) '\n",
      "                                                                 '- gnomAD '\n",
      "                                                                 'genomes',\n",
      "                                                  'Number': 'A',\n",
      "                                                  'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_controls_AF_nfe': {'Description': 'Alternate '\n",
      "                                                                 'allele '\n",
      "                                                                 'frequency in '\n",
      "                                                                 'samples of '\n",
      "                                                                 'non-Finnish '\n",
      "                                                                 'European '\n",
      "                                                                 'ancestry in '\n",
      "                                                                 'the controls '\n",
      "                                                                 'subset (nfe) '\n",
      "                                                                 '- gnomAD '\n",
      "                                                                 'genomes',\n",
      "                                                  'Number': 'A',\n",
      "                                                  'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_controls_AF_oth': {'Description': 'Alternate '\n",
      "                                                                 'allele '\n",
      "                                                                 'frequency in '\n",
      "                                                                 'samples of '\n",
      "                                                                 'uncertain '\n",
      "                                                                 'ancestry in '\n",
      "                                                                 'the controls '\n",
      "                                                                 'subset - '\n",
      "                                                                 'gnomAD '\n",
      "                                                                 'genomes',\n",
      "                                                  'Number': 'A',\n",
      "                                                  'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_controls_AF_popmax': {'Description': 'Maximum '\n",
      "                                                                    'allele '\n",
      "                                                                    'frequency '\n",
      "                                                                    'across '\n",
      "                                                                    'populations '\n",
      "                                                                    '(excluding '\n",
      "                                                                    'samples '\n",
      "                                                                    'of '\n",
      "                                                                    'Ashkenazi, '\n",
      "                                                                    'Finnish, '\n",
      "                                                                    'and '\n",
      "                                                                    'indeterminate '\n",
      "                                                                    'ancestry) '\n",
      "                                                                    'in the '\n",
      "                                                                    'controls '\n",
      "                                                                    'subset - '\n",
      "                                                                    'gnomAD '\n",
      "                                                                    'genomes',\n",
      "                                                     'Number': 'A',\n",
      "                                                     'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_controls_nhomalt': {'Description': 'Count of '\n",
      "                                                                  'homozygous '\n",
      "                                                                  'individuals '\n",
      "                                                                  'in samples '\n",
      "                                                                  'in the '\n",
      "                                                                  'controls '\n",
      "                                                                  'subset -  '\n",
      "                                                                  'gnomAD '\n",
      "                                                                  'genomes',\n",
      "                                                   'Number': 'A',\n",
      "                                                   'Type': 'Integer'},\n",
      "          'ISEQ_GNOMAD_GENOMES_controls_nhomalt_popmax': {'Description': 'Count '\n",
      "                                                                         'of '\n",
      "                                                                         'homozygous '\n",
      "                                                                         'individuals '\n",
      "                                                                         'in '\n",
      "                                                                         'the '\n",
      "                                                                         'population '\n",
      "                                                                         'with '\n",
      "                                                                         'the '\n",
      "                                                                         'maximum '\n",
      "                                                                         'allele '\n",
      "                                                                         'frequency '\n",
      "                                                                         'in '\n",
      "                                                                         'the '\n",
      "                                                                         'controls '\n",
      "                                                                         'subset '\n",
      "                                                                         '- '\n",
      "                                                                         'gnomAD '\n",
      "                                                                         'genomes',\n",
      "                                                          'Number': 'A',\n",
      "                                                          'Type': 'Integer'},\n",
      "          'ISEQ_GNOMAD_GENOMES_controls_popmax': {'Description': 'Population '\n",
      "                                                                 'with maximum '\n",
      "                                                                 'AF in the '\n",
      "                                                                 'controls '\n",
      "                                                                 'subset - '\n",
      "                                                                 'gnomAD '\n",
      "                                                                 'genomes',\n",
      "                                                  'Number': 'A',\n",
      "                                                  'Type': 'String'},\n",
      "          'ISEQ_GNOMAD_GENOMES_decoy': {'Description': 'Variant falls within a '\n",
      "                                                       'reference decoy region '\n",
      "                                                       '- gnomAD genomes',\n",
      "                                        'Number': '0',\n",
      "                                        'Type': 'Flag'},\n",
      "          'ISEQ_GNOMAD_GENOMES_lcr': {'Description': 'Variant falls within a '\n",
      "                                                     'low complexity region - '\n",
      "                                                     'gnomAD genomes',\n",
      "                                      'Number': '0',\n",
      "                                      'Type': 'Flag'},\n",
      "          'ISEQ_GNOMAD_GENOMES_nhomalt': {'Description': 'Count of homozygous '\n",
      "                                                         'individuals in '\n",
      "                                                         'samples - gnomAD '\n",
      "                                                         'genomes',\n",
      "                                          'Number': 'A',\n",
      "                                          'Type': 'Integer'},\n",
      "          'ISEQ_GNOMAD_GENOMES_nhomalt_popmax': {'Description': 'Count of '\n",
      "                                                                'homozygous '\n",
      "                                                                'individuals '\n",
      "                                                                'in the '\n",
      "                                                                'population '\n",
      "                                                                'with the '\n",
      "                                                                'maximum '\n",
      "                                                                'allele '\n",
      "                                                                'frequency - '\n",
      "                                                                'gnomAD '\n",
      "                                                                'genomes',\n",
      "                                                 'Number': 'A',\n",
      "                                                 'Type': 'Integer'},\n",
      "          'ISEQ_GNOMAD_GENOMES_non_neuro_AF': {'Description': 'Alternate '\n",
      "                                                              'allele '\n",
      "                                                              'frequency in '\n",
      "                                                              'samples in the '\n",
      "                                                              'non_neuro '\n",
      "                                                              'subset -  '\n",
      "                                                              'gnomAD genomes',\n",
      "                                               'Number': 'A',\n",
      "                                               'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_non_neuro_AF_afr': {'Description': 'Alternate '\n",
      "                                                                  'allele '\n",
      "                                                                  'frequency '\n",
      "                                                                  'in samples '\n",
      "                                                                  'of '\n",
      "                                                                  'African-American '\n",
      "                                                                  'ancestry in '\n",
      "                                                                  'the '\n",
      "                                                                  'non_neuro '\n",
      "                                                                  'subset '\n",
      "                                                                  '(afr) - '\n",
      "                                                                  'gnomAD '\n",
      "                                                                  'genomes',\n",
      "                                                   'Number': 'A',\n",
      "                                                   'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_non_neuro_AF_amr': {'Description': 'Alternate '\n",
      "                                                                  'allele '\n",
      "                                                                  'frequency '\n",
      "                                                                  'in samples '\n",
      "                                                                  'of Latino '\n",
      "                                                                  'ancestry in '\n",
      "                                                                  'the '\n",
      "                                                                  'non_neuro '\n",
      "                                                                  'subset '\n",
      "                                                                  '(amr) - '\n",
      "                                                                  'gnomAD '\n",
      "                                                                  'genomes',\n",
      "                                                   'Number': 'A',\n",
      "                                                   'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_non_neuro_AF_asj': {'Description': 'Alternate '\n",
      "                                                                  'allele '\n",
      "                                                                  'frequency '\n",
      "                                                                  'in samples '\n",
      "                                                                  'of '\n",
      "                                                                  'Ashkenazi '\n",
      "                                                                  'Jewish '\n",
      "                                                                  'ancestry in '\n",
      "                                                                  'the '\n",
      "                                                                  'non_neuro '\n",
      "                                                                  'subset '\n",
      "                                                                  '(asj) - '\n",
      "                                                                  'gnomAD '\n",
      "                                                                  'genomes',\n",
      "                                                   'Number': 'A',\n",
      "                                                   'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_non_neuro_AF_eas': {'Description': 'Alternate '\n",
      "                                                                  'allele '\n",
      "                                                                  'frequency '\n",
      "                                                                  'in samples '\n",
      "                                                                  'of East '\n",
      "                                                                  'Asian '\n",
      "                                                                  'ancestry in '\n",
      "                                                                  'the '\n",
      "                                                                  'non_neuro '\n",
      "                                                                  'subset '\n",
      "                                                                  '(eas) - '\n",
      "                                                                  'gnomAD '\n",
      "                                                                  'genomes',\n",
      "                                                   'Number': 'A',\n",
      "                                                   'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_non_neuro_AF_fin': {'Description': 'Alternate '\n",
      "                                                                  'allele '\n",
      "                                                                  'frequency '\n",
      "                                                                  'in samples '\n",
      "                                                                  'of Finnish '\n",
      "                                                                  'ancestry in '\n",
      "                                                                  'the '\n",
      "                                                                  'non_neuro '\n",
      "                                                                  'subset  '\n",
      "                                                                  '(fin) - '\n",
      "                                                                  'gnomAD '\n",
      "                                                                  'genomes',\n",
      "                                                   'Number': 'A',\n",
      "                                                   'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_non_neuro_AF_nfe': {'Description': 'Alternate '\n",
      "                                                                  'allele '\n",
      "                                                                  'frequency '\n",
      "                                                                  'in samples '\n",
      "                                                                  'of '\n",
      "                                                                  'non-Finnish '\n",
      "                                                                  'European '\n",
      "                                                                  'ancestry in '\n",
      "                                                                  'the '\n",
      "                                                                  'non_neuro '\n",
      "                                                                  'subset '\n",
      "                                                                  '(nfe) - '\n",
      "                                                                  'gnomAD '\n",
      "                                                                  'genomes',\n",
      "                                                   'Number': 'A',\n",
      "                                                   'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_non_neuro_AF_oth': {'Description': 'Alternate '\n",
      "                                                                  'allele '\n",
      "                                                                  'frequency '\n",
      "                                                                  'in samples '\n",
      "                                                                  'of '\n",
      "                                                                  'uncertain '\n",
      "                                                                  'ancestry in '\n",
      "                                                                  'the '\n",
      "                                                                  'non_neuro '\n",
      "                                                                  'subset - '\n",
      "                                                                  'gnomAD '\n",
      "                                                                  'genomes',\n",
      "                                                   'Number': 'A',\n",
      "                                                   'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_non_neuro_AF_popmax': {'Description': 'Maximum '\n",
      "                                                                     'allele '\n",
      "                                                                     'frequency '\n",
      "                                                                     'across '\n",
      "                                                                     'populations '\n",
      "                                                                     '(excluding '\n",
      "                                                                     'samples '\n",
      "                                                                     'of '\n",
      "                                                                     'Ashkenazi, '\n",
      "                                                                     'Finnish, '\n",
      "                                                                     'and '\n",
      "                                                                     'indeterminate '\n",
      "                                                                     'ancestry) '\n",
      "                                                                     'in the '\n",
      "                                                                     'non_neuro '\n",
      "                                                                     'subset - '\n",
      "                                                                     'gnomAD '\n",
      "                                                                     'genomes',\n",
      "                                                      'Number': 'A',\n",
      "                                                      'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_non_neuro_nhomalt': {'Description': 'Count of '\n",
      "                                                                   'homozygous '\n",
      "                                                                   'individuals '\n",
      "                                                                   'in samples '\n",
      "                                                                   'in the '\n",
      "                                                                   'non_neuro '\n",
      "                                                                   'subset -  '\n",
      "                                                                   'gnomAD '\n",
      "                                                                   'genomes',\n",
      "                                                    'Number': 'A',\n",
      "                                                    'Type': 'Integer'},\n",
      "          'ISEQ_GNOMAD_GENOMES_non_neuro_nhomalt_popmax': {'Description': 'Count '\n",
      "                                                                          'of '\n",
      "                                                                          'homozygous '\n",
      "                                                                          'individuals '\n",
      "                                                                          'in '\n",
      "                                                                          'the '\n",
      "                                                                          'population '\n",
      "                                                                          'with '\n",
      "                                                                          'the '\n",
      "                                                                          'maximum '\n",
      "                                                                          'allele '\n",
      "                                                                          'frequency '\n",
      "                                                                          'in '\n",
      "                                                                          'the '\n",
      "                                                                          'non_neuro '\n",
      "                                                                          'subset '\n",
      "                                                                          '- '\n",
      "                                                                          'gnomAD '\n",
      "                                                                          'genomes',\n",
      "                                                           'Number': 'A',\n",
      "                                                           'Type': 'Integer'},\n",
      "          'ISEQ_GNOMAD_GENOMES_non_neuro_popmax': {'Description': 'Population '\n",
      "                                                                  'with '\n",
      "                                                                  'maximum AF '\n",
      "                                                                  'in the '\n",
      "                                                                  'non_neuro '\n",
      "                                                                  'subset - '\n",
      "                                                                  'gnomAD '\n",
      "                                                                  'genomes',\n",
      "                                                   'Number': 'A',\n",
      "                                                   'Type': 'String'},\n",
      "          'ISEQ_GNOMAD_GENOMES_nonpar': {'Description': 'Variant (on sex '\n",
      "                                                        'chromosome) falls '\n",
      "                                                        'outside a '\n",
      "                                                        'pseudoautosomal '\n",
      "                                                        'region - gnomAD '\n",
      "                                                        'genomes',\n",
      "                                         'Number': '0',\n",
      "                                         'Type': 'Flag'},\n",
      "          'ISEQ_GNOMAD_GENOMES_popmax': {'Description': 'Population with '\n",
      "                                                        'maximum AF -  gnomAD '\n",
      "                                                        'genomes',\n",
      "                                         'Number': 'A',\n",
      "                                         'Type': 'String'},\n",
      "          'ISEQ_GNOMAD_GENOMES_popmax_AF': {'Description': 'Maximum allele '\n",
      "                                                           'frequency across '\n",
      "                                                           'populations '\n",
      "                                                           '(excluding samples '\n",
      "                                                           'of Ashkenazi, '\n",
      "                                                           'Finnish, and '\n",
      "                                                           'indeterminate '\n",
      "                                                           'ancestry) - gnomAD '\n",
      "                                                           'genomes',\n",
      "                                            'Number': 'A',\n",
      "                                            'Type': 'Float'},\n",
      "          'ISEQ_GNOMAD_GENOMES_segdup': {'Description': 'Variant falls within '\n",
      "                                                        'a segmental '\n",
      "                                                        'duplication region - '\n",
      "                                                        'gnomAD genomes',\n",
      "                                         'Number': '0',\n",
      "                                         'Type': 'Flag'},\n",
      "          'ISEQ_HIGHEST_IMPACT': {'Description': 'The highest impact of an '\n",
      "                                                 'allele.',\n",
      "                                  'Number': 'A',\n",
      "                                  'Type': 'String'},\n",
      "          'ISEQ_HPO_DISEASES': {'Description': 'Names of diseases associated '\n",
      "                                               'with genes that overlap the '\n",
      "                                               'allele (source: Human Phenotye '\n",
      "                                               'Ontology (HPO)). Names of '\n",
      "                                               'diseases associated with a '\n",
      "                                               \"given gene are '^' - \"\n",
      "                                               'delimited. Grups of diseases '\n",
      "                                               \"associated with genes are ':' \"\n",
      "                                               '- delimited. The order of '\n",
      "                                               'those groups is the same as '\n",
      "                                               'the order of genes in '\n",
      "                                               'ISEQ_GENES_NAMES field.',\n",
      "                                'Number': 'A',\n",
      "                                'Type': 'String'},\n",
      "          'ISEQ_HPO_INHERITANCE': {'Description': 'Modes of inheritance '\n",
      "                                                  'associated with genes that '\n",
      "                                                  'overlap the allele (source: '\n",
      "                                                  'Human Phenotye Ontology '\n",
      "                                                  '(HPO)). Modes of '\n",
      "                                                  'inheritance associated with '\n",
      "                                                  \"a given gene are '^' - \"\n",
      "                                                  'delimited. Grups of modes '\n",
      "                                                  'of inheritance associated '\n",
      "                                                  \"with genes are ':' - \"\n",
      "                                                  'delimited. The order of '\n",
      "                                                  'those groups is the same as '\n",
      "                                                  'the order of genes in '\n",
      "                                                  'ISEQ_GENES_NAMES field.',\n",
      "                                   'Number': 'A',\n",
      "                                   'Type': 'String'},\n",
      "          'ISEQ_HPO_PHENOTYPES': {'Description': 'Names of phenotypes '\n",
      "                                                 'associated with genes that '\n",
      "                                                 'overlap the allele (source: '\n",
      "                                                 'Human Phenotye Ontology '\n",
      "                                                 '(HPO)). Phenotypes '\n",
      "                                                 'associated with a given gene '\n",
      "                                                 \"are '^' - delimited. Grups \"\n",
      "                                                 'of phenotypes associated '\n",
      "                                                 \"with genes are ':' - \"\n",
      "                                                 'delimited. The order of '\n",
      "                                                 'those groups is the same as '\n",
      "                                                 'the order of genes in '\n",
      "                                                 'ISEQ_GENES_NAMES field.',\n",
      "                                  'Number': 'A',\n",
      "                                  'Type': 'String'},\n",
      "          'ISEQ_KG_AF': {'Description': 'Alternate allele frequency in samples '\n",
      "                                        '- 1000 Genomes',\n",
      "                         'Number': 'A',\n",
      "                         'Type': 'Float'},\n",
      "          'ISEQ_KG_AFR_AF': {'Description': 'Alternate allele frequency in '\n",
      "                                            'samples - 1000 Genomes, African '\n",
      "                                            'population (AFR)',\n",
      "                             'Number': 'A',\n",
      "                             'Type': 'Float'},\n",
      "          'ISEQ_KG_AMR_AF': {'Description': 'Alternate allele frequency in '\n",
      "                                            'samples - 1000 Genomes, Mixed '\n",
      "                                            'American population (AMR)',\n",
      "                             'Number': 'A',\n",
      "                             'Type': 'Float'},\n",
      "          'ISEQ_KG_EAS_AF': {'Description': 'Alternate allele frequency in '\n",
      "                                            'samples - 1000 Genomes, East '\n",
      "                                            'Asian population (EAS)',\n",
      "                             'Number': 'A',\n",
      "                             'Type': 'Float'},\n",
      "          'ISEQ_KG_EUR_AF': {'Description': 'Alternate allele frequency in '\n",
      "                                            'samples - 1000 Genomes, European '\n",
      "                                            'population (EUR)',\n",
      "                             'Number': 'A',\n",
      "                             'Type': 'Float'},\n",
      "          'ISEQ_KG_SAS_AF': {'Description': 'Alternate allele frequency in '\n",
      "                                            'samples - 1000 Genomes, South '\n",
      "                                            'Asian population (SAS)',\n",
      "                             'Number': 'A',\n",
      "                             'Type': 'Float'},\n",
      "          'ISEQ_M_CAP': {'Description': 'M-CAP score',\n",
      "                         'Number': 'A',\n",
      "                         'Type': 'Float'},\n",
      "          'ISEQ_M_CAP_Sensitivity': {'Description': 'Interpretable M-CAP '\n",
      "                                                    'sensitivity score. '\n",
      "                                                    'Sensitivity score  0.95 '\n",
      "                                                    'is possibly pathogenic.',\n",
      "                                     'Number': 'A',\n",
      "                                     'Type': 'Float'},\n",
      "          'ISEQ_SIFT4G': {'Description': 'SIFT4G score annotations. Format: '\n",
      "                                         'Transcript_Id | Gene_Name | Region | '\n",
      "                                         'HGVS.p | SIFT_score : (...) : '\n",
      "                                         'Transcript_Id | Gene_Name | Region | '\n",
      "                                         'HGVS.p | SIFT_score',\n",
      "                          'Number': 'A',\n",
      "                          'Type': 'String'},\n",
      "          'ISEQ_SIFT4G_MIN': {'Description': 'Minimal SIFT4G score for a '\n",
      "                                             'variant',\n",
      "                              'Number': 'A',\n",
      "                              'Type': 'Float'},\n",
      "          'ISEQ_VARIANT_CLINVAR_DISEASES': {'Description': 'Diseases '\n",
      "                                                           'associated with '\n",
      "                                                           'the variant in '\n",
      "                                                           'ClinVar. Format: '\n",
      "                                                           'disease_name : '\n",
      "                                                           '(...) : '\n",
      "                                                           'disease_name',\n",
      "                                            'Number': 'A',\n",
      "                                            'Type': 'String'},\n",
      "          'ISEQ_VARIANT_CLINVAR_DISEASES_INCLUDED': {'Description': 'Diseases '\n",
      "                                                                    'associated '\n",
      "                                                                    'with a '\n",
      "                                                                    'haplotype '\n",
      "                                                                    'or '\n",
      "                                                                    'genotype '\n",
      "                                                                    'that '\n",
      "                                                                    'includes '\n",
      "                                                                    'the '\n",
      "                                                                    'variant '\n",
      "                                                                    'in '\n",
      "                                                                    'ClinVar. '\n",
      "                                                                    'Format '\n",
      "                                                                    'disease_name '\n",
      "                                                                    ': (...) : '\n",
      "                                                                    'disease_name',\n",
      "                                                     'Number': 'A',\n",
      "                                                     'Type': 'String'},\n",
      "          'InbreedingCoeff': {'Description': 'Inbreeding coefficient as '\n",
      "                                             'estimated from the genotype '\n",
      "                                             'likelihoods per-sample when '\n",
      "                                             'compared against the '\n",
      "                                             'Hardy-Weinberg expectation',\n",
      "                              'Number': '1',\n",
      "                              'Type': 'Float'},\n",
      "          'LOF': {'Description': 'Predicted loss of function effects for this '\n",
      "                                 \"variant. Format: 'Gene_Name | Gene_ID | \"\n",
      "                                 'Number_of_transcripts_in_gene | '\n",
      "                                 \"Percent_of_transcripts_affected'\",\n",
      "                  'Number': 'A',\n",
      "                  'Type': 'String'},\n",
      "          'MITOMAP_AC': {'Description': 'Allele count in GenBank out of 49135 '\n",
      "                                        'sequences - MITOMAP',\n",
      "                         'Number': 'A',\n",
      "                         'Type': 'Float'},\n",
      "          'MITOMAP_AF': {'Description': 'Allele Frequency in Genbank - MITOMAP',\n",
      "                         'Number': 'A',\n",
      "                         'Type': 'Float'},\n",
      "          'MITOMAP_DISEASE': {'Description': 'Putative Disease Association - '\n",
      "                                             'MITOMAP. Format: disease : (...) '\n",
      "                                             ': disease',\n",
      "                              'Number': 'A',\n",
      "                              'Type': 'String'},\n",
      "          'MITOMAP_DISEASE_STATUS': {'Description': 'Disease Association '\n",
      "                                                    'Status - MITOMAP',\n",
      "                                     'Number': 'A',\n",
      "                                     'Type': 'String'},\n",
      "          'MLEAC': {'Description': 'Maximum likelihood expectation (MLE) for '\n",
      "                                   'the allele counts, for each ALT allele, in '\n",
      "                                   'the same order as listed',\n",
      "                    'Number': 'A',\n",
      "                    'Type': 'Integer'},\n",
      "          'MLEAF': {'Description': 'Maximum likelihood expectation (MLE) for '\n",
      "                                   'the allele frequency, for each ALT allele, '\n",
      "                                   'in the same order as listed',\n",
      "                    'Number': 'A',\n",
      "                    'Type': 'Float'},\n",
      "          'MQ': {'Description': 'RMS mapping quality',\n",
      "                 'Number': '1',\n",
      "                 'Type': 'Float'},\n",
      "          'MQRankSum': {'Description': 'Z-score From Wilcoxon rank sum test of '\n",
      "                                       'Alt vs. Ref read mapping qualities',\n",
      "                        'Number': '1',\n",
      "                        'Type': 'Float'},\n",
      "          'MULTIALLELIC': {'Description': 'Multiallelic site. Determined by '\n",
      "                                          'counting ALT alleles observed in '\n",
      "                                          'samples in input VCF.',\n",
      "                           'Number': '0',\n",
      "                           'Type': 'Flag'},\n",
      "          'NMD': {'Description': 'Predicted nonsense mediated decay effects '\n",
      "                                 \"for this variant. Format: 'Gene_Name | \"\n",
      "                                 'Gene_ID | Number_of_transcripts_in_gene | '\n",
      "                                 \"Percent_of_transcripts_affected'\",\n",
      "                  'Number': 'A',\n",
      "                  'Type': 'String'},\n",
      "          'PhastCons100way': {'Description': 'PhastCons100way conservation '\n",
      "                                             'score',\n",
      "                              'Number': 'A',\n",
      "                              'Type': 'Float'},\n",
      "          'PhyloP100way': {'Description': 'PhyloP100way score',\n",
      "                           'Number': 'A',\n",
      "                           'Type': 'Float'},\n",
      "          'QD': {'Description': 'Variant Confidence/Quality by Depth',\n",
      "                 'Number': '1',\n",
      "                 'Type': 'Float'},\n",
      "          'ReadPosRankSum': {'Description': 'Z-score from Wilcoxon rank sum '\n",
      "                                            'test of Alt vs. Ref read position '\n",
      "                                            'bias',\n",
      "                             'Number': '1',\n",
      "                             'Type': 'Float'},\n",
      "          'SOR': {'Description': 'Symmetric Odds Ratio of 2x2 contingency '\n",
      "                                 'table to detect strand bias',\n",
      "                  'Number': '1',\n",
      "                  'Type': 'Float'}}}\n"
     ]
    }
   ],
   "source": [
    "pprint(header) #wane: ISEQ_GNOMAD_GENOMES_V3_AF_nfe ISEQ_AGGREGATED_CLINVAR_SIGNIFICANCE ANN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------------------------------------\n",
      "Global fields:\n",
      "    None\n",
      "----------------------------------------\n",
      "Column fields:\n",
      "    's': str\n",
      "----------------------------------------\n",
      "Row fields:\n",
      "    'locus': locus<GRCh38>\n",
      "    'alleles': array<str>\n",
      "    'rsid': str\n",
      "    'qual': float64\n",
      "    'filters': set<str>\n",
      "    'info': struct {\n",
      "        AC: array<int32>, \n",
      "        AF: array<float64>, \n",
      "        AN: int32, \n",
      "        BaseQRankSum: float64, \n",
      "        ClippingRankSum: float64, \n",
      "        DB: bool, \n",
      "        DP: int32, \n",
      "        ExcessHet: float64, \n",
      "        FS: float64, \n",
      "        InbreedingCoeff: float64, \n",
      "        MLEAC: array<int32>, \n",
      "        MLEAF: array<float64>, \n",
      "        MQ: float64, \n",
      "        MQRankSum: float64, \n",
      "        QD: float64, \n",
      "        ReadPosRankSum: float64, \n",
      "        SOR: float64, \n",
      "        MULTIALLELIC: bool, \n",
      "        ALTS: array<str>, \n",
      "        ALTS_pos: int32, \n",
      "        ALTS_indices: array<str>, \n",
      "        ISEQ_GNOMAD_EXOMES_AC: array<int32>, \n",
      "        ISEQ_GNOMAD_EXOMES_AN: int32, \n",
      "        ISEQ_GNOMAD_EXOMES_AF: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_nhomalt: array<int32>, \n",
      "        ISEQ_GNOMAD_EXOMES_controls_AF: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_controls_nhomalt: array<int32>, \n",
      "        ISEQ_GNOMAD_EXOMES_non_neuro_AF: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_non_neuro_nhomalt: array<int32>, \n",
      "        ISEQ_GNOMAD_EXOMES_non_cancer_AF: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_non_cancer_nhomalt: array<int32>, \n",
      "        ISEQ_GNOMAD_EXOMES_popmax: array<str>, \n",
      "        ISEQ_GNOMAD_EXOMES_popmax_AF: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_nhomalt_popmax: array<int32>, \n",
      "        ISEQ_GNOMAD_EXOMES_controls_popmax: array<str>, \n",
      "        ISEQ_GNOMAD_EXOMES_controls_AF_popmax: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_controls_nhomalt_popmax: array<int32>, \n",
      "        ISEQ_GNOMAD_EXOMES_non_neuro_popmax: array<str>, \n",
      "        ISEQ_GNOMAD_EXOMES_non_neuro_AF_popmax: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_non_neuro_nhomalt_popmax: array<int32>, \n",
      "        ISEQ_GNOMAD_EXOMES_non_cancer_popmax: array<str>, \n",
      "        ISEQ_GNOMAD_EXOMES_non_cancer_AF_popmax: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_non_cancer_nhomalt_popmax: array<int32>, \n",
      "        ISEQ_GNOMAD_EXOMES_AF_afr: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_controls_AF_afr: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_non_neuro_AF_afr: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_non_cancer_AF_afr: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_AF_amr: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_controls_AF_amr: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_non_neuro_AF_amr: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_non_cancer_AF_amr: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_AF_asj: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_controls_AF_asj: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_non_neuro_AF_asj: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_non_cancer_AF_asj: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_AF_eas: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_controls_AF_eas: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_non_neuro_AF_eas: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_non_cancer_AF_eas: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_AF_eas_subpopulations: array<str>, \n",
      "        ISEQ_GNOMAD_EXOMES_AF_fin: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_controls_AF_fin: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_non_neuro_AF_fin: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_non_cancer_AF_fin: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_AF_nfe: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_controls_AF_nfe: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_non_neuro_AF_nfe: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_non_cancer_AF_nfe: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_AF_nfe_subpopulations: array<str>, \n",
      "        ISEQ_GNOMAD_EXOMES_AF_sas: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_controls_AF_sas: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_non_neuro_AF_sas: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_non_cancer_AF_sas: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_AF_oth: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_controls_AF_oth: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_non_neuro_AF_oth: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_non_cancer_AF_oth: array<float64>, \n",
      "        ISEQ_GNOMAD_EXOMES_segdup: bool, \n",
      "        ISEQ_GNOMAD_EXOMES_lcr: bool, \n",
      "        ISEQ_GNOMAD_EXOMES_decoy: bool, \n",
      "        ISEQ_GNOMAD_EXOMES_nonpar: bool, \n",
      "        ISEQ_GNOMAD_EXOMES_FILTER_STATUS: array<str>, \n",
      "        ISEQ_GNOMAD_GENOMES_AC: array<int32>, \n",
      "        ISEQ_GNOMAD_GENOMES_AN: int32, \n",
      "        ISEQ_GNOMAD_GENOMES_AF: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_nhomalt: array<int32>, \n",
      "        ISEQ_GNOMAD_GENOMES_controls_AF: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_controls_nhomalt: array<int32>, \n",
      "        ISEQ_GNOMAD_GENOMES_non_neuro_AF: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_non_neuro_nhomalt: array<int32>, \n",
      "        ISEQ_GNOMAD_GENOMES_popmax: array<str>, \n",
      "        ISEQ_GNOMAD_GENOMES_popmax_AF: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_nhomalt_popmax: array<int32>, \n",
      "        ISEQ_GNOMAD_GENOMES_controls_popmax: array<str>, \n",
      "        ISEQ_GNOMAD_GENOMES_controls_AF_popmax: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_controls_nhomalt_popmax: array<int32>, \n",
      "        ISEQ_GNOMAD_GENOMES_non_neuro_popmax: array<str>, \n",
      "        ISEQ_GNOMAD_GENOMES_non_neuro_AF_popmax: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_non_neuro_nhomalt_popmax: array<int32>, \n",
      "        ISEQ_GNOMAD_GENOMES_AF_afr: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_controls_AF_afr: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_non_neuro_AF_afr: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_AF_amr: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_controls_AF_amr: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_non_neuro_AF_amr: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_AF_asj: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_controls_AF_asj: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_non_neuro_AF_asj: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_AF_eas: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_controls_AF_eas: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_non_neuro_AF_eas: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_AF_fin: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_controls_AF_fin: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_non_neuro_AF_fin: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_AF_nfe: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_controls_AF_nfe: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_non_neuro_AF_nfe: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_AF_nfe_subpopulations: array<str>, \n",
      "        ISEQ_GNOMAD_GENOMES_AF_oth: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_controls_AF_oth: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_non_neuro_AF_oth: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_segdup: bool, \n",
      "        ISEQ_GNOMAD_GENOMES_lcr: bool, \n",
      "        ISEQ_GNOMAD_GENOMES_decoy: bool, \n",
      "        ISEQ_GNOMAD_GENOMES_nonpar: bool, \n",
      "        ISEQ_GNOMAD_GENOMES_FILTER_STATUS: array<str>, \n",
      "        ISEQ_GNOMAD_GENOMES_V3_AC: array<int32>, \n",
      "        ISEQ_GNOMAD_GENOMES_V3_AN: int32, \n",
      "        ISEQ_GNOMAD_GENOMES_V3_AF: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_V3_nhomalt: array<int32>, \n",
      "        ISEQ_GNOMAD_GENOMES_V3_popmax: array<str>, \n",
      "        ISEQ_GNOMAD_GENOMES_V3_popmax_AF: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_V3_nhomalt_popmax: array<str>, \n",
      "        ISEQ_GNOMAD_GENOMES_V3_AF_ami: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_V3_AF_afr: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_V3_AF_amr: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_V3_AF_asj: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_V3_AF_eas: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_V3_AF_fin: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_V3_AF_nfe: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_V3_AF_sas: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_V3_AF_oth: array<float64>, \n",
      "        ISEQ_GNOMAD_GENOMES_V3_lcr: bool, \n",
      "        ISEQ_GNOMAD_GENOMES_V3_nonpar: bool, \n",
      "        ISEQ_GNOMAD_GENOMES_V3_FILTER_STATUS: array<str>, \n",
      "        ISEQ_EXAC_AC: array<float64>, \n",
      "        ISEQ_EXAC_AN: int32, \n",
      "        ISEQ_EXAC_AF: array<float64>, \n",
      "        ISEQ_EXAC_POPMAX: array<str>, \n",
      "        ISEQ_EXAC_POPMAX_AF: array<float64>, \n",
      "        ISEQ_EXAC_AFR_AF: array<float64>, \n",
      "        ISEQ_EXAC_AMR_AF: array<float64>, \n",
      "        ISEQ_EXAC_EAS_AF: array<float64>, \n",
      "        ISEQ_EXAC_FIN_AF: array<float64>, \n",
      "        ISEQ_EXAC_NFE_AF: array<float64>, \n",
      "        ISEQ_EXAC_OTH_AF: array<float64>, \n",
      "        ISEQ_EXAC_SAS_AF: array<float64>, \n",
      "        ISEQ_EXAC_FILTER_STATUS: array<str>, \n",
      "        ISEQ_KG_AF: array<float64>, \n",
      "        ISEQ_KG_EAS_AF: array<float64>, \n",
      "        ISEQ_KG_EUR_AF: array<float64>, \n",
      "        ISEQ_KG_AFR_AF: array<float64>, \n",
      "        ISEQ_KG_AMR_AF: array<float64>, \n",
      "        ISEQ_KG_SAS_AF: array<float64>, \n",
      "        ISEQ_ESP6500_EA_AF: array<float64>, \n",
      "        ISEQ_ESP6500_AA_AF: array<float64>, \n",
      "        ISEQ_ESP6500_AN: int32, \n",
      "        ISEQ_ESP6500_AF: array<float64>, \n",
      "        MITOMAP_AC: array<float64>, \n",
      "        MITOMAP_AF: array<float64>, \n",
      "        ANN: array<str>, \n",
      "        ISEQ_GENES_NAMES: array<str>, \n",
      "        ISEQ_HIGHEST_IMPACT: array<str>, \n",
      "        LOF: array<str>, \n",
      "        NMD: array<str>, \n",
      "        ISEQ_DBSNP_RS: array<str>, \n",
      "        ISEQ_M_CAP: array<float64>, \n",
      "        ISEQ_M_CAP_Sensitivity: array<float64>, \n",
      "        ISEQ_SIFT4G_MIN: array<float64>, \n",
      "        ISEQ_SIFT4G: array<str>, \n",
      "        ISEQ_GERP: array<float64>, \n",
      "        PhastCons100way: array<float64>, \n",
      "        PhyloP100way: array<float64>, \n",
      "        ISEQ_CLINVAR_ALLELE_ID: array<str>, \n",
      "        ISEQ_CLINVAR_VARIATION_ID: array<str>, \n",
      "        ISEQ_VARIANT_CLINVAR_DISEASES: array<str>, \n",
      "        ISEQ_VARIANT_CLINVAR_DISEASES_INCLUDED: array<str>, \n",
      "        ISEQ_AGGREGATED_CLINVAR_SIGNIFICANCE: array<str>, \n",
      "        ISEQ_CLINVAR_SIGNIFICANCE: array<str>, \n",
      "        ISEQ_AGGREGATED_CLINVAR_SIGNIFICANCE_INCLUDED: array<str>, \n",
      "        ISEQ_CLINVAR_SIGNIFICANCE_INCLUDED: array<str>, \n",
      "        ISEQ_CLINVAR_REVIEW_STATUS: array<str>, \n",
      "        MITOMAP_DISEASE: array<str>, \n",
      "        MITOMAP_DISEASE_STATUS: array<str>, \n",
      "        ISEQ_CLINVAR_DISEASES: array<str>, \n",
      "        ISEQ_HPO_DISEASES: array<str>, \n",
      "        ISEQ_HPO_PHENOTYPES: array<str>, \n",
      "        ISEQ_HPO_INHERITANCE: array<str>\n",
      "    }\n",
      "----------------------------------------\n",
      "Entry fields:\n",
      "    'AD': array<int32>\n",
      "    'DP': int32\n",
      "    'GQ': int32\n",
      "    'GT': call\n",
      "    'PGT': call\n",
      "    'PID': str\n",
      "    'PL': array<int32>\n",
      "----------------------------------------\n",
      "Column key: ['s']\n",
      "Row key: ['locus', 'alleles']\n",
      "----------------------------------------\n"
     ]
    }
   ],
   "source": [
    "mt.describe() #see whats inside (AFAllele Frequency for each ALT allele, in the same order as listed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table><thead style=\"font-weight: bold;\"><tr><td>locus</td><td>alleles</td></tr>\n",
       "<tr><td>locus&lt;GRCh38&gt;</td><td>array&lt;str&gt;</td></tr>\n",
       "</thead><tbody><tr><td>chr1:10146</td><td>[&quot;AC&quot;,&quot;*&quot;]</td></tr>\n",
       "<tr><td>chr1:10177</td><td>[&quot;A&quot;,&quot;*&quot;]</td></tr>\n",
       "<tr><td>chr1:10622</td><td>[&quot;T&quot;,&quot;*&quot;]</td></tr>\n",
       "<tr><td>chr1:10623</td><td>[&quot;T&quot;,&quot;*&quot;]</td></tr>\n",
       "<tr><td>chr1:30923</td><td>[&quot;G&quot;,&quot;*&quot;]</td></tr>\n",
       "</tbody></table><p style=\"background: #fdd; padding: 0.4em;\">showing top 5 rows</p>\n"
      ],
      "text/plain": [
       "+---------------+------------+\n",
       "| locus         | alleles    |\n",
       "+---------------+------------+\n",
       "| locus<GRCh38> | array<str> |\n",
       "+---------------+------------+\n",
       "| chr1:10146    | [\"AC\",\"*\"] |\n",
       "| chr1:10177    | [\"A\",\"*\"]  |\n",
       "| chr1:10622    | [\"T\",\"*\"]  |\n",
       "| chr1:10623    | [\"T\",\"*\"]  |\n",
       "| chr1:30923    | [\"G\",\"*\"]  |\n",
       "+---------------+------------+\n",
       "showing top 5 rows"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mt.rows().select().show(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='2.1'></a>\n",
    "\n",
    "## 2.1 Removing the star alleles\n",
    "\n",
    "###These are orphaned stars and shouldn't be here\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 363,
   "metadata": {},
   "outputs": [],
   "source": [
    "mt = mt.filter_rows(mt.alleles.contains('*'), keep = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 364,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "151"
      ]
     },
     "execution_count": 364,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mt.count_cols()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 365,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "47373"
      ]
     },
     "execution_count": 365,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mt.count_rows()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can use rows along with select to pull out 5 variants. The select method takes either a string refering to a field name in the table, or a Hail Expression. Here, we leave the arguments blank to keep only the row key fields, locus and alleles."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 366,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table><thead style=\"font-weight: bold;\"><tr><td>locus</td><td>alleles</td></tr>\n",
       "<tr><td>locus&lt;GRCh38&gt;</td><td>array&lt;str&gt;</td></tr>\n",
       "</thead><tbody><tr><td>chr1:69968</td><td>[&quot;A&quot;,&quot;G&quot;]</td></tr>\n",
       "<tr><td>chr1:183189</td><td>[&quot;G&quot;,&quot;C&quot;]</td></tr>\n",
       "<tr><td>chr1:183238</td><td>[&quot;G&quot;,&quot;C&quot;]</td></tr>\n",
       "<tr><td>chr1:183937</td><td>[&quot;G&quot;,&quot;A&quot;]</td></tr>\n",
       "<tr><td>chr1:184994</td><td>[&quot;G&quot;,&quot;C&quot;]</td></tr>\n",
       "</tbody></table><p style=\"background: #fdd; padding: 0.4em;\">showing top 5 rows</p>\n",
       "<p style=\"background: #fdd; padding: 0.4em;\">showing the first 0 of 151 columns</p>\n"
      ],
      "text/plain": [
       "+---------------+------------+\n",
       "| locus         | alleles    |\n",
       "+---------------+------------+\n",
       "| locus<GRCh38> | array<str> |\n",
       "+---------------+------------+\n",
       "| chr1:69968    | [\"A\",\"G\"]  |\n",
       "| chr1:183189   | [\"G\",\"C\"]  |\n",
       "| chr1:183238   | [\"G\",\"C\"]  |\n",
       "| chr1:183937   | [\"G\",\"A\"]  |\n",
       "| chr1:184994   | [\"G\",\"C\"]  |\n",
       "+---------------+------------+\n",
       "showing top 5 rows\n",
       "showing the first 0 of 151 columns"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mt.entry.show(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 367,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Struct(AD=[7, 0], DP=7, GQ=21, GT=Call(alleles=[0, 0], phased=False), PGT=None, PID=None, PL=[0, 21, 239]),\n",
       " Struct(AD=[0, 0], DP=0, GQ=None, GT=None, PGT=None, PID=None, PL=None),\n",
       " Struct(AD=[0, 0], DP=0, GQ=None, GT=None, PGT=None, PID=None, PL=None),\n",
       " Struct(AD=[0, 0], DP=0, GQ=None, GT=None, PGT=None, PID=None, PL=None),\n",
       " Struct(AD=[6, 0], DP=6, GQ=18, GT=Call(alleles=[0, 0], phased=False), PGT=None, PID=None, PL=[0, 18, 173])]"
      ]
     },
     "execution_count": 367,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mt.entry.take(5) #create a list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 368,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Struct(AD=[10, 0], DP=10, GQ=30, GT=Call(alleles=[0, 0], phased=False), PGT=None, PID=None, PL=[0, 30, 327])"
      ]
     },
     "execution_count": 368,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mt.entry.take(200)[130] #but works slow!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table><thead style=\"font-weight: bold;\"><tr><td>s</td></tr>\n",
       "<tr><td>str</td></tr>\n",
       "</thead><tbody><tr><td>&quot;S_136&quot;</td></tr>\n",
       "<tr><td>&quot;S_170c&quot;</td></tr>\n",
       "<tr><td>&quot;S_170d&quot;</td></tr>\n",
       "<tr><td>&quot;S_6981&quot;</td></tr>\n",
       "<tr><td>&quot;S_6982&quot;</td></tr>\n",
       "</tbody></table><p style=\"background: #fdd; padding: 0.4em;\">showing top 5 rows</p>\n"
      ],
      "text/plain": [
       "+----------+\n",
       "| s        |\n",
       "+----------+\n",
       "| str      |\n",
       "+----------+\n",
       "| \"S_136\"  |\n",
       "| \"S_170c\" |\n",
       "| \"S_170d\" |\n",
       "| \"S_6981\" |\n",
       "| \"S_6982\" |\n",
       "+----------+\n",
       "showing top 5 rows"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mt.s.show(5) #samples\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--------------------------------------------------------\n",
      "Type:\n",
      "        str\n",
      "--------------------------------------------------------\n",
      "Source:\n",
      "    <hail.matrixtable.MatrixTable object at 0x7f93c162d410>\n",
      "Index:\n",
      "    ['column']\n",
      "--------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "mt.s.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 442,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['S_136', 'S_170c', 'S_170d', 'S_6981', 'S_6982']"
      ]
     },
     "execution_count": 442,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mt.s.take(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 369,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table><thead style=\"font-weight: bold;\"><tr><td>locus</td><td>alleles</td><td>S_136.GT</td><td>S_170c.GT</td><td>S_170d.GT</td><td>S_6981.GT</td></tr>\n",
       "<tr><td>locus&lt;GRCh38&gt;</td><td>array&lt;str&gt;</td><td>call</td><td>call</td><td>call</td><td>call</td></tr>\n",
       "</thead><tbody><tr><td>chr1:69968</td><td>[&quot;A&quot;,&quot;G&quot;]</td><td>0/0</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "<tr><td>chr1:183189</td><td>[&quot;G&quot;,&quot;C&quot;]</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td></tr>\n",
       "<tr><td>chr1:183238</td><td>[&quot;G&quot;,&quot;C&quot;]</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td></tr>\n",
       "<tr><td>chr1:183937</td><td>[&quot;G&quot;,&quot;A&quot;]</td><td>0/0</td><td>0/1</td><td>0/1</td><td>0/0</td></tr>\n",
       "<tr><td>chr1:184994</td><td>[&quot;G&quot;,&quot;C&quot;]</td><td>0/0</td><td>0/0</td><td>NA</td><td>0/0</td></tr>\n",
       "<tr><td>chr1:185336</td><td>[&quot;C&quot;,&quot;T&quot;]</td><td>0/0</td><td>0/1</td><td>0/0</td><td>0/0</td></tr>\n",
       "<tr><td>chr1:185497</td><td>[&quot;G&quot;,&quot;A&quot;]</td><td>0/0</td><td>0/0</td><td>0/1</td><td>0/0</td></tr>\n",
       "<tr><td>chr1:185550</td><td>[&quot;G&quot;,&quot;A&quot;]</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td></tr>\n",
       "<tr><td>chr1:186338</td><td>[&quot;T&quot;,&quot;G&quot;]</td><td>0/1</td><td>0/0</td><td>0/1</td><td>0/1</td></tr>\n",
       "<tr><td>chr1:186341</td><td>[&quot;T&quot;,&quot;G&quot;]</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td></tr>\n",
       "</tbody></table><p style=\"background: #fdd; padding: 0.4em;\">showing top 10 rows</p>\n",
       "<p style=\"background: #fdd; padding: 0.4em;\">showing the first 4 of 151 columns</p>\n"
      ],
      "text/plain": [
       "+---------------+------------+----------+-----------+-----------+-----------+\n",
       "| locus         | alleles    | S_136.GT | S_170c.GT | S_170d.GT | S_6981.GT |\n",
       "+---------------+------------+----------+-----------+-----------+-----------+\n",
       "| locus<GRCh38> | array<str> | call     | call      | call      | call      |\n",
       "+---------------+------------+----------+-----------+-----------+-----------+\n",
       "| chr1:69968    | [\"A\",\"G\"]  | 0/0      | NA        | NA        | NA        |\n",
       "| chr1:183189   | [\"G\",\"C\"]  | 0/0      | 0/0       | 0/0       | 0/0       |\n",
       "| chr1:183238   | [\"G\",\"C\"]  | 0/0      | 0/0       | 0/0       | 0/0       |\n",
       "| chr1:183937   | [\"G\",\"A\"]  | 0/0      | 0/1       | 0/1       | 0/0       |\n",
       "| chr1:184994   | [\"G\",\"C\"]  | 0/0      | 0/0       | NA        | 0/0       |\n",
       "| chr1:185336   | [\"C\",\"T\"]  | 0/0      | 0/1       | 0/0       | 0/0       |\n",
       "| chr1:185497   | [\"G\",\"A\"]  | 0/0      | 0/0       | 0/1       | 0/0       |\n",
       "| chr1:185550   | [\"G\",\"A\"]  | 0/0      | 0/0       | 0/0       | 0/0       |\n",
       "| chr1:186338   | [\"T\",\"G\"]  | 0/1      | 0/0       | 0/1       | 0/1       |\n",
       "| chr1:186341   | [\"T\",\"G\"]  | 0/0      | 0/0       | 0/0       | 0/0       |\n",
       "+---------------+------------+----------+-----------+-----------+-----------+\n",
       "showing top 10 rows\n",
       "showing the first 4 of 151 columns"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mt.GT.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Struct(AC=[2], AF=[0.083], AN=24, BaseQRankSum=-0.27, ClippingRankSum=-0.0, DB=True, DP=8231, ExcessHet=3.9794, FS=0.0, InbreedingCoeff=0.1448, MLEAC=[3], MLEAF=[0.125], MQ=1.84, MQRankSum=-0.0, QD=30.0, ReadPosRankSum=2.37, SOR=1.061, MULTIALLELIC=True, ALTS=['C', '*'], ALTS_pos=10177, ALTS_indices=['2'], ISEQ_GNOMAD_EXOMES_AC=None, ISEQ_GNOMAD_EXOMES_AN=None, ISEQ_GNOMAD_EXOMES_AF=None, ISEQ_GNOMAD_EXOMES_nhomalt=None, ISEQ_GNOMAD_EXOMES_controls_AF=None, ISEQ_GNOMAD_EXOMES_controls_nhomalt=None, ISEQ_GNOMAD_EXOMES_non_neuro_AF=None, ISEQ_GNOMAD_EXOMES_non_neuro_nhomalt=None, ISEQ_GNOMAD_EXOMES_non_cancer_AF=None, ISEQ_GNOMAD_EXOMES_non_cancer_nhomalt=None, ISEQ_GNOMAD_EXOMES_popmax=None, ISEQ_GNOMAD_EXOMES_popmax_AF=None, ISEQ_GNOMAD_EXOMES_nhomalt_popmax=None, ISEQ_GNOMAD_EXOMES_controls_popmax=None, ISEQ_GNOMAD_EXOMES_controls_AF_popmax=None, ISEQ_GNOMAD_EXOMES_controls_nhomalt_popmax=None, ISEQ_GNOMAD_EXOMES_non_neuro_popmax=None, ISEQ_GNOMAD_EXOMES_non_neuro_AF_popmax=None, ISEQ_GNOMAD_EXOMES_non_neuro_nhomalt_popmax=None, ISEQ_GNOMAD_EXOMES_non_cancer_popmax=None, ISEQ_GNOMAD_EXOMES_non_cancer_AF_popmax=None, ISEQ_GNOMAD_EXOMES_non_cancer_nhomalt_popmax=None, ISEQ_GNOMAD_EXOMES_AF_afr=None, ISEQ_GNOMAD_EXOMES_controls_AF_afr=None, ISEQ_GNOMAD_EXOMES_non_neuro_AF_afr=None, ISEQ_GNOMAD_EXOMES_non_cancer_AF_afr=None, ISEQ_GNOMAD_EXOMES_AF_amr=None, ISEQ_GNOMAD_EXOMES_controls_AF_amr=None, ISEQ_GNOMAD_EXOMES_non_neuro_AF_amr=None, ISEQ_GNOMAD_EXOMES_non_cancer_AF_amr=None, ISEQ_GNOMAD_EXOMES_AF_asj=None, ISEQ_GNOMAD_EXOMES_controls_AF_asj=None, ISEQ_GNOMAD_EXOMES_non_neuro_AF_asj=None, ISEQ_GNOMAD_EXOMES_non_cancer_AF_asj=None, ISEQ_GNOMAD_EXOMES_AF_eas=None, ISEQ_GNOMAD_EXOMES_controls_AF_eas=None, ISEQ_GNOMAD_EXOMES_non_neuro_AF_eas=None, ISEQ_GNOMAD_EXOMES_non_cancer_AF_eas=None, ISEQ_GNOMAD_EXOMES_AF_eas_subpopulations=None, ISEQ_GNOMAD_EXOMES_AF_fin=None, ISEQ_GNOMAD_EXOMES_controls_AF_fin=None, ISEQ_GNOMAD_EXOMES_non_neuro_AF_fin=None, ISEQ_GNOMAD_EXOMES_non_cancer_AF_fin=None, ISEQ_GNOMAD_EXOMES_AF_nfe=None, ISEQ_GNOMAD_EXOMES_controls_AF_nfe=None, ISEQ_GNOMAD_EXOMES_non_neuro_AF_nfe=None, ISEQ_GNOMAD_EXOMES_non_cancer_AF_nfe=None, ISEQ_GNOMAD_EXOMES_AF_nfe_subpopulations=None, ISEQ_GNOMAD_EXOMES_AF_sas=None, ISEQ_GNOMAD_EXOMES_controls_AF_sas=None, ISEQ_GNOMAD_EXOMES_non_neuro_AF_sas=None, ISEQ_GNOMAD_EXOMES_non_cancer_AF_sas=None, ISEQ_GNOMAD_EXOMES_AF_oth=None, ISEQ_GNOMAD_EXOMES_controls_AF_oth=None, ISEQ_GNOMAD_EXOMES_non_neuro_AF_oth=None, ISEQ_GNOMAD_EXOMES_non_cancer_AF_oth=None, ISEQ_GNOMAD_EXOMES_segdup=False, ISEQ_GNOMAD_EXOMES_lcr=False, ISEQ_GNOMAD_EXOMES_decoy=False, ISEQ_GNOMAD_EXOMES_nonpar=False, ISEQ_GNOMAD_EXOMES_FILTER_STATUS=None, ISEQ_GNOMAD_GENOMES_AC=None, ISEQ_GNOMAD_GENOMES_AN=None, ISEQ_GNOMAD_GENOMES_AF=None, ISEQ_GNOMAD_GENOMES_nhomalt=None, ISEQ_GNOMAD_GENOMES_controls_AF=None, ISEQ_GNOMAD_GENOMES_controls_nhomalt=None, ISEQ_GNOMAD_GENOMES_non_neuro_AF=None, ISEQ_GNOMAD_GENOMES_non_neuro_nhomalt=None, ISEQ_GNOMAD_GENOMES_popmax=None, ISEQ_GNOMAD_GENOMES_popmax_AF=None, ISEQ_GNOMAD_GENOMES_nhomalt_popmax=None, ISEQ_GNOMAD_GENOMES_controls_popmax=None, ISEQ_GNOMAD_GENOMES_controls_AF_popmax=None, ISEQ_GNOMAD_GENOMES_controls_nhomalt_popmax=None, ISEQ_GNOMAD_GENOMES_non_neuro_popmax=None, ISEQ_GNOMAD_GENOMES_non_neuro_AF_popmax=None, ISEQ_GNOMAD_GENOMES_non_neuro_nhomalt_popmax=None, ISEQ_GNOMAD_GENOMES_AF_afr=None, ISEQ_GNOMAD_GENOMES_controls_AF_afr=None, ISEQ_GNOMAD_GENOMES_non_neuro_AF_afr=None, ISEQ_GNOMAD_GENOMES_AF_amr=None, ISEQ_GNOMAD_GENOMES_controls_AF_amr=None, ISEQ_GNOMAD_GENOMES_non_neuro_AF_amr=None, ISEQ_GNOMAD_GENOMES_AF_asj=None, ISEQ_GNOMAD_GENOMES_controls_AF_asj=None, ISEQ_GNOMAD_GENOMES_non_neuro_AF_asj=None, ISEQ_GNOMAD_GENOMES_AF_eas=None, ISEQ_GNOMAD_GENOMES_controls_AF_eas=None, ISEQ_GNOMAD_GENOMES_non_neuro_AF_eas=None, ISEQ_GNOMAD_GENOMES_AF_fin=None, ISEQ_GNOMAD_GENOMES_controls_AF_fin=None, ISEQ_GNOMAD_GENOMES_non_neuro_AF_fin=None, ISEQ_GNOMAD_GENOMES_AF_nfe=None, ISEQ_GNOMAD_GENOMES_controls_AF_nfe=None, ISEQ_GNOMAD_GENOMES_non_neuro_AF_nfe=None, ISEQ_GNOMAD_GENOMES_AF_nfe_subpopulations=None, ISEQ_GNOMAD_GENOMES_AF_oth=None, ISEQ_GNOMAD_GENOMES_controls_AF_oth=None, ISEQ_GNOMAD_GENOMES_non_neuro_AF_oth=None, ISEQ_GNOMAD_GENOMES_segdup=False, ISEQ_GNOMAD_GENOMES_lcr=False, ISEQ_GNOMAD_GENOMES_decoy=False, ISEQ_GNOMAD_GENOMES_nonpar=False, ISEQ_GNOMAD_GENOMES_FILTER_STATUS=None, ISEQ_GNOMAD_GENOMES_V3_AC=None, ISEQ_GNOMAD_GENOMES_V3_AN=None, ISEQ_GNOMAD_GENOMES_V3_AF=None, ISEQ_GNOMAD_GENOMES_V3_nhomalt=None, ISEQ_GNOMAD_GENOMES_V3_popmax=None, ISEQ_GNOMAD_GENOMES_V3_popmax_AF=None, ISEQ_GNOMAD_GENOMES_V3_nhomalt_popmax=None, ISEQ_GNOMAD_GENOMES_V3_AF_ami=None, ISEQ_GNOMAD_GENOMES_V3_AF_afr=None, ISEQ_GNOMAD_GENOMES_V3_AF_amr=None, ISEQ_GNOMAD_GENOMES_V3_AF_asj=None, ISEQ_GNOMAD_GENOMES_V3_AF_eas=None, ISEQ_GNOMAD_GENOMES_V3_AF_fin=None, ISEQ_GNOMAD_GENOMES_V3_AF_nfe=None, ISEQ_GNOMAD_GENOMES_V3_AF_sas=None, ISEQ_GNOMAD_GENOMES_V3_AF_oth=None, ISEQ_GNOMAD_GENOMES_V3_lcr=False, ISEQ_GNOMAD_GENOMES_V3_nonpar=False, ISEQ_GNOMAD_GENOMES_V3_FILTER_STATUS=None, ISEQ_EXAC_AC=None, ISEQ_EXAC_AN=None, ISEQ_EXAC_AF=None, ISEQ_EXAC_POPMAX=None, ISEQ_EXAC_POPMAX_AF=None, ISEQ_EXAC_AFR_AF=None, ISEQ_EXAC_AMR_AF=None, ISEQ_EXAC_EAS_AF=None, ISEQ_EXAC_FIN_AF=None, ISEQ_EXAC_NFE_AF=None, ISEQ_EXAC_OTH_AF=None, ISEQ_EXAC_SAS_AF=None, ISEQ_EXAC_FILTER_STATUS=None, ISEQ_KG_AF=None, ISEQ_KG_EAS_AF=None, ISEQ_KG_EUR_AF=None, ISEQ_KG_AFR_AF=None, ISEQ_KG_AMR_AF=None, ISEQ_KG_SAS_AF=None, ISEQ_ESP6500_EA_AF=None, ISEQ_ESP6500_AA_AF=None, ISEQ_ESP6500_AN=None, ISEQ_ESP6500_AF=None, MITOMAP_AC=None, MITOMAP_AF=None, ANN=None, ISEQ_GENES_NAMES=None, ISEQ_HIGHEST_IMPACT=None, LOF=None, NMD=None, ISEQ_DBSNP_RS=None, ISEQ_M_CAP=None, ISEQ_M_CAP_Sensitivity=None, ISEQ_SIFT4G_MIN=None, ISEQ_SIFT4G=None, ISEQ_GERP=None, PhastCons100way=None, PhyloP100way=None, ISEQ_CLINVAR_ALLELE_ID=None, ISEQ_CLINVAR_VARIATION_ID=None, ISEQ_VARIANT_CLINVAR_DISEASES=None, ISEQ_VARIANT_CLINVAR_DISEASES_INCLUDED=None, ISEQ_AGGREGATED_CLINVAR_SIGNIFICANCE=None, ISEQ_CLINVAR_SIGNIFICANCE=None, ISEQ_AGGREGATED_CLINVAR_SIGNIFICANCE_INCLUDED=None, ISEQ_CLINVAR_SIGNIFICANCE_INCLUDED=None, ISEQ_CLINVAR_REVIEW_STATUS=None, MITOMAP_DISEASE=None, MITOMAP_DISEASE_STATUS=None, ISEQ_CLINVAR_DISEASES=None, ISEQ_HPO_DISEASES=None, ISEQ_HPO_PHENOTYPES=None, ISEQ_HPO_INHERITANCE=None)"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mt.info.take(2)[1]   #print an example info for a vartiant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 371,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--------------------------------------------------------\n",
      "Type:\n",
      "        array<str>\n",
      "--------------------------------------------------------\n",
      "Source:\n",
      "    <hail.matrixtable.MatrixTable object at 0x7f939ceb6490>\n",
      "Index:\n",
      "    ['row']\n",
      "--------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "mt.info.ISEQ_CLINVAR_REVIEW_STATUS.describe() #describe a field"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 372,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<p><b>Variant summary:</b></p><ul><li><p>Total variants: 47373</p></li><li><p>Alleles per variant:</p><table><thead style=\"font-weight: bold;\"><tr><th>Number of alleles</th><th>Count</th></tr></thead><tbody><tr><td>2</td><td>47373</td></tr></tbody></table></li><li><p>Counts by allele type:</p><table><thead style=\"font-weight: bold;\"><tr><th>Allele type</th><th>Count</th></tr></thead><tbody><tr><td>SNP</td><td>41030</td></tr><tr><td>Deletion</td><td>3351</td></tr><tr><td>Insertion</td><td>2992</td></tr></tbody></table></li><li><p>Variants per contig:</p><table><thead style=\"font-weight: bold;\"><tr><th>Contig</th><th>Count</th></tr></thead><tbody><tr><td>chr1</td><td>4499</td></tr><tr><td>chr2</td><td>2769</td></tr><tr><td>chr3</td><td>2436</td></tr><tr><td>chr4</td><td>1549</td></tr><tr><td>chr5</td><td>2578</td></tr><tr><td>chr6</td><td>1900</td></tr><tr><td>chr7</td><td>2886</td></tr><tr><td>chr8</td><td>1458</td></tr><tr><td>chr9</td><td>1812</td></tr><tr><td>chr10</td><td>1567</td></tr><tr><td>chr11</td><td>2776</td></tr><tr><td>chr12</td><td>2232</td></tr><tr><td>chr13</td><td>636</td></tr><tr><td>chr14</td><td>1607</td></tr><tr><td>chr15</td><td>1399</td></tr><tr><td>chr16</td><td>2154</td></tr><tr><td>chr17</td><td>6266</td></tr><tr><td>chr18</td><td>556</td></tr><tr><td>chr19</td><td>3145</td></tr><tr><td>chr20</td><td>850</td></tr><tr><td>chr21</td><td>394</td></tr><tr><td>chr22</td><td>797</td></tr><tr><td>chrX</td><td>1016</td></tr><tr><td>chrY</td><td>16</td></tr><tr><td>chrM</td><td>75</td></tr></tbody></table></li></ul>"
      ],
      "text/plain": [
       "==============================\n",
       "Number of variants: 47373\n",
       "==============================\n",
       "Alleles per variant\n",
       "-------------------\n",
       "  2 alleles: 47373 variants\n",
       "==============================\n",
       "Variants per contig\n",
       "-------------------\n",
       "   chr1: 4499 variants\n",
       "   chr2: 2769 variants\n",
       "   chr3: 2436 variants\n",
       "   chr4: 1549 variants\n",
       "   chr5: 2578 variants\n",
       "   chr6: 1900 variants\n",
       "   chr7: 2886 variants\n",
       "   chr8: 1458 variants\n",
       "   chr9: 1812 variants\n",
       "  chr10: 1567 variants\n",
       "  chr11: 2776 variants\n",
       "  chr12: 2232 variants\n",
       "  chr13: 636 variants\n",
       "  chr14: 1607 variants\n",
       "  chr15: 1399 variants\n",
       "  chr16: 2154 variants\n",
       "  chr17: 6266 variants\n",
       "  chr18: 556 variants\n",
       "  chr19: 3145 variants\n",
       "  chr20: 850 variants\n",
       "  chr21: 394 variants\n",
       "  chr22: 797 variants\n",
       "   chrX: 1016 variants\n",
       "   chrY: 16 variants\n",
       "   chrM: 75 variants\n",
       "==============================\n",
       "Allele type distribution\n",
       "------------------------\n",
       "        SNP: 41030 alternate alleles\n",
       "   Deletion: 3351 alternate alleles\n",
       "  Insertion: 2992 alternate alleles\n",
       "=============================="
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "hl.summarize_variants(mt)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='3'></a> \n",
    "## 3. Do simple aggregations, filtering and plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 447,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{False: 6678428, True: 474895}"
      ]
     },
     "execution_count": 447,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mt.aggregate_entries(hl.agg.counter(mt.info.MULTIALLELIC)) #count occurence of given value, but has troubles with the array fields"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "exercise: Try to see some variants where ISEQ_GENES_NAMES is.defined:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 448,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table><thead style=\"font-weight: bold;\"><tr><td>locus</td><td>alleles</td><td><expr></td></tr>\n",
       "<tr><td>locus&lt;GRCh38&gt;</td><td>array&lt;str&gt;</td><td>array&lt;str&gt;</td></tr>\n",
       "</thead><tbody><tr><td>chr1:69968</td><td>[&quot;A&quot;,&quot;G&quot;]</td><td>[&quot;OR4F5&quot;]</td></tr>\n",
       "<tr><td>chr1:183189</td><td>[&quot;G&quot;,&quot;C&quot;]</td><td>[&quot;FO538757.2&quot;]</td></tr>\n",
       "<tr><td>chr1:183238</td><td>[&quot;G&quot;,&quot;C&quot;]</td><td>[&quot;FO538757.2&quot;]</td></tr>\n",
       "<tr><td>chr1:183937</td><td>[&quot;G&quot;,&quot;A&quot;]</td><td>[&quot;FO538757.2&quot;]</td></tr>\n",
       "<tr><td>chr1:184994</td><td>[&quot;G&quot;,&quot;C&quot;]</td><td>[&quot;FO538757.1&quot;]</td></tr>\n",
       "</tbody></table><p style=\"background: #fdd; padding: 0.4em;\">showing top 5 rows</p>\n"
      ],
      "text/plain": [
       "+---------------+------------+----------------+\n",
       "| locus         | alleles    | <expr>         |\n",
       "+---------------+------------+----------------+\n",
       "| locus<GRCh38> | array<str> | array<str>     |\n",
       "+---------------+------------+----------------+\n",
       "| chr1:69968    | [\"A\",\"G\"]  | [\"OR4F5\"]      |\n",
       "| chr1:183189   | [\"G\",\"C\"]  | [\"FO538757.2\"] |\n",
       "| chr1:183238   | [\"G\",\"C\"]  | [\"FO538757.2\"] |\n",
       "| chr1:183937   | [\"G\",\"A\"]  | [\"FO538757.2\"] |\n",
       "| chr1:184994   | [\"G\",\"C\"]  | [\"FO538757.1\"] |\n",
       "+---------------+------------+----------------+\n",
       "showing top 5 rows"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mt.filter_rows(hl.is_defined(mt.info.ISEQ_GENES_NAMES)).info.ISEQ_GENES_NAMES.show(5) #is_defined() filters out empty/nondefined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 449,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: 6651990, 1: 344890, 2: 84194, None: 72249}"
      ]
     },
     "execution_count": 449,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mt.aggregate_entries(hl.agg.counter(mt.GT.n_alt_alleles())) #distribution of genotype calls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 450,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9898999388116544"
      ]
     },
     "execution_count": 450,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mt.aggregate_entries(hl.agg.fraction(hl.is_defined(mt.GT))) #calculate the call rate directly"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "annotate the mt with call rate per variant: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 451,
   "metadata": {},
   "outputs": [],
   "source": [
    "mt = mt.annotate_rows(call_rate = hl.agg.fraction(hl.is_defined(mt.GT)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 452,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--------------------------------------------------------\n",
      "Type:\n",
      "        float64\n",
      "--------------------------------------------------------\n",
      "Source:\n",
      "    <hail.matrixtable.MatrixTable object at 0x7f935fd37ad0>\n",
      "Index:\n",
      "    ['row']\n",
      "--------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "mt.call_rate.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 453,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<p>47373 records.</p><b>call_rate</b> (<i>float64</i>):<ul><table><tbody><tr><td>Non-missing</td><td>47373 (100.00%)</td></tr><tr><td>Missing</td><td>0</td></tr><tr><td>Minimum</td><td>0.01</td></tr><tr><td>Maximum</td><td>1.00</td></tr><tr><td>Mean</td><td>0.99</td></tr><tr><td>Std Dev</td><td>0.06</td></tr></tbody></table></ul>"
      ],
      "text/plain": [
       "\n",
       "47373 records.\n",
       "\n",
       "- call_rate (float64):\n",
       "  Non-missing: 47373 (100.00%)\n",
       "      Missing: 0\n",
       "      Minimum: 0.01\n",
       "      Maximum: 1.00\n",
       "         Mean: 0.99\n",
       "      Std Dev: 0.06"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mt.call_rate.summarize()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 460,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "  <div class=\"bk-root\" id=\"c302bff1-da87-43a0-8c6f-9c9bf80dbf1e\" data-root-id=\"7245\"></div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "(function(root) {\n",
       "  function embed_document(root) {\n",
       "    \n",
       "  var docs_json = {\"19848e0f-6572-45cc-8365-fd056b960702\":{\"roots\":{\"references\":[{\"attributes\":{\"background_fill_color\":{\"value\":\"#EEEEEE\"},\"below\":[{\"id\":\"7256\",\"type\":\"LinearAxis\"}],\"center\":[{\"id\":\"7260\",\"type\":\"Grid\"},{\"id\":\"7265\",\"type\":\"Grid\"},{\"id\":\"7290\",\"type\":\"Legend\"}],\"left\":[{\"id\":\"7261\",\"type\":\"LinearAxis\"}],\"renderers\":[{\"id\":\"7282\",\"type\":\"GlyphRenderer\"}],\"title\":{\"id\":\"7246\",\"type\":\"Title\"},\"toolbar\":{\"id\":\"7272\",\"type\":\"Toolbar\"},\"x_range\":{\"id\":\"7248\",\"type\":\"Range1d\"},\"x_scale\":{\"id\":\"7252\",\"type\":\"LinearScale\"},\"y_range\":{\"id\":\"7250\",\"type\":\"DataRange1d\"},\"y_scale\":{\"id\":\"7254\",\"type\":\"LinearScale\"}},\"id\":\"7245\",\"subtype\":\"Figure\",\"type\":\"Plot\"},{\"attributes\":{\"active_drag\":\"auto\",\"active_inspect\":\"auto\",\"active_multi\":null,\"active_scroll\":\"auto\",\"active_tap\":\"auto\",\"tools\":[{\"id\":\"7266\",\"type\":\"PanTool\"},{\"id\":\"7267\",\"type\":\"WheelZoomTool\"},{\"id\":\"7268\",\"type\":\"BoxZoomTool\"},{\"id\":\"7269\",\"type\":\"SaveTool\"},{\"id\":\"7270\",\"type\":\"ResetTool\"},{\"id\":\"7271\",\"type\":\"HelpTool\"}]},\"id\":\"7272\",\"type\":\"Toolbar\"},{\"attributes\":{},\"id\":\"7269\",\"type\":\"SaveTool\"},{\"attributes\":{\"label\":{\"value\":\"Call Rate\"},\"renderers\":[{\"id\":\"7282\",\"type\":\"GlyphRenderer\"}]},\"id\":\"7291\",\"type\":\"LegendItem\"},{\"attributes\":{},\"id\":\"7262\",\"type\":\"BasicTicker\"},{\"attributes\":{\"text\":\"Variant Call Rate Histogram\"},\"id\":\"7246\",\"type\":\"Title\"},{\"attributes\":{\"bottom\":{\"value\":0},\"fill_alpha\":{\"value\":0.1},\"fill_color\":{\"value\":\"#1f77b4\"},\"left\":{\"field\":\"left\"},\"line_alpha\":{\"value\":0.1},\"line_color\":{\"value\":\"#1f77b4\"},\"right\":{\"field\":\"right\"},\"top\":{\"field\":\"top\"}},\"id\":\"7281\",\"type\":\"Quad\"},{\"attributes\":{},\"id\":\"7271\",\"type\":\"HelpTool\"},{\"attributes\":{},\"id\":\"7252\",\"type\":\"LinearScale\"},{\"attributes\":{\"callback\":null,\"data\":{\"left\":[0.0,0.04,0.08,0.12,0.16,0.2,0.24,0.28,0.32,0.36,0.4,0.44,0.48,0.52,0.56,0.6,0.64,0.68,0.72,0.76,0.8,0.84,0.88,0.92,0.96],\"right\":[0.04,0.08,0.12,0.16,0.2,0.24,0.28,0.32,0.36,0.4,0.44,0.48,0.52,0.56,0.6,0.64,0.68,0.72,0.76,0.8,0.84,0.88,0.92,0.96,1.0],\"top\":[12,9,14,16,4,7,18,10,10,20,20,32,47,31,40,56,57,77,83,111,103,137,234,442,45783]},\"selected\":{\"id\":\"7530\",\"type\":\"Selection\"},\"selection_policy\":{\"id\":\"7529\",\"type\":\"UnionRenderers\"}},\"id\":\"7279\",\"type\":\"ColumnDataSource\"},{\"attributes\":{},\"id\":\"7270\",\"type\":\"ResetTool\"},{\"attributes\":{\"callback\":null,\"end\":1.05,\"start\":-0.05},\"id\":\"7248\",\"type\":\"Range1d\"},{\"attributes\":{},\"id\":\"7285\",\"type\":\"BasicTickFormatter\"},{\"attributes\":{\"bottom_units\":\"screen\",\"fill_alpha\":{\"value\":0.5},\"fill_color\":{\"value\":\"lightgrey\"},\"left_units\":\"screen\",\"level\":\"overlay\",\"line_alpha\":{\"value\":1.0},\"line_color\":{\"value\":\"black\"},\"line_dash\":[4,4],\"line_width\":{\"value\":2},\"render_mode\":\"css\",\"right_units\":\"screen\",\"top_units\":\"screen\"},\"id\":\"7289\",\"type\":\"BoxAnnotation\"},{\"attributes\":{\"axis_label\":\"Call Rate\",\"formatter\":{\"id\":\"7287\",\"type\":\"BasicTickFormatter\"},\"ticker\":{\"id\":\"7257\",\"type\":\"BasicTicker\"}},\"id\":\"7256\",\"type\":\"LinearAxis\"},{\"attributes\":{\"callback\":null},\"id\":\"7250\",\"type\":\"DataRange1d\"},{\"attributes\":{\"ticker\":{\"id\":\"7257\",\"type\":\"BasicTicker\"}},\"id\":\"7260\",\"type\":\"Grid\"},{\"attributes\":{},\"id\":\"7529\",\"type\":\"UnionRenderers\"},{\"attributes\":{\"axis_label\":\"Frequency\",\"formatter\":{\"id\":\"7285\",\"type\":\"BasicTickFormatter\"},\"ticker\":{\"id\":\"7262\",\"type\":\"BasicTicker\"}},\"id\":\"7261\",\"type\":\"LinearAxis\"},{\"attributes\":{\"bottom\":{\"value\":0},\"fill_color\":{\"value\":\"#1f77b4\"},\"left\":{\"field\":\"left\"},\"right\":{\"field\":\"right\"},\"top\":{\"field\":\"top\"}},\"id\":\"7280\",\"type\":\"Quad\"},{\"attributes\":{\"data_source\":{\"id\":\"7279\",\"type\":\"ColumnDataSource\"},\"glyph\":{\"id\":\"7280\",\"type\":\"Quad\"},\"hover_glyph\":null,\"muted_glyph\":null,\"nonselection_glyph\":{\"id\":\"7281\",\"type\":\"Quad\"},\"selection_glyph\":null,\"view\":{\"id\":\"7283\",\"type\":\"CDSView\"}},\"id\":\"7282\",\"type\":\"GlyphRenderer\"},{\"attributes\":{},\"id\":\"7257\",\"type\":\"BasicTicker\"},{\"attributes\":{\"items\":[{\"id\":\"7291\",\"type\":\"LegendItem\"}]},\"id\":\"7290\",\"type\":\"Legend\"},{\"attributes\":{\"overlay\":{\"id\":\"7289\",\"type\":\"BoxAnnotation\"}},\"id\":\"7268\",\"type\":\"BoxZoomTool\"},{\"attributes\":{},\"id\":\"7267\",\"type\":\"WheelZoomTool\"},{\"attributes\":{},\"id\":\"7266\",\"type\":\"PanTool\"},{\"attributes\":{},\"id\":\"7254\",\"type\":\"LinearScale\"},{\"attributes\":{},\"id\":\"7530\",\"type\":\"Selection\"},{\"attributes\":{\"dimension\":1,\"ticker\":{\"id\":\"7262\",\"type\":\"BasicTicker\"}},\"id\":\"7265\",\"type\":\"Grid\"},{\"attributes\":{\"source\":{\"id\":\"7279\",\"type\":\"ColumnDataSource\"}},\"id\":\"7283\",\"type\":\"CDSView\"},{\"attributes\":{},\"id\":\"7287\",\"type\":\"BasicTickFormatter\"}],\"root_ids\":[\"7245\"]},\"title\":\"Bokeh Application\",\"version\":\"1.2.0\"}};\n",
       "  var render_items = [{\"docid\":\"19848e0f-6572-45cc-8365-fd056b960702\",\"roots\":{\"7245\":\"c302bff1-da87-43a0-8c6f-9c9bf80dbf1e\"}}];\n",
       "  root.Bokeh.embed.embed_items_notebook(docs_json, render_items);\n",
       "\n",
       "  }\n",
       "  if (root.Bokeh !== undefined) {\n",
       "    embed_document(root);\n",
       "  } else {\n",
       "    var attempts = 0;\n",
       "    var timer = setInterval(function(root) {\n",
       "      if (root.Bokeh !== undefined) {\n",
       "        embed_document(root);\n",
       "        clearInterval(timer);\n",
       "      }\n",
       "      attempts++;\n",
       "      if (attempts > 100) {\n",
       "        console.log(\"Bokeh: ERROR: Unable to run BokehJS code because BokehJS library is missing\");\n",
       "        clearInterval(timer);\n",
       "      }\n",
       "    }, 10, root)\n",
       "  }\n",
       "})(window);"
      ],
      "application/vnd.bokehjs_exec.v0+json": ""
     },
     "metadata": {
      "application/vnd.bokehjs_exec.v0+json": {
       "id": "7245"
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "p = hl.plot.histogram(mt.call_rate, bins=25, title='Variant Call Rate Histogram', range=(0,1.0), legend='Call Rate')\n",
    "show(p) #plot call rate per variant frequency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 490,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "  <div class=\"bk-root\" id=\"329c9e3a-37a2-4d85-8e44-b44012f5f9bf\" data-root-id=\"7951\"></div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "(function(root) {\n",
       "  function embed_document(root) {\n",
       "    \n",
       "  var docs_json = {\"c9131da5-7787-4499-9f12-e5ac814f5c7b\":{\"roots\":{\"references\":[{\"attributes\":{\"background_fill_color\":{\"value\":\"#EEEEEE\"},\"below\":[{\"id\":\"7962\",\"type\":\"LinearAxis\"}],\"center\":[{\"id\":\"7966\",\"type\":\"Grid\"},{\"id\":\"7971\",\"type\":\"Grid\"},{\"id\":\"7996\",\"type\":\"Legend\"}],\"left\":[{\"id\":\"7967\",\"type\":\"LinearAxis\"}],\"renderers\":[{\"id\":\"7988\",\"type\":\"GlyphRenderer\"},{\"id\":\"8001\",\"type\":\"GlyphRenderer\"}],\"title\":{\"id\":\"7952\",\"type\":\"Title\"},\"toolbar\":{\"id\":\"7978\",\"type\":\"Toolbar\"},\"x_range\":{\"id\":\"7954\",\"type\":\"Range1d\"},\"x_scale\":{\"id\":\"7958\",\"type\":\"LinearScale\"},\"y_range\":{\"id\":\"7956\",\"type\":\"DataRange1d\"},\"y_scale\":{\"id\":\"7960\",\"type\":\"LinearScale\"}},\"id\":\"7951\",\"subtype\":\"Figure\",\"type\":\"Plot\"},{\"attributes\":{\"ticker\":{\"id\":\"7963\",\"type\":\"BasicTicker\"}},\"id\":\"7966\",\"type\":\"Grid\"},{\"attributes\":{\"callback\":null,\"data\":{},\"selected\":{\"id\":\"8270\",\"type\":\"Selection\"},\"selection_policy\":{\"id\":\"8269\",\"type\":\"UnionRenderers\"}},\"id\":\"7998\",\"type\":\"ColumnDataSource\"},{\"attributes\":{\"text\":\"DP Histogram\"},\"id\":\"7952\",\"type\":\"Title\"},{\"attributes\":{\"dimension\":1,\"ticker\":{\"id\":\"7968\",\"type\":\"BasicTicker\"}},\"id\":\"7971\",\"type\":\"Grid\"},{\"attributes\":{\"data_source\":{\"id\":\"7985\",\"type\":\"ColumnDataSource\"},\"glyph\":{\"id\":\"7986\",\"type\":\"Quad\"},\"hover_glyph\":null,\"muted_glyph\":null,\"nonselection_glyph\":{\"id\":\"7987\",\"type\":\"Quad\"},\"selection_glyph\":null,\"view\":{\"id\":\"7989\",\"type\":\"CDSView\"}},\"id\":\"7988\",\"type\":\"GlyphRenderer\"},{\"attributes\":{},\"id\":\"7976\",\"type\":\"ResetTool\"},{\"attributes\":{\"data_source\":{\"id\":\"7998\",\"type\":\"ColumnDataSource\"},\"glyph\":{\"id\":\"7999\",\"type\":\"Quad\"},\"hover_glyph\":null,\"muted_glyph\":null,\"nonselection_glyph\":{\"id\":\"8000\",\"type\":\"Quad\"},\"selection_glyph\":null,\"view\":{\"id\":\"8002\",\"type\":\"CDSView\"}},\"id\":\"8001\",\"type\":\"GlyphRenderer\"},{\"attributes\":{},\"id\":\"7975\",\"type\":\"SaveTool\"},{\"attributes\":{\"callback\":null},\"id\":\"7956\",\"type\":\"DataRange1d\"},{\"attributes\":{\"overlay\":{\"id\":\"7995\",\"type\":\"BoxAnnotation\"}},\"id\":\"7974\",\"type\":\"BoxZoomTool\"},{\"attributes\":{\"bottom\":{\"value\":0},\"fill_alpha\":{\"value\":0.1},\"fill_color\":{\"value\":\"#1f77b4\"},\"left\":{\"field\":\"left\"},\"line_alpha\":{\"value\":0.1},\"line_color\":{\"value\":\"#1f77b4\"},\"right\":{\"field\":\"right\"},\"top\":{\"field\":\"top\"}},\"id\":\"7987\",\"type\":\"Quad\"},{\"attributes\":{\"bottom_units\":\"screen\",\"fill_alpha\":{\"value\":0.5},\"fill_color\":{\"value\":\"lightgrey\"},\"left_units\":\"screen\",\"level\":\"overlay\",\"line_alpha\":{\"value\":1.0},\"line_color\":{\"value\":\"black\"},\"line_dash\":[4,4],\"line_width\":{\"value\":2},\"render_mode\":\"css\",\"right_units\":\"screen\",\"top_units\":\"screen\"},\"id\":\"7995\",\"type\":\"BoxAnnotation\"},{\"attributes\":{},\"id\":\"7968\",\"type\":\"BasicTicker\"},{\"attributes\":{},\"id\":\"7977\",\"type\":\"HelpTool\"},{\"attributes\":{\"axis_label\":\"Frequency\",\"formatter\":{\"id\":\"7991\",\"type\":\"BasicTickFormatter\"},\"ticker\":{\"id\":\"7968\",\"type\":\"BasicTicker\"}},\"id\":\"7967\",\"type\":\"LinearAxis\"},{\"attributes\":{\"active_drag\":\"auto\",\"active_inspect\":\"auto\",\"active_multi\":null,\"active_scroll\":\"auto\",\"active_tap\":\"auto\",\"tools\":[{\"id\":\"7972\",\"type\":\"PanTool\"},{\"id\":\"7973\",\"type\":\"WheelZoomTool\"},{\"id\":\"7974\",\"type\":\"BoxZoomTool\"},{\"id\":\"7975\",\"type\":\"SaveTool\"},{\"id\":\"7976\",\"type\":\"ResetTool\"},{\"id\":\"7977\",\"type\":\"HelpTool\"}]},\"id\":\"7978\",\"type\":\"Toolbar\"},{\"attributes\":{\"source\":{\"id\":\"7985\",\"type\":\"ColumnDataSource\"}},\"id\":\"7989\",\"type\":\"CDSView\"},{\"attributes\":{\"source\":{\"id\":\"7998\",\"type\":\"ColumnDataSource\"}},\"id\":\"8002\",\"type\":\"CDSView\"},{\"attributes\":{},\"id\":\"7958\",\"type\":\"LinearScale\"},{\"attributes\":{},\"id\":\"7973\",\"type\":\"WheelZoomTool\"},{\"attributes\":{},\"id\":\"7993\",\"type\":\"BasicTickFormatter\"},{\"attributes\":{\"bottom\":{\"value\":0},\"fill_color\":{\"value\":\"green\"},\"left\":{\"value\":100.0},\"right\":{\"value\":103.33333333333333},\"top\":{\"value\":49584}},\"id\":\"7999\",\"type\":\"Quad\"},{\"attributes\":{\"callback\":null,\"end\":105.0,\"start\":-5.0},\"id\":\"7954\",\"type\":\"Range1d\"},{\"attributes\":{\"items\":[{\"id\":\"7997\",\"type\":\"LegendItem\"},{\"id\":\"8011\",\"type\":\"LegendItem\"}]},\"id\":\"7996\",\"type\":\"Legend\"},{\"attributes\":{},\"id\":\"7991\",\"type\":\"BasicTickFormatter\"},{\"attributes\":{},\"id\":\"7963\",\"type\":\"BasicTicker\"},{\"attributes\":{},\"id\":\"8269\",\"type\":\"UnionRenderers\"},{\"attributes\":{\"callback\":null,\"data\":{\"left\":[0.0,3.3333333333333335,6.666666666666667,10.0,13.333333333333334,16.666666666666668,20.0,23.333333333333336,26.666666666666668,30.0,33.333333333333336,36.66666666666667,40.0,43.333333333333336,46.66666666666667,50.0,53.333333333333336,56.66666666666667,60.0,63.333333333333336,66.66666666666667,70.0,73.33333333333334,76.66666666666667,80.0,83.33333333333334,86.66666666666667,90.0,93.33333333333334,96.66666666666667],\"right\":[3.3333333333333335,6.666666666666667,10.0,13.333333333333334,16.666666666666668,20.0,23.333333333333336,26.666666666666668,30.0,33.333333333333336,36.66666666666667,40.0,43.333333333333336,46.66666666666667,50.0,53.333333333333336,56.66666666666667,60.0,63.333333333333336,66.66666666666667,70.0,73.33333333333334,76.66666666666667,80.0,83.33333333333334,86.66666666666667,90.0,93.33333333333334,96.66666666666667,100.0],\"top\":[144329,167056,206059,351731,367465,511304,1077107,820500,807437,921153,860473,435425,194483,59116,35680,33017,18120,15109,15893,9877,8356,9753,6021,5514,6049,3908,3584,3862,2506,2852]},\"selected\":{\"id\":\"8009\",\"type\":\"Selection\"},\"selection_policy\":{\"id\":\"8008\",\"type\":\"UnionRenderers\"}},\"id\":\"7985\",\"type\":\"ColumnDataSource\"},{\"attributes\":{},\"id\":\"8008\",\"type\":\"UnionRenderers\"},{\"attributes\":{\"bottom\":{\"value\":0},\"fill_alpha\":{\"value\":0.1},\"fill_color\":{\"value\":\"#1f77b4\"},\"left\":{\"value\":100.0},\"line_alpha\":{\"value\":0.1},\"line_color\":{\"value\":\"#1f77b4\"},\"right\":{\"value\":103.33333333333333},\"top\":{\"value\":49584}},\"id\":\"8000\",\"type\":\"Quad\"},{\"attributes\":{},\"id\":\"7972\",\"type\":\"PanTool\"},{\"attributes\":{},\"id\":\"8009\",\"type\":\"Selection\"},{\"attributes\":{},\"id\":\"7960\",\"type\":\"LinearScale\"},{\"attributes\":{\"axis_label\":\"DP\",\"formatter\":{\"id\":\"7993\",\"type\":\"BasicTickFormatter\"},\"ticker\":{\"id\":\"7963\",\"type\":\"BasicTicker\"}},\"id\":\"7962\",\"type\":\"LinearAxis\"},{\"attributes\":{},\"id\":\"8270\",\"type\":\"Selection\"},{\"attributes\":{\"label\":{\"value\":\"Outliers Above\"},\"renderers\":[{\"id\":\"8001\",\"type\":\"GlyphRenderer\"}]},\"id\":\"8011\",\"type\":\"LegendItem\"},{\"attributes\":{\"bottom\":{\"value\":0},\"fill_color\":{\"value\":\"#1f77b4\"},\"left\":{\"field\":\"left\"},\"right\":{\"field\":\"right\"},\"top\":{\"field\":\"top\"}},\"id\":\"7986\",\"type\":\"Quad\"},{\"attributes\":{\"label\":{\"value\":\"DP\"},\"renderers\":[{\"id\":\"7988\",\"type\":\"GlyphRenderer\"}]},\"id\":\"7997\",\"type\":\"LegendItem\"}],\"root_ids\":[\"7951\"]},\"title\":\"Bokeh Application\",\"version\":\"1.2.0\"}};\n",
       "  var render_items = [{\"docid\":\"c9131da5-7787-4499-9f12-e5ac814f5c7b\",\"roots\":{\"7951\":\"329c9e3a-37a2-4d85-8e44-b44012f5f9bf\"}}];\n",
       "  root.Bokeh.embed.embed_items_notebook(docs_json, render_items);\n",
       "\n",
       "  }\n",
       "  if (root.Bokeh !== undefined) {\n",
       "    embed_document(root);\n",
       "  } else {\n",
       "    var attempts = 0;\n",
       "    var timer = setInterval(function(root) {\n",
       "      if (root.Bokeh !== undefined) {\n",
       "        embed_document(root);\n",
       "        clearInterval(timer);\n",
       "      }\n",
       "      attempts++;\n",
       "      if (attempts > 100) {\n",
       "        console.log(\"Bokeh: ERROR: Unable to run BokehJS code because BokehJS library is missing\");\n",
       "        clearInterval(timer);\n",
       "      }\n",
       "    }, 10, root)\n",
       "  }\n",
       "})(window);"
      ],
      "application/vnd.bokehjs_exec.v0+json": ""
     },
     "metadata": {
      "application/vnd.bokehjs_exec.v0+json": {
       "id": "7951"
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "p = hl.plot.histogram(mt.DP, range=(0,100), bins=30, title='DP Histogram', legend='DP')\n",
    "show(p)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='4'></a> \n",
    "## 4. Work with single sample and single variant"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='4.1'></a> \n",
    "### 4.1 Single sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 461,
   "metadata": {},
   "outputs": [],
   "source": [
    "S_136 = mt.filter_cols(mt.s == \"S_136\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 462,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--------------------------------------------------------\n",
      "Type:\n",
      "        call\n",
      "--------------------------------------------------------\n",
      "Source:\n",
      "    <hail.matrixtable.MatrixTable object at 0x7f937795a110>\n",
      "Index:\n",
      "    ['column', 'row']\n",
      "--------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "S_136.GT.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 463,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 463,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "S_136.count_cols()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 464,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "47373"
      ]
     },
     "execution_count": 464,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "S_136.count_rows()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 465,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: 44094, 1: 2291, 2: 532, None: 456}"
      ]
     },
     "execution_count": 465,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "S_136.aggregate_entries(hl.agg.counter(S_136.GT.n_alt_alleles()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 413,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table><thead style=\"font-weight: bold;\"><tr><td>locus</td><td>alleles</td><td>S_136.GT</td></tr>\n",
       "<tr><td>locus&lt;GRCh38&gt;</td><td>array&lt;str&gt;</td><td>call</td></tr>\n",
       "</thead><tbody><tr><td>chr1:10146</td><td>[&quot;AC&quot;,&quot;*&quot;]</td><td>0/0</td></tr>\n",
       "<tr><td>chr1:10177</td><td>[&quot;A&quot;,&quot;*&quot;]</td><td>NA</td></tr>\n",
       "<tr><td>chr1:10622</td><td>[&quot;T&quot;,&quot;*&quot;]</td><td>NA</td></tr>\n",
       "<tr><td>chr1:10623</td><td>[&quot;T&quot;,&quot;*&quot;]</td><td>NA</td></tr>\n",
       "<tr><td>chr1:30923</td><td>[&quot;G&quot;,&quot;*&quot;]</td><td>NA</td></tr>\n",
       "<tr><td>chr1:54714</td><td>[&quot;TTCTTTCTTTCTTTC&quot;,&quot;*&quot;]</td><td>0/0</td></tr>\n",
       "<tr><td>chr1:54715</td><td>[&quot;TC&quot;,&quot;*&quot;]</td><td>0/0</td></tr>\n",
       "<tr><td>chr1:54718</td><td>[&quot;TTCTTTC&quot;,&quot;*&quot;]</td><td>0/0</td></tr>\n",
       "<tr><td>chr1:54720</td><td>[&quot;CTTTCTTTCTTTCT&quot;,&quot;*&quot;]</td><td>0/0</td></tr>\n",
       "<tr><td>chr1:54724</td><td>[&quot;CT&quot;,&quot;*&quot;]</td><td>0/1</td></tr>\n",
       "</tbody></table><p style=\"background: #fdd; padding: 0.4em;\">showing top 10 rows</p>\n"
      ],
      "text/plain": [
       "+---------------+-------------------------+----------+\n",
       "| locus         | alleles                 | S_136.GT |\n",
       "+---------------+-------------------------+----------+\n",
       "| locus<GRCh38> | array<str>              | call     |\n",
       "+---------------+-------------------------+----------+\n",
       "| chr1:10146    | [\"AC\",\"*\"]              | 0/0      |\n",
       "| chr1:10177    | [\"A\",\"*\"]               | NA       |\n",
       "| chr1:10622    | [\"T\",\"*\"]               | NA       |\n",
       "| chr1:10623    | [\"T\",\"*\"]               | NA       |\n",
       "| chr1:30923    | [\"G\",\"*\"]               | NA       |\n",
       "| chr1:54714    | [\"TTCTTTCTTTCTTTC\",\"*\"] | 0/0      |\n",
       "| chr1:54715    | [\"TC\",\"*\"]              | 0/0      |\n",
       "| chr1:54718    | [\"TTCTTTC\",\"*\"]         | 0/0      |\n",
       "| chr1:54720    | [\"CTTTCTTTCTTTCT\",\"*\"]  | 0/0      |\n",
       "| chr1:54724    | [\"CT\",\"*\"]              | 0/1      |\n",
       "+---------------+-------------------------+----------+\n",
       "showing top 10 rows"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "S_136.GT.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{False: 44228, True: 682152}"
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "S_136.aggregate_entries(hl.agg.counter(S_136.info.MULTIALLELIC))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'ArrayExpression' object has no attribute 'is_star'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-212-8b879894dc48>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mS_136\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0malleles\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_star\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m: 'ArrayExpression' object has no attribute 'is_star'"
     ]
    }
   ],
   "source": [
    "S_136.alleles.is_star()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<bound method Expression.describe of <ArrayExpression of type array<str>>>"
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "S_136.alleles.describe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='4.2'></a> \n",
    "### 4.2 Single variant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--------------------------------------------------------\n",
      "Type:\n",
      "        locus<GRCh38>\n",
      "--------------------------------------------------------\n",
      "Source:\n",
      "    <hail.matrixtable.MatrixTable object at 0x7f93c162d410>\n",
      "Index:\n",
      "    ['row']\n",
      "--------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "mt.locus.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Locus(contig=chr1, position=183189, reference_genome=GRCh38)"
      ]
     },
     "execution_count": 186,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hl.Locus.parse('chr1:183189', reference_genome='GRCh38') #get hail locus-type field from chrZ:12345 string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [],
   "source": [
    "chr1_183189 = mt.filter_rows(mt.locus == hl.Locus.parse('chr1:183189', reference_genome='GRCh38'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-11-14 16:46:20 Hail: INFO: reading 1 of 116 data partitions\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{0: 132, 1: 19}"
      ]
     },
     "execution_count": 188,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr1_183189.aggregate_entries(hl.agg.counter(chr1_183189.GT.n_alt_alleles()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 375,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-11-19 08:29:49 Hail: INFO: reading 1 of 116 data partitions\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table><thead style=\"font-weight: bold;\"><tr><td>locus</td><td>alleles</td></tr>\n",
       "<tr><td>locus&lt;GRCh38&gt;</td><td>array&lt;str&gt;</td></tr>\n",
       "</thead><tbody><tr><td>chr1:183189</td><td>[&quot;G&quot;,&quot;C&quot;]</td></tr>\n",
       "</tbody></table>"
      ],
      "text/plain": [
       "+---------------+------------+\n",
       "| locus         | alleles    |\n",
       "+---------------+------------+\n",
       "| locus<GRCh38> | array<str> |\n",
       "+---------------+------------+\n",
       "| chr1:183189   | [\"G\",\"C\"]  |\n",
       "+---------------+------------+"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "chr1_183189.alleles.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 374,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-11-19 08:28:42 Hail: INFO: reading 1 of 116 data partitions\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[['C|missense_variant|MODERATE|FO538757.2|ENSG00000279928|transcript|ENST00000624431.1|protein_coding|2/3|c.114G>C|p.Lys38Asn|430/718|114/402|38/133||',\n",
       "  'C|downstream_gene_variant|MODIFIER|FO538757.1|ENSG00000279457|transcript|ENST00000623083.3|protein_coding||c.*2028C>G|||||1736|',\n",
       "  'C|downstream_gene_variant|MODIFIER|FO538757.1|ENSG00000279457|transcript|ENST00000623834.3|protein_coding||c.*2028C>G|||||1734|',\n",
       "  'C|downstream_gene_variant|MODIFIER|FO538757.1|ENSG00000279457|transcript|ENST00000624735.1|protein_coding||c.*1738C>G|||||1738|',\n",
       "  'C|downstream_gene_variant|MODIFIER|MIR6859-2|ENSG00000273874|transcript|ENST00000612080.1|miRNA||n.*4702C>G|||||4702|']]"
      ]
     },
     "execution_count": 374,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr1_183189.info.ANN.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 385,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-11-19 08:49:37 Hail: INFO: reading 1 of 116 data partitions\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[None]"
      ]
     },
     "execution_count": 385,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr1_183189.info.ISEQ_AGGREGATED_CLINVAR_SIGNIFICANCE.collect()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='5'></a> \n",
    "\n",
    "### 5. Explore Clinvar sinificance of detected variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 395,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<ArrayExpression of type array<str>>"
      ]
     },
     "execution_count": 395,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mt.info.ISEQ_AGGREGATED_CLINVAR_SIGNIFICANCE # hl.delimit function converts array into string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 409,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'likely_benign': 487,\n",
       " None: 44158,\n",
       " 'risk_factor': 8,\n",
       " 'benign': 363,\n",
       " 'benign:other': 1,\n",
       " 'pathogenic/likely_pathogenic': 32,\n",
       " 'uncertain_significance:other': 1,\n",
       " 'likely_pathogenic': 28,\n",
       " 'not_provided': 60,\n",
       " 'protective': 2,\n",
       " 'pathogenic:risk_factor': 2,\n",
       " 'benign/likely_benign': 330,\n",
       " 'drug_response': 5,\n",
       " 'conflicting_interpretations_of_pathogenicity': 697,\n",
       " 'benign/likely_benign:other': 3,\n",
       " 'uncertain_significance': 1074,\n",
       " 'other': 5,\n",
       " 'pathogenic': 114,\n",
       " 'conflicting_interpretations_of_pathogenicity:risk_factor': 1,\n",
       " 'affects': 2}"
      ]
     },
     "execution_count": 409,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mt.aggregate_rows(hl.agg.counter(hl.delimit(mt.info.ISEQ_AGGREGATED_CLINVAR_SIGNIFICANCE))) #yay!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 414,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BooleanExpression of type bool>"
      ]
     },
     "execution_count": 414,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hl.is_defined(mt.info.ISEQ_CLINVAR_DISEASES)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 434,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Combined_oxidative_phosphorylation_deficiency^Combined_oxidative_phosphorylation_deficiency_5^OVARIAN_DYSGENESIS_7': 1,\n",
       " 'Cerebrooculofacioskeletal_syndrome_4^ERCC1-Related_Xeroderma_Pigmentosum^Xeroderma_pigmentosum': 1,\n",
       " 'White_sponge_nevus_of_cannon': 3,\n",
       " 'Autoimmune_lymphoproliferative_syndrome^Autoimmune_lymphoproliferative_syndrome_type_2A^Neoplasm_of_stomach^Non-Hodgkin_lymphoma': 1,\n",
       " 'Cohen_syndrome': 8,\n",
       " 'Pachyonychia_congenita_2^Pachyonychia_congenita_syndrome^Steatocystoma_multiplex': 1,\n",
       " 'Otitis_media_susceptibility_to': 4,\n",
       " 'PLATELET_ABNORMALITIES_WITH_EOSINOPHILIA_AND_IMMUNE-MEDIATED_INFLAMMATORY_DISEASE': 7,\n",
       " 'Cortical_dysplasia_complex_with_other_brain_malformations_6^Michelin-tire_baby': 2,\n",
       " 'AMYOTROPHIC_LATERAL_SCLEROSIS_SUSCEPTIBILITY_TO_25^Myoclonus_intractable_neonatal^Spastic_paraplegia_10': 1,\n",
       " 'Spondyloepimetaphyseal_dysplasia_Genevieve_type': 1,\n",
       " 'Reduced_antithrombin_III_activity^Venous_thrombosis': 13,\n",
       " 'Cousin_syndrome': 1,\n",
       " 'Deafness_autosomal_recessive_101': 1,\n",
       " 'Mitochondrial_phosphate_carrier_deficiency': 7,\n",
       " '.:Age-related_macular_degeneration^Age-related_macular_degeneration_5^Cerebro-oculo-facio-skeletal_syndrome^Cockayne_syndrome^Cockayne_syndrome_B^DE_SANCTIS-CACCHIONE_SYNDROME^Lung_cancer^Neoplasm_of_lung^Premature_ovarian_failure_11^UV-sensitive_syndrome': 1,\n",
       " 'Cleft_palate_cardiac_defects_and_mental_retardation': 1,\n",
       " 'Efavirenz_response': 5,\n",
       " 'Early_infantile_epileptic_encephalopathy_55': 1,\n",
       " 'PEELING_SKIN_SYNDROME_6': 7,\n",
       " 'Thyroid_hormone_metabolism_abnormal': 5,\n",
       " 'Acatalasemia': 10,\n",
       " 'Dubin-Johnson_syndrome': 11,\n",
       " 'IMMUNODEFICIENCY_57': 1,\n",
       " 'Aortic_valve_disease_2^Craniosynostosis_7': 3,\n",
       " 'Disorders_of_Intracellular_Cobalamin_Metabolism^METHYLMALONIC_ACIDURIA_AND_HOMOCYSTINURIA_cblJ_TYPE': 4,\n",
       " 'Bile_acid_synthesis_defect_congenital_5': 3,\n",
       " 'Retinal_dystrophy_with_or_without_extraocular_anomalies': 3,\n",
       " 'Congenital_disorder_of_glycosylation^Congenital_disorder_of_glycosylation_type_2B': 3,\n",
       " 'Agammaglobulinemia_non-Bruton_type': 5,\n",
       " 'Platelet-type_bleeding_disorder_15': 2,\n",
       " 'Ciliary_dyskinesia_primary_10^Primary_ciliary_dyskinesia': 2,\n",
       " 'SPERMATOGENIC_FAILURE_38': 2,\n",
       " 'Alport_syndrome^Alport_syndrome_autosomal_recessive^Benign_familial_hematuria^Collagen_IV-related_nephropathies^Hereditary_hearing_loss_and_deafness': 2,\n",
       " 'Autism_17^Autism_spectrum_disorder': 6,\n",
       " 'Nephronophthisis^Nephronophthisis_9^Renal-hepatic-pancreatic_dysplasia_2': 2,\n",
       " 'Glycogen_storage_disease^Glycogen_storage_disease_type_VII': 2,\n",
       " 'Cardiomyopathy^Familial_hypertrophic_cardiomyopathy_16^Primary_familial_hypertrophic_cardiomyopathy': 2,\n",
       " 'Anophthalmia-microphthalmia_syndrome^Cataract_23_multiple_types': 1,\n",
       " 'Microphthalmia_syndromic_9': 5,\n",
       " 'FRAXE': 4,\n",
       " 'Orofaciodigital_syndrome_XV': 12,\n",
       " \"Deafness_autosomal_dominant_39_with_dentinogenesis_imperfecta_1^Denticles^Dentinogenesis_imperfecta_-_Shield's_type_II^Dentinogenesis_imperfecta_-_Shield's_type_III\": 8,\n",
       " 'SPINOCEREBELLAR_ATAXIA_48^Spinocerebellar_ataxia_autosomal_recessive_16': 1,\n",
       " 'NEURODEVELOPMENTAL_DISORDER_WITH_CENTRAL_AND_PERIPHERAL_MOTOR_DYSFUNCTION': 7,\n",
       " 'Joubert_syndrome^Joubert_syndrome_15': 2,\n",
       " 'Nemaline_myopathy_10': 2,\n",
       " 'Multiple_joint_dislocations_short_stature_craniofacial_dysmorphism_and_congenital_heart_defects^Spondyloepiphyseal_dysplasia_with_congenital_joint_dislocations': 2,\n",
       " 'Morquio_syndrome^Mucopolysaccharidosis_MPS-IV-A': 3,\n",
       " 'Mental_retardation_autosomal_recessive_56': 16,\n",
       " 'Upshaw-Schulman_syndrome': 3,\n",
       " 'Systemic_lupus_erythematosus^Systemic_lupus_erythematosus_11': 5,\n",
       " 'Ectodermal_dysplasia_13_hair/tooth_type': 2,\n",
       " 'Short-rib_thoracic_dysplasia_16_with_or_without_polydactyly': 1,\n",
       " 'Urocanate_hydratase_deficiency': 1,\n",
       " 'Parkinson_disease_18': 24,\n",
       " 'EPILEPTIC_ENCEPHALOPATHY_EARLY_INFANTILE_65': 1,\n",
       " 'Autoinflammation_antibody_deficiency_and_immune_dysregulation_plcg2-associated^Familial_cold_autoinflammatory_syndrome_3': 3,\n",
       " 'Muscular_dystrophy_limb-girdle_type_2y': 41,\n",
       " 'Deafness_autosomal_recessive_24^Hereditary_hearing_loss_and_deafness^Nonsyndromic_Hearing_Loss_and_Deafness_Autosomal_Recessive^Nonsyndromic_hearing_loss_and_deafness': 1,\n",
       " 'Occult_macular_dystrophy': 17,\n",
       " 'Rolandic_epilepsy_with_mental_retardation_and_speech_dyspraxia_X-linked': 1,\n",
       " 'Mental_retardation_autosomal_recessive_38^Prader-Willi_syndrome^Skin/hair/eye_pigmentation_variation_in_1': 6,\n",
       " 'Glutaryl-CoA_oxidase_deficiency': 1,\n",
       " 'Epileptic_encephalopathy_early_infantile_53^Parkinson_disease_20_early-onset': 3,\n",
       " 'Hyperphosphatasia_with_mental_retardation_syndrome_4^Hyperphosphatasia-intellectual_disability_syndrome': 6,\n",
       " 'Hypogonadotropic_hypogonadism_10_with_or_without_anosmia': 1,\n",
       " 'Diabetes_mellitus_noninsulin-dependent_5': 4,\n",
       " 'MACROTHROMBOCYTOPENIA_AUTOSOMAL_DOMINANT_TUBB1-RELATED': 3,\n",
       " 'Neural_tube_defect^Sacral_defect_with_anterior_meningocele': 1,\n",
       " 'Congenital_disorder_of_glycosylation^Congenital_disorder_of_glycosylation_type_2C': 2,\n",
       " 'Corneal_dystrophy_Fuchs_endothelial_8': 6,\n",
       " 'NEURODEVELOPMENTAL_DISORDER_WITH_SEIZURES_AND_NONEPILEPTIC_HYPERKINETIC_MOVEMENTS': 3,\n",
       " 'Amelogenesis_imperfecta_type_IF': 2,\n",
       " 'Mirage_syndrome^Tumoral_calcinosis_familial_normophosphatemic': 4,\n",
       " 'Pityriasis_rubra_pilaris^Psoriasis_susceptibility_2': 15,\n",
       " None: 35507,\n",
       " 'RH-NULL_AMORPH_TYPE': 1,\n",
       " 'Coronary_heart_disease_5': 3,\n",
       " 'Neurodegeneration_with_brain_iron_accumulation^Neurodegeneration_with_brain_iron_accumulation_4^Spastic_paraplegia_43_autosomal_recessive': 1,\n",
       " 'Ashkenazi_Jewish_disorders^Cystic_fibrosis^Susceptibility_to_malaria^Systemic_lupus_erythematosus': 2,\n",
       " 'Palmoplantar_keratoderma_mutilating_with_periorificial_keratotic_plaques^Palmoplantar_keratoderma_nonepidermolytic_focal_2': 8,\n",
       " 'Focal_segmental_glomerulosclerosis_9^Ventriculomegaly_with_cystic_kidney_disease': 8,\n",
       " 'Congenital_disorder_of_glycosylation^Congenital_disorder_of_glycosylation_type_1N': 2,\n",
       " 'Thiourea_tasting': 1,\n",
       " 'NEURODEVELOPMENTAL_DISORDER_WITH_OR_WITHOUT_SEIZURES_AND_GAIT_ABNORMALITIES': 1,\n",
       " 'Bruck_syndrome_2': 3,\n",
       " 'Amyotrophic_lateral_sclerosis_18': 23,\n",
       " 'Congenital_heart_defects_multiple_types_2': 1,\n",
       " 'Variegate_porphyria': 1,\n",
       " 'Conduction_disorder_of_the_heart': 3,\n",
       " '.:Cutaneous_malignant_melanoma_5^Cutaneous_melanoma^Hereditary_cutaneous_melanoma^Increased_analgesia_from_kappa-opioid_receptor_agonist_female-specific^Malignant_melanoma_of_skin^Oculocutaneous_albinism^Skin/hair/eye_pigmentation_variation_in_2^Tyrosinase-positive_oculocutaneous_albinism': 9,\n",
       " 'MICROCEPHALY_21_PRIMARY_AUTOSOMAL_RECESSIVE': 6,\n",
       " 'Diabetes_mellitus_type_2^Maturity-onset_diabetes_of_the_young_type_4^Neonatal_diabetes_mellitus^Pancreatic_agenesis_congenital^Permanent_neonatal_diabetes_mellitus': 1,\n",
       " 'Ciliary_dyskinesia_primary_19^Primary_ciliary_dyskinesia': 1,\n",
       " 'Hereditary_hearing_loss_and_deafness^Peroxisome_biogenesis_disorder_10A^Peroxisome_biogenesis_disorder_10b^Peroxisome_biogenesis_disorders^Peroxisome_biogenesis_disorders_Zellweger_syndrome_spectrum': 1,\n",
       " 'Myopathy_centronuclear_1^Myotubular_myopathy': 3,\n",
       " 'Congenital_muscular_dystrophy^Congenital_muscular_dystrophy-dystroglycanopathy_with_mental_retardation_type_B1^Limb-girdle_muscular_dystrophy-dystroglycanopathy_type_C1^Muscular_dystrophy-dystroglycanopathy_(congenital_with_brain_and_eye_anomalies)_type_A_1^alpha-Dystroglycan_related_dystrophy': 2,\n",
       " 'Brachydactyly_with_hypertension': 2,\n",
       " 'Posterior_column_ataxia_with_retinitis_pigmentosa': 1,\n",
       " 'Ichthyosis_leukocyte_vacuoles_alopecia_and_sclerosing_cholangitis': 2,\n",
       " 'Cerebellar_ataxia_mental_retardation_and_dysequilibrium_syndrome_1': 8,\n",
       " 'Retinitis_pigmentosa^Retinitis_pigmentosa_62': 2,\n",
       " 'Deafness_autosomal_recessive_6^Hereditary_hearing_loss_and_deafness^Nonsyndromic_Hearing_Loss_and_Deafness_Autosomal_Recessive^Nonsyndromic_hearing_loss_and_deafness': 3,\n",
       " 'Autism_spectrum_disorder^Autism_susceptibility_to_X-linked_4': 1,\n",
       " 'Ashkenazi_Jewish_disorders^Tay-Sachs_disease': 3,\n",
       " 'Knobloch_syndrome_1': 10,\n",
       " 'Congenital_central_hypoventilation^Hirschsprung_disease^Hirschsprung_disease_3^Pheochromocytoma': 1,\n",
       " 'Cataract_congenital_zonular_with_sutural_opacities': 1,\n",
       " 'Dihydropyrimidinase_deficiency': 11,\n",
       " 'Coffin-Siris_syndrome_5^Meningioma_familial': 1,\n",
       " 'Diarrhea_7': 1,\n",
       " 'Long_QT_syndrome_14^Ventricular_tachycardia_catecholaminergic_polymorphic_4': 10,\n",
       " 'Congenital_stationary_night_blindness^Congenital_stationary_night_blindness_type_1B': 4,\n",
       " 'Colorectal_cancer_susceptibility_to_12^Facial_dysmorphism_immunodeficiency_livedo_and_short_stature^INTRAUTERINE_GROWTH_RETARDATION_METAPHYSEAL_DYSPLASIA_ADRENAL_HYPOPLASIA_CONGENITA_GENITAL_ANOMALIES_AND_IMMUNODEFICIENCY': 9,\n",
       " 'VERTEBRAL_CARDIAC_RENAL_AND_LIMB_DEFECTS_SYNDROME_1': 1,\n",
       " 'Adenosine_triphosphate_elevated_of_erythrocytes^Pyruvate_kinase_deficiency_of_red_cells': 3,\n",
       " 'Alagille_syndrome_2^Arteriohepatic_dysplasia^Hajdu-Cheney_syndrome': 5,\n",
       " \"Ectodermal_dysplasia_'pure'_hair-nail_type\": 2,\n",
       " 'Dicarboxylic_aminoaciduria^Schizophrenia^Schizophrenia_18': 3,\n",
       " 'Corpus_callosum_agenesis_of_with_facial_anomalies_and_cerebellar_ataxia': 8,\n",
       " 'Autosomal_recessive_congenital_ichthyosis_10^Congenital_ichthyosis_of_skin': 1,\n",
       " 'Deafness_autosomal_recessive_49^Hereditary_hearing_loss_and_deafness^Nonsyndromic_Hearing_Loss_and_Deafness_Autosomal_Recessive^Nonsyndromic_hearing_loss_and_deafness': 1,\n",
       " 'Bombay_phenotype^Vitamin_b12_plasma_level_quantitative_trait_locus_1': 2,\n",
       " 'Autosomal_dominant_non-syndromic_sensorineural_deafness_type_DFNA^Deafness_autosomal_dominant_64^Hereditary_hearing_loss_and_deafness^Nonsyndromic_hearing_loss_and_deafness': 1,\n",
       " 'Microcephaly_with_or_without_chorioretinopathy_lymphedema_or_mental_retardation': 3,\n",
       " 'Hereditary_hearing_loss_and_deafness^Peroxisome_biogenesis_disorder_5B^Peroxisome_biogenesis_disorder_5a_(zellweger)^Peroxisome_biogenesis_disorders^Peroxisome_biogenesis_disorders_Zellweger_syndrome_spectrum': 1,\n",
       " 'Cardiomyopathy^Dilated_cardiomyopathy_1AA^Familial_dilated_cardiomyopathy^Primary_dilated_cardiomyopathy^Primary_familial_hypertrophic_cardiomyopathy': 2,\n",
       " 'Androgen_resistance_syndrome^Bulbo-spinal_atrophy_X-linked^Hypospadias_1_X-linked^Malignant_tumor_of_prostate^Neoplasm_of_the_genitourinary_tract^Partial_androgen_insensitivity_syndrome': 7,\n",
       " 'Cornelia_de_Lange_syndrome_4^De_Lange_syndrome^Mungan_syndrome': 2,\n",
       " 'KHAN-KHAN-KATSANIS_SYNDROME': 3,\n",
       " 'Acute_myeloid_leukemia_M6_type^Lethal_congenital_contracture_syndrome_2': 3,\n",
       " 'Epileptic_encephalopathy_early_infantile_32': 1,\n",
       " 'Ciliary_dyskinesia_primary_20^Primary_ciliary_dyskinesia': 1,\n",
       " 'Xeroderma_pigmentosum^Xeroderma_pigmentosum_type_1': 1,\n",
       " 'PULMONARY_ALVEOLAR_MICROLITHIASIS': 2,\n",
       " 'IMMUNODEFICIENCY_15A^Immunodeficiency_15': 6,\n",
       " 'Retinitis_pigmentosa^Retinitis_pigmentosa_69': 1,\n",
       " 'Thyrotoxic_periodic_paralysis_2': 28,\n",
       " 'Cataract_45': 4,\n",
       " 'DEVELOPMENTAL_DELAY_WITH_OR_WITHOUT_DYSMORPHIC_FACIES_AND_AUTISM': 4,\n",
       " 'Glycogen_phosphorylase_kinase_deficiency^Glycogen_storage_disease^Glycogen_storage_disease_IXd': 1,\n",
       " 'Disorder_of_organic_acid_metabolism^Disorders_of_Intracellular_Cobalamin_Metabolism^Methylmalonic_acidemia^Methylmalonic_acidemia_with_homocystinuria_cblD': 3,\n",
       " 'Megalencephaly-polymicrogyria-polydactyly-hydrocephalus_syndrome_1:.': 1,\n",
       " 'Joubert_syndrome_22': 4,\n",
       " 'Limb-Girdle_Muscular_Dystrophies_Autosomal_Dominant^Limb-girdle_muscular_dystrophy^Limb-girdle_muscular_dystrophy_type_1E': 1,\n",
       " 'Familial_episodic_pain_syndrome_1': 4,\n",
       " 'Complement_component_8_deficiency_type_1': 3,\n",
       " 'CHARGE_association^Hypogonadism_with_anosmia^Kallmann_syndrome_5': 4,\n",
       " 'Autoimmune_thrombocytopenia': 2,\n",
       " 'Cataract_30': 7,\n",
       " 'Hypogonadotropic_hypogonadism_7_with_or_without_anosmia^Hypomyelinating_leukodystrophy_7^Hypomyelinating_leukodystrophy_8_with_or_without_oligodontia_and/or_hypogonadotropic_hypogonadism^Pol_III-related_leukodystrophy': 1,\n",
       " \"Arthrogryposis_distal_with_impaired_proprioception_and_touch^Distal_arthrogryposis^Gordon's_syndrome^Marden-Walker_syndrome^Oculomelic_amyoplasia\": 9,\n",
       " 'Hereditary_diffuse_gastric_cancer^Microvascular_complications_of_diabetes^Microvascular_complications_of_diabetes_4^Osteomyelitis_sterile_multifocal_with_periostitis_and_pustulosis': 1,\n",
       " 'Carbohydrate-deficient_glycoprotein_syndrome_type_II^Congenital_disorder_of_glycosylation': 1,\n",
       " 'Charcot-Marie-Tooth_disease^Charcot-Marie-Tooth_disease_axonal_type_2O^Charcot-Marie-Tooth_disease_type_2^Mental_retardation_autosomal_dominant_13^Spinal_muscular_atrophy_lower_extremity_predominant_1_autosomal_dominant': 4,\n",
       " 'Mulibrey_nanism_syndrome': 9,\n",
       " 'Deafness_autosomal_recessive_44^Hereditary_hearing_loss_and_deafness^Nonsyndromic_Hearing_Loss_and_Deafness_Autosomal_Recessive^Nonsyndromic_hearing_loss_and_deafness': 2,\n",
       " 'Bardet-Biedl_syndrome^Bardet-Biedl_syndrome_9': 3,\n",
       " 'Adult_junctional_epidermolysis_bullosa^Epidermolysis_bullosa_junctionalis_with_pyloric_atresia^Epidermolysis_bullosa_simplex^Epidermolysis_bullosa_simplex_Cockayne-Touraine_type^Epidermolysis_bullosa_junctional': 10,\n",
       " 'Congenital_generalized_lipodystrophy^Congenital_generalized_lipodystrophy_type_1': 1,\n",
       " 'Galactosemia^UDPglucose-4-epimerase_deficiency': 2,\n",
       " 'Bethlem_myopathy_1^Collagen_VI-related_myopathy^Congenital_muscular_dystrophy^Myosclerosis_autosomal_recessive^Ullrich_congenital_muscular_dystrophy_1': 9,\n",
       " 'SHUKLA-VERNON_SYNDROME': 3,\n",
       " 'Spermatogenic_failure_9': 4,\n",
       " 'CONGENITAL_HEART_DEFECTS_AND_SKELETAL_MALFORMATIONS_SYNDROME^Chronic_myelogenous_leukemia': 1,\n",
       " 'Fanconi_renotubular_syndrome_3': 3,\n",
       " 'Nephronophthisis_14': 4,\n",
       " 'Natural_killer_cell_and_glucocorticoid_deficiency_with_DNA_repair_defect': 3,\n",
       " 'Adenylate_kinase_deficiency_hemolytic_anemia_due_to': 5,\n",
       " 'Congenital_hypothyroidism^Hyperthyroidism_familial_gestational^Hyperthyroidism_nonautoimmune^Hypothyroidism_congenital_nongoitrous_1': 1,\n",
       " 'Ciliary_dyskinesia_primary_30^Primary_ciliary_dyskinesia': 1,\n",
       " 'Microvascular_complications_of_diabetes^Microvascular_complications_of_diabetes_6': 1,\n",
       " 'Nephrotic_syndrome_type_5_with_or_without_ocular_abnormalities^Pierson_syndrome': 3,\n",
       " 'Carpenter_syndrome_2': 4,\n",
       " 'Epileptic_encephalopathy_early_infantile_23': 3,\n",
       " 'Fanconi_anemia^Fanconi_anemia_complementation_group_L': 3,\n",
       " 'Carcinoma_of_colon^Mosaic_variegated_aneuploidy_syndrome^Mosaic_variegated_aneuploidy_syndrome_1^Premature_chromatid_separation_trait': 6,\n",
       " 'Cromer_blood_group_system^Protein-losing_enteropathy': 2,\n",
       " 'Spastic_paraplegia_7': 8,\n",
       " 'Ischemic_stroke': 2,\n",
       " 'Deficiency_of_2-methylbutyryl-CoA_dehydrogenase': 1,\n",
       " '3-METHYLGLUTACONIC_ACIDURIA_TYPE_IX': 1,\n",
       " 'Ataxia_sensory_autosomal_dominant': 2,\n",
       " 'Anonychia': 1,\n",
       " 'DEAFNESS_AUTOSOMAL_RECESSIVE_115': 2,\n",
       " 'Mental_retardation_X-linked_syndromic_Turner_type': 5,\n",
       " 'Retinitis_pigmentosa^Retinitis_pigmentosa_45': 5,\n",
       " 'NEURODEGENERATION_CHILDHOOD-ONSET_WITH_BRAIN_ATROPHY': 3,\n",
       " 'ALG9_congenital_disorder_of_glycosylation^Congenital_disorder_of_glycosylation^Gillessen-Kaesbach-Nishimura_syndrome:.': 1,\n",
       " 'Chronic_infantile_neurological_cutaneous_and_articular_syndrome^DEAFNESS_AUTOSOMAL_DOMINANT_34_WITH_OR_WITHOUT_INFLAMMATION^Familial_amyloid_nephropathy_with_urticaria_AND_deafness^Familial_cold_urticaria^Keratitis_fugax_hereditaria': 1,\n",
       " 'Premature_ovarian_failure_10': 4,\n",
       " 'Gracile_bone_dysplasia^Kenny-Caffey_syndrome_type_2': 4,\n",
       " 'Charcot-Marie-Tooth_disease^Charcot-Marie-Tooth_disease_type_2F^Charcot-Marie-Tooth_disease_type_2^Distal_hereditary_motor_neuronopathy^Distal_hereditary_motor_neuronopathy_type_2B': 14,\n",
       " 'Familial_cancer_of_breast^Neoplasm_of_the_breast': 31,\n",
       " 'Dystonia_26_myoclonic': 1,\n",
       " 'Acute_lymphoid_leukemia^Chronic_myelogenous_leukemia': 3,\n",
       " 'Hypogonadotropic_hypogonadism_16_with_or_without_anosmia': 1,\n",
       " 'Hereditary_multiple_osteochondromas^Multiple_exostoses_type_2^Seizures_scoliosis_and_macrocephaly_syndrome': 1,\n",
       " 'Cerebral_creatine_deficiency_syndrome^Deficiency_of_guanidinoacetate_methyltransferase': 1,\n",
       " 'Cone-rod_dystrophy_15': 4,\n",
       " 'Parkinson_disease^Parkinson_disease_11': 7,\n",
       " 'Carcinoma_of_colon^Familial_colorectal_cancer^Hereditary_nonpolyposis_colorectal_cancer_type_6^Loeys-Dietz_syndrome^Loeys-Dietz_syndrome_2^Lynch_syndrome^Malignant_tumor_of_esophagus^Thoracic_aortic_aneurysm_and_aortic_dissection': 1,\n",
       " 'Cone_dystrophy_3': 1,\n",
       " 'FANCONI_ANEMIA_COMPLEMENTATION_GROUP_W': 1,\n",
       " 'Lethal_multiple_pterygium_syndrome^Multiple_pterygium_syndrome_Escobar_type': 2,\n",
       " 'Juvenile_neuronal_ceroid_lipofuscinosis^Neuronal_ceroid_lipofuscinosis': 1,\n",
       " 'Autosomal_dominant_torsion_dystonia_4^Leukodystrophy_hypomyelinating_6': 1,\n",
       " 'Aicardi_Goutieres_syndrome^Aicardi_Goutieres_syndrome_3': 2,\n",
       " 'Neuropathy_hereditary_sensory_and_autonomic_type_VIII': 1,\n",
       " 'Cataract_autosomal_dominant': 1,\n",
       " 'CHARGE_association^Hypogonadotropic_hypogonadism_7_with_or_without_anosmia': 1,\n",
       " 'Glycerol_release_during_exercise_defective': 2,\n",
       " 'Hereditary_pancreatitis:.': 3,\n",
       " 'Spastic_paraplegia_49_autosomal_recessive': 6,\n",
       " 'Cerebral_cavernous_malformation^Cerebral_cavernous_malformations_2': 3,\n",
       " 'Dystonia_25': 2,\n",
       " 'Cutis_laxa_autosomal_dominant^Cutis_laxa_autosomal_dominant_1^Supravalvar_aortic_stenosis^Williams_syndrome': 7,\n",
       " 'Cataract_33_multiple_types:.': 2,\n",
       " 'Bare_lymphocyte_syndrome_2': 4,\n",
       " 'Epilepsy_familial_temporal_lobe_7^Epilepsy_lateral_temporal_lobe_autosomal_dominant^Lissencephaly^Lissencephaly_2^Malformation_of_cortical_development': 8,\n",
       " 'Fanconi_anemia_complementation_group_Q^XFE_progeroid_syndrome^Xeroderma_pigmentosum^Xeroderma_pigmentosum_group_F': 2,\n",
       " 'Familial_partial_lipodystrophy^Familial_partial_lipodystrophy_6': 5,\n",
       " 'Duane-radial_ray_syndrome^IVIC_syndrome': 2,\n",
       " 'CILIARY_DYSKINESIA_PRIMARY_41': 7,\n",
       " 'Glycogen_storage_disease^Glycogen_storage_disease_type_X': 13,\n",
       " 'Acute_myeloid_leukemia^Gastrointestinal_stroma_tumor^Malignant_tumor_of_testis^Mastocytosis^Neoplasm_of_the_genitourinary_tract^Partial_albinism': 2,\n",
       " 'Cafe-au-lait_macules_with_pulmonary_stenosis^Juvenile_myelomonocytic_leukemia^Neurofibromatosis_familial_spinal^Neurofibromatosis_type_1^Neurofibromatosis-Noonan_syndrome^Rasopathy': 7,\n",
       " 'Adams-Oliver_syndrome^Adams-Oliver_syndrome_2': 6,\n",
       " 'Breast-ovarian_cancer_familial_4^Familial_cancer_of_breast^Hereditary_breast_and_ovarian_cancer_syndrome^Neoplasm_of_the_breast': 1,\n",
       " 'Uncombable_hair_syndrome_2': 4,\n",
       " 'INTELLECTUAL_DEVELOPMENTAL_DISORDER_AUTOSOMAL_RECESSIVE_70': 2,\n",
       " 'Erythrocytosis_familial_4^Familial_erythrocytosis': 1,\n",
       " 'Mental_retardation_autosomal_recessive_27': 2,\n",
       " 'Insulin-dependent_diabetes_mellitus_secretory_diarrhea_syndrome': 1,\n",
       " 'Asthma-related_traits_susceptibility_to_7^Schizophrenia': 3,\n",
       " 'Epileptic_encephalopathy_early_infantile_45': 1,\n",
       " 'CATSPER-Related_Male_Infertility': 2,\n",
       " 'Spastic_paraplegia_50_autosomal_recessive': 5,\n",
       " 'IRAK4_deficiency^Invasive_pneumococcal_disease_recurrent_isolated_1': 1,\n",
       " 'Body_mass_index_quantitative_trait_locus_12^Monogenic_Non-Syndromic_Obesity_Autosomal_Recessive^Proprotein_convertase_1/3_deficiency': 2,\n",
       " 'LOW_DENSITY_LIPOPROTEIN_CHOLESTEROL_LEVEL_QUANTITATIVE_TRAIT_LOCUS_7': 4,\n",
       " 'HENNEKAM_LYMPHANGIECTASIA-LYMPHEDEMA_SYNDROME_3': 1,\n",
       " 'Robinow_syndrome^Robinow_syndrome_autosomal_dominant_3': 1,\n",
       " 'Hypogonadotropic_hypogonadism_17_with_or_without_anosmia^Hypogonadotropic_hypogonadism_7_with_or_without_anosmia': 4,\n",
       " 'NEURODEGENERATION_EARLY-ONSET_WITH_CHOREOATHETOID_MOVEMENTS_AND_MICROCYTIC_ANEMIA': 1,\n",
       " 'Immunodeficiency_14': 1,\n",
       " 'SPERMATOGENIC_FAILURE_34': 12,\n",
       " 'Heterotaxy_visceral_8_autosomal': 4,\n",
       " 'Microphthalmia_isolated_8': 2,\n",
       " 'Ataxia_combined_cerebellar_and_peripheral_with_hearing_loss_and_diabetes_mellitus': 1,\n",
       " 'Deficiency_of_butyrylcholine_esterase': 1,\n",
       " 'Microcephaly_18_primary_autosomal_dominant': 4,\n",
       " 'Premature_ovarian_failure^Premature_ovarian_failure_5': 1,\n",
       " 'Seborrhea-like_dermatitis_with_psoriasiform_elements': 6,\n",
       " 'Pontocerebellar_hypoplasia_type_1^Pontocerebellar_hypoplasia_type_1b^Pontoneocerebellar_hypoplasia': 1,\n",
       " 'Gonadotropin-independent_familial_sexual_precocity^Leydig_cell_agenesis^Precocious_puberty:.': 1,\n",
       " 'BONE_MARROW_FAILURE_SYNDROME_4': 1,\n",
       " 'Abdominal_obesity-metabolic_syndrome_3': 7,\n",
       " 'Spinocerebellar_ataxia_12': 1,\n",
       " 'X-Linked_Mental_Retardation_41': 1,\n",
       " 'Uridine_5-prime_monophosphate_hydrolase_deficiency_hemolytic_anemia_due_to': 1,\n",
       " 'Alpha-methylacyl-CoA_racemase_deficiency^Bile_acid_synthesis_defect_congenital_4': 2,\n",
       " 'Loeys-Dietz_syndrome^Loeys-Dietz_syndrome_3^Thoracic_aortic_aneurysm_and_aortic_dissection': 1,\n",
       " 'Congenital_stationary_night_blindness_type_1D': 4,\n",
       " 'T-cell_immunodeficiency_congenital_alopecia_and_nail_dystrophy': 3,\n",
       " 'Cerebro-costo-mandibular_syndrome': 4,\n",
       " 'Oculocutaneous_albinism^Skin/hair/eye_pigmentation_variation_in_1^Tyrosinase-positive_oculocutaneous_albinism': 2,\n",
       " 'Hereditary_hearing_loss_and_deafness^Leigh_syndrome^Mitochondrial_diseases^Pyruvate_dehydrogenase_E1-alpha_deficiency^Pyruvate_dehydrogenase_complex_deficiency': 1,\n",
       " 'Congenital_contractural_arachnodactyly^Macular_degeneration_early-onset': 11,\n",
       " 'MENTAL_RETARDATION_AUTOSOMAL_RECESSIVE_61': 4,\n",
       " 'Osteogenesis_imperfecta^Osteogenesis_imperfecta_type_7': 2,\n",
       " 'Mental_retardation_autosomal_recessive_2': 4,\n",
       " 'Focal_segmental_glomerulosclerosis_4_susceptibility_to': 2,\n",
       " 'Tooth_agenesis_selective_3': 1,\n",
       " 'Atransferrinemia': 3,\n",
       " 'Diarrhea_4_malabsorptive_congenital': 1,\n",
       " 'Severe_congenital_neutropenia^Severe_congenital_neutropenia_X-linked^Thrombocytopenia_X-linked^Wiskott-Aldrich_syndrome': 1,\n",
       " 'COMBINED_IMMUNODEFICIENCY_AND_MEGALOBLASTIC_ANEMIA_WITH_OR_WITHOUT_HYPERHOMOCYSTEINEMIA^Neural_tube_defects_folate-sensitive': 1,\n",
       " 'OSTEOCHONDRODYSPLASIA_BRACHYDACTYLY_AND_OVERLAPPING_MALFORMED_DIGITS': 1,\n",
       " 'Gaze_palsy_familial_horizontal_with_progressive_scoliosis': 3,\n",
       " 'Parkinson_disease^Parkinson_disease_17': 1,\n",
       " 'Brown-Vialetto-Van_Laere_syndrome_1^Progressive_bulbar_palsy_of_childhood': 1,\n",
       " 'Hypermanganesemia_with_dystonia_2^Hyperostosis_cranialis_interna': 1,\n",
       " 'Susceptibility_to_malaria': 3,\n",
       " 'GONADAL_DYSGENESIS_DYSMORPHIC_FACIES_RETINAL_DYSTROPHY_AND_MYOPATHY^SPERMATOGENIC_FAILURE_36': 1,\n",
       " 'Early_infantile_epileptic_encephalopathy_34^Epilepsy_idiopathic_generalized_susceptibility_to_14': 9,\n",
       " 'Familial_hyperinsulinism^Hyperinsulinemic_hypoglycemia_familial_5^Insulin-resistant_diabetes_mellitus_AND_acanthosis_nigricans^Leprechaunism_syndrome^Pineal_hyperplasia_AND_diabetes_mellitus_syndrome': 1,\n",
       " 'Amyotrophic_lateral_sclerosis^Amyotrophic_lateral_sclerosis_type_1^Distal_hereditary_motor_neuronopathy_type_7B^Perry_syndrome': 2,\n",
       " 'MITOCHONDRIAL_COMPLEX_I_DEFICIENCY_NUCLEAR_TYPE_14:.': 1,\n",
       " 'DEAFNESS_AUTOSOMAL_RECESSIVE_99': 7,\n",
       " 'Congenital_muscular_dystrophy^Congenital_muscular_dystrophy-dystroglycanopathy_with_brain_and_eye_anomalies_type_A2^Congenital_muscular_dystrophy-dystroglycanopathy_with_mental_retardation_type_B2^Limb-girdle_muscular_dystrophy-dystroglycanopathy_type_C2^Muscular_dystrophy-dystroglycanopathy_(congenital_with_brain_and_eye_anomalies)_type_A_1^alpha-Dystroglycan_related_dystrophy': 3,\n",
       " 'Carney_complex_variant^Distal_arthrogryposis^Hecht_syndrome': 12,\n",
       " 'Resting_heart_rate^SHORT_SLEEP_FAMILIAL_NATURAL_2': 1,\n",
       " 'Bardet-Biedl_syndrome^Bardet-Biedl_syndrome_4': 2,\n",
       " 'Nemaline_myopathy^Nemaline_myopathy_7': 1,\n",
       " 'Homocystinuria_due_to_CBS_deficiency': 1,\n",
       " 'Apparent_mineralocorticoid_excess': 3,\n",
       " 'Wilson-Turner_X-linked_mental_retardation_syndrome': 1,\n",
       " 'Metachromatic_leukodystrophy': 3,\n",
       " 'Basal_cell_carcinoma_multiple^Craniofacial_malformations_asymmetric_with_polysyndactyly_and_abnormal_skin_and_gut_development': 4,\n",
       " 'Nasopharyngeal_carcinoma_susceptibility_to_3': 5,\n",
       " 'Mucopolysaccharidosis_type_VI': 118,\n",
       " 'Congenital_disorder_of_glycosylation_type_1w': 1,\n",
       " 'Congenital_ichthyosis_of_skin^Ichthyosis_congenital_autosomal_recessive_12': 1,\n",
       " 'Myopathy_areflexia_respiratory_distress_and_dysphagia_early-onset': 1,\n",
       " 'Camptodactyly_arthropathy_coxa_vara_pericarditis_syndrome': 10,\n",
       " \"Hereditary_hearing_loss_and_deafness^Leber's_optic_atrophy^Mitochondrial_DNA-Associated_Leigh_Syndrome_and_NARP^Mitochondrial_diseases^Neuropathy_ataxia_retinitis_pigmentosa_syndrome^Retinitis_pigmentosa\": 11,\n",
       " 'Hereditary_hearing_loss_and_deafness^Multiple_Epiphyseal_Dysplasia_Dominant^Multiple_epiphyseal_dysplasia_2^Stickler_syndrome^Stickler_syndrome_type_5': 4,\n",
       " 'MITOCHONDRIAL_COMPLEX_I_DEFICIENCY_NUCLEAR_TYPE_6': 3,\n",
       " 'Peroxisomal_fatty_acyl-coa_reductase_1_disorder': 1,\n",
       " 'Familial_idiopathic_hypercalciuria': 8,\n",
       " 'Metaphyseal_anadysplasia_2': 1,\n",
       " 'BLEPHAROCHEILODONTIC_SYNDROME_2': 4,\n",
       " 'Joubert_syndrome^Joubert_syndrome_10^Oral-facial-digital_syndrome^Retinitis_Pigmentosa_23^Simpson-Golabi-Behmel_syndrome_type_2': 3,\n",
       " 'Treacher_Collins_syndrome^Treacher_Collins_syndrome_1': 14,\n",
       " 'Elliptocytosis_2^Hereditary_elliptocytosis^Hereditary_pyropoikilocytosis^Hereditary_spherocytosis^Spherocytosis_type_3': 14,\n",
       " 'Holt-Oram_syndrome': 2,\n",
       " 'Orofaciodigital_syndrome_xiv': 11,\n",
       " 'Fanconi_anemia^Fanconi_anemia_complementation_group_B^VACTERL_association_with_hydrocephaly_X-linked': 1,\n",
       " 'Charcot-Marie-Tooth_disease^Charcot-Marie-Tooth_disease_type_4^Charcot-Marie-Tooth_disease_demyelinating_type_4F^Dejerine-Sottas_disease': 7,\n",
       " 'ARTHROGRYPOSIS_MULTIPLEX_CONGENITA_MYOGENIC_TYPE^Emery-Dreifuss_muscular_dystrophy_4_autosomal_dominant^Spinocerebellar_ataxia_autosomal_recessive_8': 19,\n",
       " 'Cataract_autosomal_recessive_congenital_5^Sengers_syndrome': 2,\n",
       " 'Congenital_myasthenic_syndrome^Congenital_myasthenic_syndrome_acetazolamide-responsive^Hereditary_hearing_loss_and_deafness^Hyperkalemic_Periodic_Paralysis_Type_1^Hypokalemic_periodic_paralysis^Hypokalemic_periodic_paralysis_1^Hypokalemic_periodic_paralysis_type_2^Leigh_syndrome^Mitochondrial_diseases^Paramyotonia_congenita_of_von_Eulenburg^Potassium_aggravated_myotonia': 23,\n",
       " 'Autoimmune_disease_syndromic_multisystem:.': 1,\n",
       " 'Digitotalar_dysmorphism^Distal_arthrogryposis^Distal_arthrogryposis_type_1B^Lethal_congenital_contracture_syndrome_4^MYOPATHY_CONGENITAL_WITH_TREMOR': 2,\n",
       " 'Fucosyltransferase_6_deficiency': 1,\n",
       " 'Congenital_disorder_of_glycosylation^Congenital_disorder_of_glycosylation_type_1K': 1,\n",
       " 'Primary_autosomal_recessive_microcephaly^Primary_autosomal_recessive_microcephaly_5': 5,\n",
       " 'Desmosterolosis': 1,\n",
       " 'Complement_component_6_deficiency': 1,\n",
       " 'Odontoonychodermal_dysplasia^Schopf-Schulz-Passarge_syndrome^Tooth_agenesis_selective_4': 3,\n",
       " 'Familial_cancer_of_breast^Lung_cancer^Neoplasm_of_lung^Neoplasm_of_the_breast^Rhabdomyosarcoma_1': 4,\n",
       " 'EHLERS-DANLOS_SYNDROME_CLASSIC-LIKE_2': 3,\n",
       " 'Acromesomelic_dysplasia_Hunter_Thompson_type^Brachydactyly_type_a1_c^Fibular_hypoplasia_and_complex_brachydactyly^Grebe_syndrome^Multiple_synostoses_syndrome_2^Osteoarthritis_of_hip^Symphalangism_proximal_1b^Type_A2_brachydactyly^Type_C_brachydactyly': 1,\n",
       " 'Rajab_syndrome': 1,\n",
       " 'Anterior_segment_dysgenesis_7': 3,\n",
       " 'Martsolf_syndrome^Warburg_micro_syndrome^Warburg_micro_syndrome_2': 124,\n",
       " 'Acromesomelic_dysplasia_Demirhan_type^Brachydactyly_type_a1_d^Type_A2_brachydactyly': 1,\n",
       " 'Glomerulopathy_with_fibronectin_deposits_2^Plasma_fibronectin_deficiency^Spondylometaphyseal_dysplasia_-_Sutcliffe_type': 20,\n",
       " 'Joubert_syndrome_23^Short-rib_thoracic_dysplasia_14_with_polydactyly': 6,\n",
       " 'Chromosome_Xq26.3_duplication_syndrome^Pituitary_adenoma_growth_hormone-secreting_2': 2,\n",
       " 'Birk_Barel_mental_retardation_dysmorphism_syndrome': 1,\n",
       " 'Cerebral_creatine_deficiency_syndrome^Creatine_deficiency_X-linked': 1,\n",
       " 'KBG_syndrome': 4,\n",
       " 'Craniosynostosis_1^Craniosynostosis_syndrome^Robinow_Sorauf_syndrome^SWEENEY-COX_SYNDROME^Saethre-Chotzen_syndrome': 1,\n",
       " 'Nephrotic_syndrome_type_6': 1,\n",
       " 'Protein_Z_deficiency': 3,\n",
       " 'Combined_saposin_deficiency^Gaucher_disease^Gaucher_disease_atypical_due_to_saposin_C_deficiency^Krabbe_disease_atypical_due_to_Saposin_A_deficiency^Sphingolipid_activator_protein_1_deficiency': 3,\n",
       " 'Primary_autosomal_recessive_microcephaly^Primary_autosomal_recessive_microcephaly_13': 3,\n",
       " 'Autosomal_recessive_woolly_hair_3': 3,\n",
       " 'Psychomotor_retardation_epilepsy_and_craniofacial_dysmorphism': 1,\n",
       " 'Mental_retardation_autosomal_recessive_49': 2,\n",
       " 'Familial_hypercholesterolemias^Hypercholesterolemia_autosomal_dominant_3': 2,\n",
       " 'NEURODEVELOPMENTAL_DISORDER_WITH_SEIZURES_AND_SPEECH_AND_WALKING_IMPAIRMENT': 2,\n",
       " 'Diamond-Blackfan_anemia^Diamond-Blackfan_anemia_8': 1,\n",
       " 'Congenital_defect_of_folate_absorption': 2,\n",
       " 'Retinitis_pigmentosa^Retinitis_pigmentosa_38': 3,\n",
       " 'Corticosterone_18-monooxygenase_deficiency^Corticosterone_methyloxidase_type_1_deficiency^Corticosterone_methyloxidase_type_2_deficiency': 2,\n",
       " 'Warsaw_breakage_syndrome': 7,\n",
       " 'Abetalipoproteinaemia^Metabolic_syndrome_X': 1,\n",
       " 'Fanconi_anemia_complementation_group_U': 2,\n",
       " 'Birk-Landau-Perez_syndrome': 2,\n",
       " 'Mental_retardation_autosomal_dominant_41^Pierpont_syndrome': 1,\n",
       " 'Spinocerebellar_ataxia_23': 3,\n",
       " 'DEAFNESS_AUTOSOMAL_RECESSIVE_57^Hereditary_hearing_loss_and_deafness^Usher_syndrome_type_2^Usher_syndrome_type_2A^Usher_syndrome_type_2C': 5,\n",
       " 'Agammaglobulinemia_X-linked_type_2': 2,\n",
       " 'GALLOWAY-MOWAT_SYNDROME_8^NEPHROTIC_SYNDROME_TYPE_18': 4,\n",
       " 'Hemorrhage_intracerebral_susceptibility_to^Porencephaly_2': 1,\n",
       " 'Ataxia_spastic_1_autosomal_dominant^MYASTHENIC_SYNDROME_CONGENITAL_25_PRESYNAPTIC': 1,\n",
       " 'Cerebral_palsy_spastic_quadriplegic_3': 1,\n",
       " 'Age-related_macular_degeneration^Age-related_macular_degeneration_2^Cone-rod_dystrophy_3^Retinitis_pigmentosa^Retinitis_pigmentosa_19^Stargardt_disease^Stargardt_disease_1': 10,\n",
       " 'Opitz-Frias_syndrome': 1,\n",
       " 'Ventricular_tachycardia_catecholaminergic_polymorphic_3': 1,\n",
       " 'Congenital_lactase_deficiency': 4,\n",
       " 'Synpolydactyly_2': 3,\n",
       " 'Carcinoma_of_colon^Familial_colorectal_cancer^Lynch_syndrome^Lynch_syndrome_I': 1,\n",
       " 'DEAFNESS_AUTOSOMAL_RECESSIVE_109': 1,\n",
       " 'Craniofacial_anomalies_and_anterior_segment_dysgenesis_syndrome^Keratoconus_1': 5,\n",
       " 'Thyroid_cancer_nonmedullary_5^Venous_thrombosis': 1,\n",
       " 'Deafness_autosomal_dominant_56': 6,\n",
       " 'Encephalopathy_due_to_defective_mitochondrial_and_peroxisomal_fission_2': 11,\n",
       " 'Weill-Marchesani_syndrome_4': 1,\n",
       " 'Progesterone_resistance': 1,\n",
       " 'Hyperphosphatasia_with_mental_retardation_syndrome_3^Hyperphosphatasia-intellectual_disability_syndrome': 1,\n",
       " 'Spinal_muscular_atrophy': 2,\n",
       " 'Orofacial_cleft_15': 6,\n",
       " 'Mosaic_variegated_aneuploidy_syndrome^Mosaic_variegated_aneuploidy_syndrome_2': 2,\n",
       " 'Lethal_multiple_pterygium_syndrome^Myasthenic_syndrome_congenital_3a_slow-channel^Myasthenic_syndrome_congenital_3b_fast-channel^Myasthenic_syndrome_congenital_3c_associated_with_acetylcholine_receptor_deficiency': 1,\n",
       " 'Spondyloepimetaphyseal_dysplasia_Faden-Alkuraya_type': 2,\n",
       " 'Common_variable_immunodeficiency_10': 4,\n",
       " 'Multiple_congenital_anomalies-hypotonia-seizures_syndrome_1': 5,\n",
       " 'Pontocerebellar_hypoplasia_type_2^Pontocerebellar_hypoplasia_type_2f^Pontoneocerebellar_hypoplasia': 1,\n",
       " 'Atrial_fibrillation^Atrial_fibrillation_familial_4^Long_QT_syndrome_6^Romano-Ward_syndrome': 1,\n",
       " 'Hemochromatosis_type_5^Hereditary_hemochromatosis': 8,\n",
       " 'RENAL_HYPODYSPLASIA/APLASIA_3': 2,\n",
       " 'Anauxetic_dysplasia_2': 3,\n",
       " 'Hypohidrotic_X-linked_ectodermal_dysplasia^Hypohidrotic_ectodermal_dysplasia^Tooth_agenesis_selective_X-linked_1': 1,\n",
       " 'URIDINE-CYTIDINEURIA': 1,\n",
       " 'Immunodeficiency_20': 1,\n",
       " 'Acromelic_frontonasal_dysostosis^NEURODEVELOPMENTAL_DISORDER_WITH_MOVEMENT_ABNORMALITIES_ABNORMAL_GAIT_AND_AUTISTIC_FEATURES': 5,\n",
       " 'Mitochondrial_complex_III_deficiency_nuclear_type_4': 2,\n",
       " 'DEAFNESS_AUTOSOMAL_DOMINANT_37^Fibrochondrogenesis^Hereditary_hearing_loss_and_deafness^Intervertebral_disc_disorder^Marshall_syndrome^Stickler_syndrome^Stickler_syndrome_type_2': 5,\n",
       " 'Coenzyme_Q10_deficiency_primary_2': 1,\n",
       " 'Ashkenazi_Jewish_disorders^Mucolipidosis_type_IV': 3,\n",
       " 'Salt_and_pepper_developmental_regression_syndrome': 1,\n",
       " 'Polyneuropathy_hearing_loss_ataxia_retinitis_pigmentosa_and_cataract': 1,\n",
       " 'Doyne_honeycomb_retinal_dystrophy': 1,\n",
       " 'Segawa_syndrome_autosomal_recessive': 2,\n",
       " 'Ceroid_lipofuscinosis_neuronal_5^Neuronal_ceroid_lipofuscinosis': 3,\n",
       " 'Congenital_stationary_night_blindness_type_1E': 20,\n",
       " 'Spastic_paraplegia_18': 1,\n",
       " 'Poikiloderma_hereditary_fibrosing_with_tendon_contractures_myopathy_and_pulmonary_fibrosis': 1,\n",
       " 'Meckel_syndrome_13^OROFACIODIGITAL_SYNDROME_XVI': 1,\n",
       " 'DEAFNESS_AUTOSOMAL_DOMINANT_74': 3,\n",
       " 'Acrodysostosis_1_with_or_without_hormone_resistance^Atrial_myxoma_familial^Carney_complex^Carney_complex_type_1^Pigmented_nodular_adrenocortical_disease_primary_1^Primary_pigmented_nodular_adrenocortical_disease': 4,\n",
       " 'Epidermolysis_bullosa_simplex_generalized_with_scarring_and_hair_loss': 4,\n",
       " 'Epileptic_encephalopathy_early_infantile_49': 2,\n",
       " 'Adult_hypophosphatasia^Childhood_hypophosphatasia^Hypophosphatasia^Infantile_hypophosphatasia': 1,\n",
       " 'NEURODEVELOPMENTAL_DISORDER_WITH_PROGRESSIVE_MICROCEPHALY_SPASTICITY_AND_BRAIN_ANOMALIES': 30,\n",
       " 'McLeod_neuroacanthocytosis_syndrome': 1,\n",
       " 'MICROCEPHALY_23_PRIMARY_AUTOSOMAL_RECESSIVE': 3,\n",
       " 'Autosomal_dominant_non-syndromic_sensorineural_deafness_type_DFNA^Deafness_autosomal_dominant_4^Hereditary_hearing_loss_and_deafness^Nonsyndromic_hearing_loss_and_deafness^Peripheral_neuropathy_myopathy_hoarseness_and_hearing_loss': 12,\n",
       " 'Auditory_neuropathy_autosomal_dominant_1': 1,\n",
       " 'Smith-McCort_dysplasia_2': 2,\n",
       " 'Early_infantile_epileptic_encephalopathy_59^NEURODEVELOPMENTAL_DISORDER_WITH_POOR_LANGUAGE_AND_LOSS_OF_HAND_SKILLS^Tobacco_addiction_susceptibility_to': 1,\n",
       " 'Keratolytic_winter_erythema': 6,\n",
       " 'Immunodeficiency_17^Immunodeficiency_due_to_defect_in_CD3-gamma': 1,\n",
       " 'Hodgkin_lymphoma': 1,\n",
       " 'Epilepsy_familial_focal_with_variable_foci_1': 1,\n",
       " 'Preeclampsia/eclampsia_5': 11,\n",
       " 'Ehlers-Danlos_syndrome_musculocontractural_type_2': 2,\n",
       " 'Anti-plasmin_deficiency_congenital': 3,\n",
       " 'Acrodysostosis_2_with_or_without_hormone_resistance': 219,\n",
       " 'Mental_retardation_X-linked_104': 1,\n",
       " 'PARAGANGLIOMAS_6': 1,\n",
       " 'Autosomal_recessive_polycystic_kidney_disease': 94,\n",
       " 'Primary_autosomal_recessive_microcephaly^Primary_autosomal_recessive_microcephaly_10': 1,\n",
       " 'Mental_retardation_63_X-linked': 1,\n",
       " '.:Hypogonadism_diabetes_mellitus_alopecia_mental_retardation_and_electrocardiographic_abnormalities^Neurodegeneration_with_brain_iron_accumulation': 1,\n",
       " 'Hypogonadism_with_anosmia^Kallmann_syndrome_6': 1,\n",
       " 'Hermansky_Pudlak_syndrome_2^Hermansky-Pudlak_syndrome': 5,\n",
       " 'Microcephaly-capillary_malformation_syndrome': 1,\n",
       " 'Congenital_hydrocephalus^Hydrocephalus^Hydrocephalus_congenital_2_with_or_without_brain_or_eye_anomalies': 11,\n",
       " 'Pseudohypoaldosteronism_type_2E^Pseudohypoaldosteronism_type_2': 6,\n",
       " 'Ataxia-telangiectasia_syndrome^Familial_cancer_of_breast^Neoplasm_of_the_breast': 8,\n",
       " 'BLOOD_GROUP--LUTHERAN_INHIBITOR^Congenital_dyserythropoietic_anemia^Congenital_dyserythropoietic_anemia_type_IV^Fetal_hemoglobin_quantitative_trait_locus_6': 1,\n",
       " 'Mental_retardation_autosomal_dominant_18': 1,\n",
       " 'Adult_neuronal_ceroid_lipofuscinosis^Ceroid_lipofuscinosis_neuronal_6^Neuronal_ceroid_lipofuscinosis': 2,\n",
       " 'Telangiectasia_hereditary_hemorrhagic_type_5': 2,\n",
       " 'Hermansky-Pudlak_syndrome^Hermansky-Pudlak_syndrome_7': 1,\n",
       " 'Congenital_hypothyroidism^Thyroid_dysgenesis': 2,\n",
       " 'Factor_VII_deficiency^Myocardial_infarction_1': 1,\n",
       " 'Dyskeratosis_congenita^Dyskeratosis_congenita_X-linked^Hoyeraal_Hreidarsson_syndrome': 1,\n",
       " 'Cholestasis_progressive_familial_intrahepatic_5': 1,\n",
       " 'Anophthalmia-microphthalmia_syndrome^Microphthalmia_isolated_3': 2,\n",
       " 'Niemann-Pick_disease_type_C1^Niemann-Pick_disease_type_C^Sphingomyelin/cholesterol_lipidosis': 3,\n",
       " 'Myasthenic_syndrome_congenital_21_presynaptic': 1,\n",
       " 'MITOCHONDRIAL_COMPLEX_I_DEFICIENCY_NUCLEAR_TYPE_22': 2,\n",
       " 'Ataxia-oculomotor_apraxia_4^Early_infantile_epileptic_encephalopathy_10': 2,\n",
       " 'Epidermolysis_bullosa_simplex_autosomal_recessive_2^Neuropathy_hereditary_sensory_and_autonomic_type_VI': 18,\n",
       " 'Corneal_fragility_keratoglobus_blue_sclerae_AND_joint_hypermobility': 17,\n",
       " 'Amyotrophic_lateral_sclerosis^Amyotrophic_lateral_sclerosis_type_1^Charcot-Marie-Tooth_disease_axonal_type_2CC': 2,\n",
       " 'Spastic_ataxia_Charlevoix-Saguenay_type': 9,\n",
       " 'Early_infantile_epileptic_encephalopathy_61': 1,\n",
       " 'MENTAL_RETARDATION_AUTOSOMAL_RECESSIVE_66': 1,\n",
       " 'Carcinoma_of_colon^Carcinoma_of_pancreas^Familial_cancer_of_breast^Familial_colorectal_cancer^Malignant_tumor_of_testis^Neoplasm_of_the_breast^Neoplasm_of_the_genitourinary_tract^Peutz-Jeghers_syndrome': 29,\n",
       " 'Basal_cell_carcinoma_multiple^Gorlin_syndrome^Holoprosencephaly_7^Holoprosencephaly_sequence': 5,\n",
       " 'Neurodegeneration_with_brain_iron_accumulation_6^PONTOCEREBELLAR_HYPOPLASIA_TYPE_12': 1,\n",
       " 'L-2-hydroxyglutaric_aciduria': 4,\n",
       " 'Parkinson_disease_23_autosomal_recessive_early-onset': 7,\n",
       " 'Ashkenazi_Jewish_disorders^Bronchiectasis_with_or_without_elevated_sweat_chloride_1^Congenital_bilateral_absence_of_the_vas_deferens^Cystic_fibrosis^Hereditary_pancreatitis^Ivacaftor_response': 5,\n",
       " 'Mucopolysaccharidosis_type_VII': 1,\n",
       " 'Renal_hamartomas_nephroblastomatosis_and_fetal_gigantism': 5,\n",
       " 'Bethlem_myopathy_1^Collagen_VI-related_myopathy^Congenital_muscular_dystrophy^Dystonia_27^Ullrich_congenital_muscular_dystrophy_1': 8,\n",
       " 'Seckel_syndrome_9': 22,\n",
       " 'Dyggve-Melchior-Clausen_syndrome^Smith-McCort_dysplasia_1': 1,\n",
       " 'Retinitis_pigmentosa^Retinitis_pigmentosa_26': 2,\n",
       " 'Keratosis_pilaris': 12,\n",
       " 'Spastic_paraplegia_46_autosomal_recessive': 5,\n",
       " '.:Retinitis_pigmentosa_78': 2,\n",
       " 'Budd-Chiari_syndrome^Factor_V_deficiency^Ischemic_stroke^Recurrent_abortion^Thrombophilia_due_to_activated_protein_C_resistance^Venous_thrombosis': 7,\n",
       " 'Hypomyelination_global_cerebral': 1,\n",
       " 'Leber_congenital_amaurosis^Leber_congenital_amaurosis_14^Retinitis_pigmentosa': 1,\n",
       " 'Hypotrichosis_and_recurrent_skin_vesicles': 3,\n",
       " 'Deafness_autosomal_recessive_22^Hereditary_hearing_loss_and_deafness^Nonsyndromic_Hearing_Loss_and_Deafness_Autosomal_Recessive^Nonsyndromic_hearing_loss_and_deafness': 4,\n",
       " \"Alpha-2-macroglobulin_deficiency^Alzheimer's_disease\": 2,\n",
       " 'Mental_retardation_autosomal_recessive_34': 2,\n",
       " 'Congenital_disorder_of_glycosylation^Congenital_disorder_of_glycosylation_type_2H': 2,\n",
       " 'Amyotrophic_lateral_sclerosis_22_with_or_without_frontotemporal_dementia': 4,\n",
       " 'Infantile_liver_failure_syndrome_2^Short_stature_optic_nerve_atrophy_and_Pelger-Huet_anomaly': 4,\n",
       " 'Autosomal_dominant_striatal_neurodegeneration^Dyskinesia_limb_and_orofacial_infantile-onset^Striatal_degeneration_autosomal_dominant_2': 2,\n",
       " 'Agammaglobulinemia_8_autosomal_dominant': 4,\n",
       " 'Roberts-SC_phocomelia_syndrome': 1,\n",
       " 'Glaucoma^Glaucoma_1_open_angle_F': 4,\n",
       " 'SHORT_STATURE_HEARING_LOSS_RETINITIS_PIGMENTOSA_AND_DISTINCTIVE_FACIES': 1,\n",
       " 'Nephrotic_syndrome_type_13': 6,\n",
       " 'BLOOD_GROUP--FROESE^BLOOD_GROUP--SWANN_SYSTEM^BLOOD_GROUP--WALDNER_TYPE^BLOOD_GROUP--WRIGHT_ANTIGEN^Hereditary_spherocytosis^Ovalocytosis_southeast_Asian^Pseudohyperkalemia_Cardiff^Renal_tubular_acidosis_distal_autosomal_dominant^Renal_tubular_acidosis_distal_with_hemolytic_anemia^Spherocytosis_type_4^Susceptibility_to_malaria': 9,\n",
       " 'Epilepsy_juvenile_absence^Idiopathic_generalized_epilepsy^Juvenile_myoclonic_epilepsy': 6,\n",
       " 'Spinocerebellar_ataxia_autosomal_recessive_2': 3,\n",
       " 'Arthrogryposis_multiplex_congenita_neurogenic_with_myelin_defect': 3,\n",
       " 'Cutaneous_malignant_melanoma_3^Cutaneous_melanoma^Hereditary_cutaneous_melanoma^Malignant_melanoma_of_skin': 4,\n",
       " 'Alpha-fetoprotein_deficiency^Alpha-fetoprotein_hereditary_persistence_of': 2,\n",
       " 'EPILEPTIC_ENCEPHALOPATHY_EARLY_INFANTILE_68': 6,\n",
       " 'Dopamine_beta_hydroxylase_deficiency': 3,\n",
       " 'Parkinson_disease^Parkinson_disease_7': 1,\n",
       " 'Gray_platelet_syndrome': 7,\n",
       " 'Spastic_paraplegia_64_autosomal_recessive': 5,\n",
       " 'PONTOCEREBELLAR_HYPOPLASIA_TYPE_11^Pontoneocerebellar_hypoplasia': 3,\n",
       " 'Arthrogryposis_renal_dysfunction_and_cholestasis_2': 1,\n",
       " 'Renal_dysplasia_diffuse_cystic': 3,\n",
       " 'Spondyloepimetaphyseal_dysplasia_with_multiple_dislocations': 2,\n",
       " 'Retinitis_pigmentosa^Retinitis_pigmentosa_31': 1,\n",
       " 'Deafness_autosomal_recessive_9^Hereditary_hearing_loss_and_deafness^Nonsyndromic_Hearing_Loss_and_Deafness_Autosomal_Recessive^Nonsyndromic_hearing_loss_and_deafness': 3,\n",
       " 'Amish_lethal_microcephaly^Striatal_necrosis_bilateral_and_progressive_polyneuropathy': 1,\n",
       " 'Age-related_macular_degeneration^Age-related_macular_degeneration_5^Cerebro-oculo-facio-skeletal_syndrome^Cockayne_syndrome^Cockayne_syndrome_B^DE_SANCTIS-CACCHIONE_SYNDROME^Lung_cancer^Neoplasm_of_lung^Premature_ovarian_failure_11^UV-sensitive_syndrome': 4,\n",
       " 'Juvenile_neuronal_ceroid_lipofuscinosis^Neuronal_ceroid_lipofuscinosis:.': 1,\n",
       " 'Diamond-Blackfan_anemia^Diamond-Blackfan_anemia_4': 2,\n",
       " 'Body_mass_index_quantitative_trait_locus_10': 1,\n",
       " 'Hypomyelination_and_Congenital_Cataract': 1,\n",
       " 'Trichomegaly': 1,\n",
       " 'FETAL_AKINESIA_DEFORMATION_SEQUENCE_4': 6,\n",
       " 'Rheumatoid_arthritis': 3,\n",
       " 'Autoimmune_lymphoproliferative_syndrome_type_III': 4,\n",
       " 'Hypogonadism_diabetes_mellitus_alopecia_mental_retardation_and_electrocardiographic_abnormalities^Neurodegeneration_with_brain_iron_accumulation': 1,\n",
       " 'FIBROMATOSIS_GINGIVAL_5^Wilms_tumor_6': 5,\n",
       " 'Ritscher-schinzel_syndrome_2': 2,\n",
       " 'AORTIC_VALVE_DISEASE_3': 3,\n",
       " 'Ichthyosis_bullosa_of_Siemens': 6,\n",
       " 'Xeroderma_pigmentosum^Xeroderma_pigmentosum_variant_type': 3,\n",
       " 'Spermatogenic_failure_14': 1,\n",
       " 'Arrhythmogenic_right_ventricular_cardiomyopathy^Arrhythmogenic_right_ventricular_cardiomyopathy_type_9^Cardiomyopathy': 5,\n",
       " 'OPHTHALMOPLEGIA_EXTERNAL_WITH_RIB_AND_VERTEBRAL_ANOMALIES': 1,\n",
       " 'Acute_myeloid_leukemia^Aplastic_anemia^Cutaneous_malignant_melanoma_9^Cutaneous_melanoma^Dyskeratosis_congenita^Dyskeratosis_congenita_autosomal_dominant^Dyskeratosis_congenita_autosomal_dominant_2^Hereditary_cutaneous_melanoma^Idiopathic_fibrosing_alveolitis_chronic_form^Malignant_melanoma_of_skin^Pulmonary_fibrosis_and/or_bone_marrow_failure_telomere-related_1': 1,\n",
       " 'EXTRAORAL_HALITOSIS_DUE_TO_METHANETHIOL_OXIDASE_DEFICIENCY': 7,\n",
       " 'Anophthalmia-microphthalmia_syndrome^Septo-optic_dysplasia_sequence': 1,\n",
       " 'Retinitis_pigmentosa_84': 4,\n",
       " 'Smith-Magenis_syndrome': 12,\n",
       " 'Oculoauricular_syndrome': 1,\n",
       " 'Non-ketotic_hyperglycinemia': 3,\n",
       " 'KLEEFSTRA_SYNDROME_2': 10,\n",
       " 'Glycogen_storage_disease^Glycogen_storage_disease_type_13': 2,\n",
       " 'Deafness_and_myopia': 3,\n",
       " 'Cardiomyopathy^Dilated_cardiomyopathy_1G^Distal_myopathy_Markesbery-Griggs_type^Familial_dilated_cardiomyopathy^Familial_hypertrophic_cardiomyopathy_9^Hereditary_myopathy_with_early_respiratory_failure^Limb-girdle_muscular_dystrophy^Limb-girdle_muscular_dystrophy_autosomal_recessive^Limb-girdle_muscular_dystrophy_type_2J^Myopathy_early-onset_with_fatal_cardiomyopathy^Primary_dilated_cardiomyopathy^Primary_familial_hypertrophic_cardiomyopathy': 83,\n",
       " 'Avascular_necrosis_of_femoral_head_primary_2^Brachyrachia_(short_spine_dysplasia)^Charcot-Marie-Tooth_disease^Charcot-Marie-Tooth_disease_type_2C^Charcot-Marie-Tooth_disease_type_2^Digital_arthropathy-brachydactyly_familial^Distal_spinal_muscular_atrophy_congenital_nonprogressive^Familial_avascular_necrosis_of_the_femoral_head^Metatrophic_dysplasia^Parastremmatic_dwarfism^Scapuloperoneal_spinal_muscular_atrophy^Sodium_serum_level_quantitative_trait_locus_1^Spondyloepiphyseal_dysplasia_Maroteaux_type^Spondylometaphyseal_dysplasia_Kozlowski_type': 2,\n",
       " 'Ciliary_dyskinesia_primary_34^Primary_ciliary_dyskinesia': 1,\n",
       " 'Hereditary_Nephrotic_Syndromes_Autosomal_Recessive^Hereditary_nephrotic_syndrome^Nephrotic_syndrome_idiopathic_steroid-resistant': 2,\n",
       " 'Dyskeratosis_congenita^Dyskeratosis_congenita_autosomal_dominant^Dyskeratosis_congenita_autosomal_dominant_3^Revesz_syndrome': 3,\n",
       " 'Plasminogen_activator_inhibitor_type_1_deficiency': 4,\n",
       " 'Nystagmus_6_congenital_X-linked^Ocular_albinism_type_I': 1,\n",
       " 'Retinitis_pigmentosa_78': 2,\n",
       " 'Carcinoma_of_colon^Diarrhea_5_with_tufting_enteropathy_congenital^Familial_colorectal_cancer^Hereditary_nonpolyposis_colorectal_cancer_type_8^Lynch_syndrome': 1,\n",
       " 'Aminoacylase_1_deficiency:.': 3,\n",
       " 'Immunodeficiency-centromeric_instability-facial_anomalies_syndrome_3': 1,\n",
       " 'Raine_syndrome': 2,\n",
       " 'CARDIAC-UROGENITAL_SYNDROME^ENCEPHALITIS/ENCEPHALOPATHY_MILD_WITH_REVERSIBLE_MYELIN_VACUOLIZATION': 3,\n",
       " 'Congenital_disorder_of_glycosylation^Epileptic_encephalopathy_early_infantile_36': 2,\n",
       " 'Cerebrooculofacioskeletal_syndrome_3^Xeroderma_pigmentosum^Xeroderma_pigmentosum_group_G:.': 1,\n",
       " 'Hereditary_hearing_loss_and_deafness^Mitochondrial_DNA_Deletion_Syndromes^Mitochondrial_DNA_Depletion_Syndrome_Encephalomyopathic_Form^Mitochondrial_DNA_depletion_syndrome_9_(encephalomyopathic_with_methylmalonic_aciduria)^Mitochondrial_diseases': 6,\n",
       " 'Deficiency_of_UDPglucose-hexose-1-phosphate_uridylyltransferase^Galactosemia': 1,\n",
       " 'Polycystic_liver_disease_2': 2,\n",
       " 'Disorder_of_fatty_acid_metabolism^Glutaric_aciduria_type_2': 2,\n",
       " 'High_molecular_weight_kininogen_deficiency': 1,\n",
       " 'Hypotonia_infantile_with_psychomotor_retardation_and_characteristic_facies_2': 5,\n",
       " 'Charcot-Marie-Tooth_disease^Charcot-Marie-Tooth_disease_type_4^Hemolytic_anemia_due_to_hexokinase_deficiency^NEURODEVELOPMENTAL_DISORDER_WITH_VISUAL_DEFECTS_AND_BRAIN_ANOMALIES^Neuropathy_hereditary_motor_and_sensory_Russe_type^Retinitis_pigmentosa_79': 83,\n",
       " 'Cryptophthalmos_unilateral_or_bilateral_isolated^FRASER_SYNDROME_2': 8,\n",
       " 'Immunodeficiency_with_hyper_IgM_type_1': 1,\n",
       " 'Deafness_autosomal_recessive_39^Hereditary_hearing_loss_and_deafness^Nonsyndromic_Hearing_Loss_and_Deafness_Autosomal_Recessive^Nonsyndromic_hearing_loss_and_deafness': 1,\n",
       " 'Spondylometaphyseal_dysplasia_Sedaghatian_type': 10,\n",
       " 'Meckel_syndrome_type_10^Meckel-Gruber_syndrome': 1,\n",
       " 'Platelet-type_bleeding_disorder_13_susceptibility_to': 1,\n",
       " 'Brugada_syndrome^Brugada_syndrome_4': 8,\n",
       " 'Hypogonadism_with_anosmia^Kallmann_syndrome_1': 1,\n",
       " 'Nonpersistence_of_intestinal_lactase': 11,\n",
       " 'Diabetes_mellitus_noninsulin-dependent_1': 5,\n",
       " 'Age-related_macular_degeneration^Age-related_macular_degeneration_9^Atypical_hemolytic_uremic_syndrome^Atypical_hemolytic-uremic_syndrome_5^Complement_component_3_deficiency_autosomal_recessive^Hemolytic-uremic_syndrome': 3,\n",
       " 'Currarino_triad': 4,\n",
       " 'Autoimmune_disease_6': 3,\n",
       " 'Familial_cancer_of_breast^Intellectual_developmental_disorder_with_gastrointestinal_difficulties_and_high_pain_threshold^Neoplasm_of_the_breast': 2,\n",
       " 'Aspergillosis_susceptibility_to': 1,\n",
       " 'Erythrokeratodermia_variabilis_et_progressiva_2': 2,\n",
       " 'Baller-Gerold_syndrome^Rapadilino_syndrome^Rothmund-Thomson_syndrome': 4,\n",
       " \"Juvenile_myopathy_encephalopathy_lactic_acidosis_AND_stroke^Leber's_optic_atrophy^Mitochondrial_DNA-Associated_Leigh_Syndrome_and_NARP^Retinitis_pigmentosa\": 8,\n",
       " 'Enhanced_s-cone_syndrome^Retinitis_pigmentosa^Retinitis_pigmentosa_37': 1,\n",
       " 'Acid-labile_subunit_deficiency': 3,\n",
       " 'Hypoprebetalipoproteinemia_acanthocytosis_retinitis_pigmentosa_and_pallidal_degeneration^Neurodegeneration_with_brain_iron_accumulation^Pigmentary_pallidal_degeneration': 2,\n",
       " 'CFHR5_deficiency^CFHR5-Related_Dense_Deposit_Disease_/_Membranoproliferative_Glomerulonephritis_Type_II^Mesangiocapillary_glomerulonephritis_type_II': 2,\n",
       " 'Andersen_Tawil_syndrome^Atrial_fibrillation^Atrial_fibrillation_familial_9^Short_QT_syndrome_3^short_QT_syndrome': 1,\n",
       " 'Autoinflammation_with_infantile_enterocolitis^Familial_cold_autoinflammatory_syndrome_4': 3,\n",
       " 'EPILEPSY_FAMILIAL_FOCAL_WITH_VARIABLE_FOCI_4^Early_infantile_epileptic_encephalopathy_62': 3,\n",
       " 'Anophthalmia-microphthalmia_syndrome^Holoprosencephaly_2^Holoprosencephaly_sequence^SCHIZENCEPHALY': 2,\n",
       " 'Microcephaly_and_chorioretinopathy_autosomal_recessive_2': 1,\n",
       " 'Amyotrophic_lateral_sclerosis_19': 3,\n",
       " 'Recombination_rate_quantitative_trait_locus_1': 2,\n",
       " 'Disorder_of_amino_acid_metabolism^Disorder_of_organic_acid_metabolism^Maple_syrup_urine_disease': 1,\n",
       " 'Retinitis_pigmentosa^Retinitis_pigmentosa_48': 1,\n",
       " 'Proliferative_vasculopathy_and_hydranencephaly-hydrocephaly_syndrome': 2,\n",
       " 'Arterial_tortuosity_syndrome': 1,\n",
       " 'Renal_hypouricemia_2': 2,\n",
       " 'Congenital_disorder_of_glycosylation^Congenital_disorder_of_glycosylation_type_1N:.': 1,\n",
       " 'ANEMIA_CONGENITAL_DYSERYTHROPOIETIC_TYPE_Ib': 1,\n",
       " 'Ehlers-Danlos-like_syndrome_due_to_tenascin-X_deficiency^Vesicoureteral_reflux_8': 10,\n",
       " 'Disorder_of_organic_acid_metabolism^Isovaleryl-CoA_dehydrogenase_deficiency': 3,\n",
       " 'Autoimmune_disease_multisystem_infantile-onset_1^Hyperimmunoglobulin_E_syndrome': 32,\n",
       " 'Short-sleeper': 2,\n",
       " 'Age-related_macular_degeneration^Age-related_macular_degeneration_4^Atypical_hemolytic_uremic_syndrome^Atypical_hemolytic-uremic_syndrome_1^Basal_laminar_drusen^Factor_H_deficiency^Hemolytic-uremic_syndrome': 2,\n",
       " 'Myopia_high_with_cataract_and_vitreoretinal_degeneration': 1,\n",
       " 'Anophthalmia-microphthalmia_syndrome^Microphthalmia_isolated_2^Microphthalmia_isolated_with_coloboma_3': 1,\n",
       " 'Megaloblastic_anemia_due_to_inborn_errors_of_metabolism': 10,\n",
       " 'Meretoja_syndrome:.:.': 5,\n",
       " 'Primary_hypomagnesemia': 2,\n",
       " 'Episodic_pain_syndrome_familial_2': 3,\n",
       " 'Epileptic_encephalopathy_early_infantile_38': 1,\n",
       " 'Bleeding_disorder_platelet-type_21': 3,\n",
       " 'Spastic_paraplegia_51_autosomal_recessive^Stuttering_familial_persistent_1': 1,\n",
       " 'Familial_cirrhosis': 6,\n",
       " 'EPILEPTIC_ENCEPHALOPATHY_EARLY_INFANTILE_57': 1,\n",
       " 'AL-RAQAD_SYNDROME': 1,\n",
       " 'Body_mass_index_quantitative_trait_locus_9': 3,\n",
       " 'Pituitary_hormone_deficiency_combined^Pituitary_hormone_deficiency_combined_2': 3,\n",
       " 'Achondrogenesis_type_II^Avascular_necrosis_of_the_head_of_femur^Coxa_plana^Czech_dysplasia_metatarsal_type^Epiphyseal_dysplasia_multiple_with_myopia_and_conductive_deafness^Familial_avascular_necrosis_of_the_femoral_head^Hereditary_hearing_loss_and_deafness^Kniest_dysplasia^Osteoarthritis_with_mild_chondrodysplasia^Platyspondylic_lethal_skeletal_dysplasia_Torrance_type^Spondyloepiphyseal_dysplasia^Spondyloepiphyseal_dysplasia_stanescu_type^Spondylometaphyseal_dysplasia^Spondyloperipheral_dysplasia^Stickler_syndrome^Stickler_syndrome_type_1^Stickler_syndrome_type_I_nonsyndromic_ocular^Type_II_Collagenopathies': 4,\n",
       " 'Immunoglobulin_kappa_light_chain_deficiency': 1,\n",
       " 'Primary_autosomal_recessive_microcephaly^Primary_autosomal_recessive_microcephaly_6^Seckel_syndrome_4': 1,\n",
       " 'Cardiac_valvular_dysplasia_X-linked^FG_syndrome_2^Frontometaphyseal_dysplasia^Intestinal_pseudoobstruction_neuronal_chronic_idiopathic_X-linked^Melnick-Needles_syndrome^Oto-palato-digital_syndrome_type_I^Oto-palato-digital_syndrome_type_II^Otopalatodigital_Spectrum_Disorders^Periventricular_nodular_heterotopia^Periventricular_nodular_heterotopia_1^Terminal_osseous_dysplasia': 21,\n",
       " 'Ashkenazi_Jewish_disorders^Cystic_fibrosis^Diaphyseal_dysplasia^INFLAMMATORY_BOWEL_DISEASE_IMMUNODEFICIENCY_AND_ENCEPHALOPATHY': 2,\n",
       " 'Alveolar_rhabdomyosarcoma': 3,\n",
       " 'Congenital_muscular_dystrophy^Muscular_dystrophy-dystroglycanopathy_(congenital_with_brain_and_eye_anomalies)_type_a_14^Muscular_dystrophy-dystroglycanopathy_(congenital_with_mental_retardation)_type_b_14^Muscular_dystrophy-dystroglycanopathy_(limb-girdle)_type_c_14^alpha-Dystroglycan_related_dystrophy': 4,\n",
       " 'Left_ventricular_noncompaction_8': 2,\n",
       " 'SPERMATOGENIC_FAILURE_22': 2,\n",
       " 'Otofaciocervical_syndrome_2': 2,\n",
       " 'Peeling_skin_syndrome_5': 1,\n",
       " 'Diabetes_mellitus_type_2^Hepatic_lipase_deficiency^High_density_lipoprotein_cholesterol_level_quantitative_trait_locus_12': 3,\n",
       " 'Combined_immunodeficiency_X-linked^Severe_combined_immunodeficiency_disease^X-linked_severe_combined_immunodeficiency': 1,\n",
       " 'Basal_ganglia_calcification_idiopathic_4^Idiopathic_basal_ganglia_calcification_1^Infantile_myofibromatosis^Infantile_myofibromatosis_1^Kosaki_overgrowth_syndrome^Malignant_eosinophil_proliferation^Premature_aging_syndrome_Penttinen_type': 2,\n",
       " 'Vitreoretinopathy_neovascular_inflammatory': 2,\n",
       " 'Deafness_autosomal_recessive_84b': 8,\n",
       " 'Pontocerebellar_hypoplasia_type_1^Pontocerebellar_hypoplasia_type_1c^Pontoneocerebellar_hypoplasia': 1,\n",
       " 'Hypogonadotropic_hypogonadism_22_with_or_without_anosmia^Hypogonadotropic_hypogonadism_7_with_or_without_anosmia': 1,\n",
       " 'Beta-D-mannosidosis': 3,\n",
       " 'Ehlers-Danlos_syndrome_with_progressive_kyphoscoliosis_myopathy_and_hearing_loss': 1,\n",
       " 'Charcot-Marie-Tooth_disease_axonal_type_2R': 1,\n",
       " 'Congenital_Fibrosis_of_the_Extraocular_Muscles_1B^Congenital_fibrosis_of_the_extraocular_muscles^Cortical_dysplasia_complex_with_other_brain_malformations_1^Fibrosis_of_extraocular_muscles_congenital_3a_with_or_without_extraocular_involvement': 1,\n",
       " 'Angiopathy_hereditary_with_nephropathy_aneurysms_and_muscle_cramps^Brain_small_vessel_disease_with_hemorrhage^Hemorrhage_intracerebral_susceptibility_to^MICROANGIOPATHY_AND_LEUKOENCEPHALOPATHY_PONTINE_AUTOSOMAL_DOMINANT^Porencephaly_1^Retinal_arteries_tortuosity_of': 1,\n",
       " 'Elliptocytosis_3^Hereditary_elliptocytosis^Hereditary_spherocytosis^Spherocytosis_type_2': 6,\n",
       " 'Acute_myeloid_leukemia^Sotos_syndrome_1': 6,\n",
       " 'Hypochromic_microcytic_anemia_with_iron_overload': 1,\n",
       " 'Uric_acid_nephrolithiasis_susceptibility_to': 1,\n",
       " '3-methylglutaconic_aciduria_type_V': 5,\n",
       " 'Lissencephaly_6_with_microcephaly': 1,\n",
       " 'Muscular_dystrophy-dystroglycanopathy_(congenital_with_brain_and_eye_anomalies)_type_a_12^Muscular_dystrophy-dystroglycanopathy_(limb-girdle)_type_c_12': 2,\n",
       " 'Ciliary_dyskinesia_primary_2^Primary_ciliary_dyskinesia': 2,\n",
       " 'Charcot-Marie-Tooth_disease^Charcot-Marie-Tooth_disease_type_4^Charcot-Marie-Tooth_disease_type_4B2': 5,\n",
       " 'Periodic_fever_menstrual_cycle-dependent': 2,\n",
       " 'Ectopia_lentis^Ectopia_lentis_et_pupillae^Ectopia_lentis_isolated_autosomal_recessive': 4,\n",
       " 'Spinocerebellar_ataxia_autosomal_recessive_10': 1,\n",
       " 'Cardiomyopathy^Caveolinopathy^Distal_myopathy_Tateyama_type^Elevated_serum_creatine_phosphokinase^Familial_hypertrophic_cardiomyopathy_1^Limb-Girdle_Muscular_Dystrophies_Autosomal_Dominant^Limb-girdle_muscular_dystrophy^Long_QT_syndrome_9^Primary_familial_hypertrophic_cardiomyopathy^Rippling_muscle_disease^Romano-Ward_syndrome': 1,\n",
       " 'Diarrhea_8_secretory_sodium_congenital': 2,\n",
       " 'GELEOPHYSIC_DYSPLASIA_3^Verloes_Bourguignon_syndrome': 1,\n",
       " 'Shprintzen_syndrome': 2,\n",
       " 'MICROCEPHALY_22_PRIMARY_AUTOSOMAL_RECESSIVE': 3,\n",
       " 'Mucopolysaccharidosis-plus_syndrome:.': 2,\n",
       " 'Anemia_nonspherocytic_hemolytic_due_to_G6PD_deficiency^Chlorpropamide_response^Glimepiride_response^Glipizide_response^Glyburide_response^Pegloticase_response^Rasburicase_response^Susceptibility_to_malaria^Tolbutamide_response': 10,\n",
       " 'Alopecia_neurologic_defects_and_endocrinopathy_syndrome': 2,\n",
       " 'Epilepsy_juvenile_absence^Epilepsy_childhood_absence_5^Epileptic_encephalopathy_early_infantile_43^Idiopathic_generalized_epilepsy': 2,\n",
       " 'Focal_cortical_dysplasia_type_II^Lymphangiomyomatosis^Tuberous_sclerosis_1^Tuberous_sclerosis_syndrome': 3,\n",
       " 'Diffuse_palmoplantar_keratoderma_Bothnian_type': 1,\n",
       " 'Fasting_plasma_glucose_level_quantitative_trait_locus_5': 2,\n",
       " 'Achondrogenesis_type_IA^Goldblatt_hypertension': 1,\n",
       " 'Craniolenticulosutural_dysplasia': 1,\n",
       " 'Deafness_autosomal_recessive_88': 1,\n",
       " '24-Dienoyl-CoA_reductase_deficiency': 5,\n",
       " 'Dyskeratosis_congenita_autosomal_recessive_6^Pulmonary_fibrosis_and/or_bone_marrow_failure_telomere-related_4': 78,\n",
       " 'Joubert_syndrome^Joubert_syndrome_2^Meckel_syndrome_type_2^Meckel-Gruber_syndrome': 1,\n",
       " 'Dysostosis_multiplex^Mucopolysaccharidosis_type_I^Mucopolysaccharidosis_MPS-I-H/S^Mucopolysaccharidosis_MPS-I-S': 4,\n",
       " 'COMBINED_OXIDATIVE_PHOSPHORYLATION_DEFICIENCY_39': 2,\n",
       " 'Stuve-Wiedemann_syndrome': 2,\n",
       " 'MENTAL_RETARDATION_AUTOSOMAL_DOMINANT_52': 4,\n",
       " 'Complement_component_c1s_deficiency^Ehlers-Danlos_syndrome_periodontal_type_2': 3,\n",
       " 'Epileptic_encephalopathy_early_infantile_50': 6,\n",
       " 'Essential_thrombocythemia^Myelofibrosis': 13,\n",
       " 'Autosomal_dominant_non-syndromic_sensorineural_deafness_type_DFNA^Deafness_autosomal_dominant_41^Hereditary_hearing_loss_and_deafness^Nonsyndromic_hearing_loss_and_deafness': 2,\n",
       " 'Deafness_nonsyndromic_sensorineural_mitochondrial^Hereditary_hearing_loss_and_deafness^Juvenile_myopathy_encephalopathy_lactic_acidosis_AND_stroke^Mitochondrial_diseases^Nonsyndromic_hearing_loss_and_deafness': 3,\n",
       " 'Cataracts_growth_hormone_deficiency_sensory_neuropathy_sensorineural_hearing_loss_and_skeletal_dysplasia': 2,\n",
       " 'Chromosome_9q_deletion_syndrome': 8,\n",
       " 'Leber_congenital_amaurosis^Leber_congenital_amaurosis_5': 1,\n",
       " 'Afibrinogenemia_congenital^Dysfibrinogenemia_congenital^Hereditary_factor_I_deficiency_disease': 2,\n",
       " 'Inflammatory_bowel_disease_14^Systemic_lupus_erythematosus^Systemic_lupus_erythematosus_10': 2,\n",
       " 'Mental_retardation_autosomal_recessive_50': 1,\n",
       " 'Kallikrein_decreased_urinary_activity_of': 1,\n",
       " 'Combined_molybdoflavoprotein_enzyme_deficiency^Hyperekplexia^Hyperekplexia_hereditary^Molybdenum_cofactor_deficiency_complementation_group_C': 1,\n",
       " 'Wiedemann-Steiner_syndrome': 9,\n",
       " 'NEURODEVELOPMENTAL_DISORDER_WITH_MICROCEPHALY_AND_STRUCTURAL_BRAIN_ANOMALIES': 1,\n",
       " 'Mental_retardation_autosomal_dominant_23': 1,\n",
       " 'Neutrophil_immunodeficiency_syndrome': 2,\n",
       " 'Alcohol_dependence^Anorexia_nervosa_1^Major_depressive_disorder^Obsessive-compulsive_disorder^Schizophrenia': 1,\n",
       " 'Hypomagnesemia_1_intestinal': 5,\n",
       " '.:Human_immunodeficiency_virus_type_1_susceptibility_to': 16,\n",
       " 'Lafora_disease^Progressive_myoclonus_epilepsy': 5,\n",
       " 'NEURODEVELOPMENTAL_DISORDER_WITH_SEVERE_MOTOR_IMPAIRMENT_AND_ABSENT_LANGUAGE': 2,\n",
       " 'Familial_hemophagocytic_lymphohistiocytosis^Hemophagocytic_lymphohistiocytosis_familial_5': 6,\n",
       " 'Familial_Mediterranean_fever^Familial_mediterranean_fever_autosomal_dominant': 4,\n",
       " 'Hydatidiform_mole': 3,\n",
       " 'Dihydropteridine_reductase_deficiency^Disorder_of_amino_acid_metabolism^Phenylketonuria': 3,\n",
       " \"Hereditary_hearing_loss_and_deafness^Leber's_optic_atrophy^Mitochondrial_DNA-Associated_Leigh_Syndrome_and_NARP^Mitochondrial_diseases^Retinitis_pigmentosa\": 3,\n",
       " 'Dronabinol_response^Flurbiprofen_response^Lesinurad_response^Piroxicam_response^Warfarin_response': 1,\n",
       " 'Congenital_myasthenic_syndrome^Familial_infantile_myasthenia': 5,\n",
       " 'Sponastrime_dysplasia': 2,\n",
       " 'Familial_cancer_of_breast^Fanconi_anemia^Fanconi_anemia_complementation_group_J^Neoplasm_of_the_breast': 4,\n",
       " 'Deafness_autosomal_recessive_36_with_or_without_vestibular_involvement^Hereditary_hearing_loss_and_deafness^Nonsyndromic_Hearing_Loss_and_Deafness_Autosomal_Recessive^Nonsyndromic_hearing_loss_and_deafness': 6,\n",
       " 'Epithelial-myoepithelial_carcinoma': 6,\n",
       " 'Diamond-Blackfan_anemia^Diamond-Blackfan_anemia_18': 4,\n",
       " 'Howel-Evans_syndrome': 6,\n",
       " 'MITOCHONDRIAL_COMPLEX_I_DEFICIENCY_NUCLEAR_TYPE_11': 1,\n",
       " 'Progressive_external_ophthalmoplegia_with_mitochondrial_DNA_deletions_autosomal_recessive_2': 1,\n",
       " 'Human_immunodeficiency_virus_type_1_susceptibility_to^Mycobacterium_tuberculosis_susceptibility_to^Neural_tube_defect': 1,\n",
       " 'Leukodystrophy_and_acquired_microcephaly_with_or_without_dystonia': 5,\n",
       " 'Mental_retardation_autosomal_dominant_29^Schinzel-Giedion_syndrome': 5,\n",
       " 'Combined_oxidative_phosphorylation_deficiency^Combined_oxidative_phosphorylation_deficiency_33': 1,\n",
       " 'Spinocerebellar_ataxia_7': 7,\n",
       " 'Anomaly_of_sex_chromosome^Chromosome_Xq28_duplication_syndrome^Mental_retardation_X-linked_syndromic_32^X_chromosome_deletion/duplication': 1,\n",
       " 'Chondrocalcinosis_2^Craniometaphyseal_dysplasia_autosomal_dominant': 1,\n",
       " 'Parkinson_disease^Parkinson_disease_6^Parkinson_disease_6_autosomal_recessive_early-onset': 4,\n",
       " 'Ciliary_dyskinesia_primary_33^Primary_ciliary_dyskinesia': 1,\n",
       " 'Craniosynostosis_2^Craniosynostosis_syndrome^Enlarged_parietal_foramina^Parietal_foramina_1^Parietal_foramina_with_cleidocranial_dysplasia': 1,\n",
       " 'Slowed_nerve_conduction_velocity_autosomal_dominant:.': 1,\n",
       " 'Dyskeratosis_congenita^Dyskeratosis_congenita_autosomal_recessive_3': 4,\n",
       " 'Lactic_aciduria_due_to_d-lactic_acid': 1,\n",
       " 'Joubert_syndrome_21': 3,\n",
       " 'Charcot-Marie-Tooth_disease_recessive_intermediate_c^Distal_spinal_muscular_atrophy_autosomal_recessive_4': 4,\n",
       " 'Autism_spectrum_disorder^Autistic_disorder_of_childhood_onset^Prader-Willi_syndrome': 2,\n",
       " 'Severe_combined_immunodeficiency_autosomal_recessive_T_cell-negative_B_cell-positive_NK_cell-positive': 2,\n",
       " 'Cleidocranial_dysostosis^Metaphyseal_dysplasia_with_maxillary_hypoplasia_and_brachydactyly': 3,\n",
       " 'Bone_mineral_density_quantitative_trait_locus_17': 3,\n",
       " 'MASP2_deficiency': 4,\n",
       " 'Disorder_of_organic_acid_metabolism^Propionyl-CoA_carboxylase_deficiency': 6,\n",
       " 'Myopathy_congenital_compton-north': 4,\n",
       " 'Osteopetrosis_autosomal_recessive_1': 4,\n",
       " 'Mental_retardation_and_distinctive_facial_features_with_or_without_cardiac_defects^Transposition_of_the_great_arteries^Transposition_of_the_great_arteries_dextro-looped_1': 5,\n",
       " 'NEURODEVELOPMENTAL_DISORDER_WITH_OR_WITHOUT_VARIABLE_BRAIN_ABNORMALITIES': 3,\n",
       " 'Arrhythmogenic_right_ventricular_cardiomyopathy^Arrhythmogenic_right_ventricular_cardiomyopathy_type_5^Cardiomyopathy^Emery-Dreifuss_muscular_dystrophy_7_autosomal_dominant': 1,\n",
       " 'Factor_X_deficiency': 2,\n",
       " 'Wilson_disease': 7,\n",
       " 'Autoimmune_thyroid_disease_3': 3,\n",
       " 'Congenital_disorder_of_glycosylation^Congenital_disorder_of_glycosylation_type_2k': 1,\n",
       " 'Carnitine_palmitoyltransferase_I_deficiency^Disorder_of_fatty_acid_metabolism': 2,\n",
       " 'Mirror_movements_3': 1,\n",
       " 'Autosomal_recessive_cutis_laxa_type_1B^Cutis_laxa_autosomal_recessive': 2,\n",
       " 'Bardet-Biedl_syndrome^Bardet-Biedl_syndrome_2^Retinitis_pigmentosa_74': 5,\n",
       " 'Anophthalmia-microphthalmia_syndrome^Focal_segmental_glomerulosclerosis_7^Renal_coloboma_syndrome': 1,\n",
       " 'Primary_ciliary_dyskinesia^Primary_ciliary_dyskinesia_23': 2,\n",
       " 'JOUBERT_SYNDROME_31^Short-rib_thoracic_dysplasia_13_with_or_without_polydactyly': 2,\n",
       " 'Ochoa_syndrome': 2,\n",
       " 'Congenital_central_hypoventilation^Familial_medullary_thyroid_carcinoma^Hirschsprung_disease^Hirschsprung_disease_1^Multiple_endocrine_neoplasia_type_2^Multiple_endocrine_neoplasia_type_2a^Multiple_endocrine_neoplasia_type_2b^Pheochromocytoma': 2,\n",
       " 'Retinitis_pigmentosa^Retinitis_pigmentosa_33': 3,\n",
       " 'Pituitary_hormone_deficiency_combined^Pituitary_hormone_deficiency_combined_1': 1,\n",
       " 'NEPHROTIC_SYNDROME_TYPE_19': 2,\n",
       " 'Deficiency_of_galactokinase^Galactosemia': 1,\n",
       " 'Intellectual_developmental_disorder_with_persistence_of_fetal_hemoglobin': 1,\n",
       " 'Atrial_septal_defect^Atrial_septal_defect_6': 2,\n",
       " 'Epidermolysis_bullosa_junctionalis_with_pyloric_atresia': 2,\n",
       " 'SPONDYLOEPIMETAPHYSEAL_DYSPLASIA_WITH_JOINT_LAXITY_TYPE_3': 2,\n",
       " 'Marinesco-Sjogren_syndrome': 1,\n",
       " 'Dyschromatosis_universalis_hereditaria_3^Langereis_blood_group^Microphthalmia_isolated_with_coloboma_7^Pseudohyperkalemia_familial_2_due_to_red_cell_leak': 5,\n",
       " 'Bladder_cancer_somatic^Neoplasm_of_the_genitourinary_tract^Osteosarcoma^Retinoblastoma^Small_cell_lung_cancer': 3,\n",
       " 'Autosomal_recessive_congenital_ichthyosis_2^Autosomal_recessive_congenital_ichthyosis_3^Congenital_ichthyosis_of_skin': 5,\n",
       " 'Bone_mineral_density_quantitative_trait_locus_12': 2,\n",
       " 'Hereditary_pancreatitis': 51,\n",
       " 'Biotin-thiamine-responsive_basal_ganglia_disease': 2,\n",
       " 'Multiple_sclerosis_susceptibility_to_5^TNF_receptor-associated_periodic_fever_syndrome_(TRAPS)': 2,\n",
       " 'Ciliary_dyskinesia_primary_22^Primary_ciliary_dyskinesia': 1,\n",
       " 'BLOOD_GROUP--LUTHERAN_SYSTEM^LuLu_phenotype': 3,\n",
       " 'Nephronophthisis_15': 6,\n",
       " 'Nephrotic_syndrome_type_15': 3,\n",
       " 'Amelogenesis_imperfecta_-_hypoplastic_autosomal_dominant_-_local^Amelogenesis_imperfecta_type_IC': 3,\n",
       " 'Asthma-related_traits_susceptibility_to_5': 2,\n",
       " 'Combined_molybdoflavoprotein_enzyme_deficiency^Molybdenum_cofactor_deficiency_complementation_group_A': 4,\n",
       " 'Alcohol_dependence': 4,\n",
       " 'Cerebellar_ataxia_Cayman_type': 1,\n",
       " 'Lung_disease_immunodeficiency_and_chromosome_breakage_syndrome': 6,\n",
       " 'Atrial_fibrillation_familial_13^Brugada_syndrome^Brugada_syndrome_5^Epileptic_encephalopathy_early_infantile_52^Generalized_epilepsy_with_febrile_seizures_plus_type_1': 3,\n",
       " 'Bainbridge-Ropers_syndrome': 4,\n",
       " 'Megaloblastic_anemia_due_to_dihydrofolate_reductase_deficiency': 19,\n",
       " 'Myelodysplastic_syndrome': 6,\n",
       " 'COFFIN-SIRIS_SYNDROME_6': 5,\n",
       " 'Retinitis_pigmentosa^Retinitis_pigmentosa_44': 1,\n",
       " 'MENTAL_RETARDATION_AUTOSOMAL_DOMINANT_49': 18,\n",
       " 'Pyogenic_arthritis_pyoderma_gangrenosum_and_acne': 3,\n",
       " \"Leber's_optic_atrophy^Mitochondrial_DNA-Associated_Leigh_Syndrome_and_NARP^Retinitis_pigmentosa\": 10,\n",
       " 'Trichothiodystrophy_2_photosensitive^Xeroderma_pigmentosum^Xeroderma_pigmentosum_complementation_group_b': 1,\n",
       " 'Cone-rod_dystrophy_5': 4,\n",
       " 'Deficiency_of_alpha-mannosidase:.': 2,\n",
       " 'Uncombable_hair_syndrome': 6,\n",
       " 'Meretoja_syndrome': 2,\n",
       " 'Generalized_epilepsy_with_febrile_seizures_plus_type_7^Hereditary_sensory_and_autonomic_neuropathy_type_II^Hereditary_sensory_and_autonomic_neuropathy_type_IIA^Indifference_to_pain_congenital_autosomal_recessive^Paroxysmal_extreme_pain_disorder^Primary_erythromelalgia^Severe_myoclonic_epilepsy_in_infancy': 7,\n",
       " 'Alveolar_soft_part_sarcoma': 9,\n",
       " 'Microphthalmia_isolated_5^Nanophthalmos_2': 2,\n",
       " 'Hepatitis_b_virus_susceptibility_to^Inflammatory_bowel_disease_25_autosomal_recessive:.': 1,\n",
       " 'Pseudoneonatal_adrenoleukodystrophy': 53,\n",
       " 'Pachyonychia_congenita_1^Pachyonychia_congenita_syndrome^Palmoplantar_keratoderma_nonepidermolytic_focal': 4,\n",
       " 'Disorder_of_organic_acid_metabolism^Methylmalonic_acidemia^Methylmalonyl-CoA_epimerase_deficiency': 1,\n",
       " 'Spinocerebellar_ataxia_autosomal_recessive_18': 3,\n",
       " 'Hypotrichosis_simplex': 5,\n",
       " 'Charcot-Marie-Tooth_disease^Charcot-Marie-Tooth_disease_dominant_intermediate_F': 2,\n",
       " 'Juvenile_myopathy_encephalopathy_lactic_acidosis_AND_stroke': 2,\n",
       " 'METABOLIC_CRISES_RECURRENT_WITH_RHABDOMYOLYSIS_CARDIAC_ARRHYTHMIAS_AND_NEURODEGENERATION': 1,\n",
       " 'Epilepsy_familial_adult_myoclonic_2': 1,\n",
       " 'EPIDERMODYSPLASIA_VERRUCIFORMIS_SUSCEPTIBILITY_TO_3': 1,\n",
       " 'Encephalopathy_progressive_with_amyotrophy_and_optic_atrophy^Hypoparathyroidism_retardation_dysmorphism_syndrome^Kenny-Caffey_syndrome_type_1': 2,\n",
       " 'Farber_disease^Jankovic_Rivera_syndrome': 4,\n",
       " 'Jarcho-Levin_syndrome^Spondylocostal_dysostosis_5': 1,\n",
       " 'Lymphedema_hereditary_ID': 1,\n",
       " 'Megaloblastic_anemia_thiamine-responsive_with_diabetes_mellitus_and_sensorineural_deafness': 1,\n",
       " 'Lissencephaly_8': 2,\n",
       " 'Essential_tremor^Hereditary_essential_tremor_1^Schizophrenia': 1,\n",
       " 'Aicardi_Goutieres_syndrome^Aicardi_Goutieres_syndrome_4': 1,\n",
       " 'Mental_retardation_autosomal_dominant_21': 1,\n",
       " 'Agammaglobulinemia_5_autosomal_dominant': 4,\n",
       " 'Glaucoma^Glaucoma_1_open_angle_O': 2,\n",
       " 'Congenital_hypothyroidism^Thyroid_dyshormonogenesis_6': 8,\n",
       " 'Hyperglycinuria^Iminoglycinuria^Neutral_1_amino_acid_transport_defect': 3,\n",
       " 'Porokeratosis_7_multiple_types': 3,\n",
       " 'Mesangiocapillary_glomerulonephritis_type_II^Nephrotic_syndrome_type_7': 1,\n",
       " 'Acth-independent_macronodular_adrenal_hyperplasia_2': 4,\n",
       " 'Blood_group_Junior_system^Uric_acid_concentration_serum_quantitative_trait_locus_1': 2,\n",
       " 'SPERMATOGENIC_FAILURE_31': 12,\n",
       " 'Uncombable_hair_syndrome_3': 5,\n",
       " 'ARTHROGRYPOSIS_DISTAL_TYPE_2B2': 1,\n",
       " 'Cold-induced_sweating_syndrome^Cold-induced_sweating_syndrome_1': 2,\n",
       " 'Microcornea_myopic_chorioretinal_atrophy_and_telecanthus': 4,\n",
       " 'Hyperlysinemia^Saccharopinuria': 6,\n",
       " 'X-Linked_mental_retardation_90': 2,\n",
       " 'Autosomal_recessive_hypophosphatemic_bone_disease': 7,\n",
       " 'Dyskeratosis_congenita^Dyskeratosis_congenita_autosomal_recessive_1': 2,\n",
       " 'Luscan-lumish_syndrome': 6,\n",
       " 'Early_infantile_epileptic_encephalopathy_15^Mental_retardation_autosomal_recessive_12': 1,\n",
       " 'Common_variable_immunodeficiency_2^Common_variable_immunodeficiency_7^Systemic_lupus_erythematosus^Systemic_lupus_erythematosus_9': 3,\n",
       " 'Coffin-Siris_syndrome_1': 10,\n",
       " 'Spastic_paraplegia_75_autosomal_recessive': 1,\n",
       " 'Familial_hypercholesterolemias^Hypercholesterolemia_autosomal_dominant_type_B^Hypobetalipoproteinemia_familial_1': 22,\n",
       " 'ENCEPHALOPATHY_PROGRESSIVE_EARLY-ONSET_WITH_BRAIN_EDEMA_AND/OR_LEUKOENCEPHALOPATHY_2': 2,\n",
       " 'Amyotrophic_lateral_sclerosis^Amyotrophic_lateral_sclerosis_16_juvenile^Distal_spinal_muscular_atrophy_autosomal_recessive_2^Juvenile_amyotrophic_lateral_sclerosis': 1,\n",
       " 'Frontotemporal_dementia_and/or_amyotrophic_lateral_sclerosis_2^Myopathy_isolated_mitochondrial_autosomal_dominant^Spinal_muscular_atrophy_jokela_type': 2,\n",
       " 'Abortive_cerebellar_ataxia^Autosomal_dominant_optic_atrophy_plus_syndrome^Dominant_hereditary_optic_atrophy^Glaucoma_normal_tension_susceptibility_to^Mitochondrial_DNA_depletion_syndrome_14_(cardioencephalomyopathic_type)': 2,\n",
       " 'Trichohepatoenteric_syndrome_2': 2,\n",
       " 'MITOCHONDRIAL_COMPLEX_I_DEFICIENCY_NUCLEAR_TYPE_2': 1,\n",
       " 'Body_mass_index_quantitative_trait_locus_14^Growth_retardation_developmental_delay_coarse_facies_and_early_death': 4,\n",
       " 'Thrombophilia_histidine-rich_glycoprotein-related': 4,\n",
       " 'Mental_retardation_X-linked_1': 1,\n",
       " 'Joubert_syndrome^Joubert_syndrome_16': 1,\n",
       " 'Familial_advanced_sleep_phase_syndrome_1': 4,\n",
       " \"DOORS_syndrome^Deafness_autosomal_dominant_65^Deafness_autosomal_recessive_86^Early_infantile_epileptic_encephalopathy_16^Epilepsy_rolandic_with_paroxysmal_exercise-induced_dystonia_and_writer's_cramp^Myoclonic_epilepsy_familial_infantile\": 1,\n",
       " 'Hereditary_hearing_loss_and_deafness^Leigh_syndrome^Mitochondrial_diseases^Pyruvate_dehydrogenase_E1-beta_deficiency^Pyruvate_dehydrogenase_complex_deficiency': 1,\n",
       " 'Neuropathy_hereditary_sensory_and_autonomic_type_1': 1,\n",
       " 'Emery-Dreifuss_muscular_dystrophy_5_autosomal_dominant': 18,\n",
       " 'Limb-girdle_muscular_dystrophy^Limb-girdle_muscular_dystrophy_autosomal_recessive^Limb-girdle_muscular_dystrophy_type_2A^MUSCULAR_DYSTROPHY_LIMB-GIRDLE_AUTOSOMAL_DOMINANT_4': 1,\n",
       " 'Deafness_autosomal_recessive_97^Hepatocellular_carcinoma^Neoplasm_of_the_genitourinary_tract^Osteofibrous_dysplasia^Renal_cell_carcinoma_papillary_1': 1,\n",
       " 'Autosomal_recessive_congenital_ichthyosis_9^Congenital_ichthyosis_of_skin': 1,\n",
       " 'Charcot-Marie-Tooth_disease_dominant_intermediate_B^Lethal_congenital_contracture_syndrome_5^Myopathy_centronuclear_1^Myotubular_myopathy': 2,\n",
       " 'Fanconi_anemia^Fanconi_anemia_complementation_group_A': 12,\n",
       " 'Maturity-onset_diabetes_of_the_young_type_7': 3,\n",
       " 'SHORT_STATURE_AMELOGENESIS_IMPERFECTA_AND_SKELETAL_DYSPLASIA_WITH_SCOLIOSIS': 1,\n",
       " 'Muscular_dystrophy-dystroglycanopathy_(congenital_with_brain_and_eye_anomalies)_type_a_11': 2,\n",
       " 'Cowden_syndrome_1^Familial_cancer_of_breast^Glioma_susceptibility_2^Macrocephaly/autism_syndrome^Malignant_tumor_of_prostate^Meningioma_familial^Neoplasm_of_the_breast^Neoplasm_of_the_genitourinary_tract^PTEN_hamartoma_tumor_syndrome': 1,\n",
       " 'Pseudoexfoliation_glaucoma': 2,\n",
       " 'Lethal_congenital_contracture_syndrome_9': 15,\n",
       " 'Combined_oxidative_phosphorylation_deficiency^Combined_oxidative_phosphorylation_deficiency_16': 1,\n",
       " 'Von_Willebrand_disease_recessive_form^von_Willebrand_disease_type_1^von_Willebrand_disease_type_2^von_Willebrand_disorder': 10,\n",
       " 'Anophthalmia-microphthalmia_syndrome': 4,\n",
       " 'FRASER_SYNDROME_3': 8,\n",
       " 'Microcephaly_short_stature_and_polymicrogyria_with_or_without_seizures': 2,\n",
       " 'Lethal_arthrogryposis_with_anterior_horn_cell_disease^Lethal_congenital_contracture_syndrome_1': 1,\n",
       " 'Combined_oxidative_phosphorylation_deficiency^Combined_oxidative_phosphorylation_deficiency_1': 3,\n",
       " 'Filippi_syndrome': 2,\n",
       " 'Cataract_4': 2,\n",
       " 'D-2-hydroxyglutaric_aciduria_1': 5,\n",
       " 'Enlarged_vestibular_aqueduct^Hereditary_hearing_loss_and_deafness^Nonsyndromic_Hearing_Loss_and_Deafness_Autosomal_Recessive^Nonsyndromic_hearing_loss_and_deafness^Pendred_syndrome': 4,\n",
       " 'Haim-Munk_syndrome^Papillon-Lefevre_syndrome^Periodontitis_aggressive_1': 7,\n",
       " 'Craniosynostosis_3^Craniosynostosis_syndrome': 1,\n",
       " 'JOINT_LAXITY_SHORT_STATURE_AND_MYOPIA': 1,\n",
       " 'Charcot-Marie-Tooth_disease^Charcot-Marie-Tooth_disease_axonal_autosomal_recessive_type_2A2B^Charcot-Marie-Tooth_disease_type_2^Charcot-Marie-Tooth_disease_type_2A^Charcot-Marie-Tooth_disease_type_2A2A^Hereditary_motor_and_sensory_neuropathy_with_optic_atrophy': 1,\n",
       " 'Aortic_aneurysm_familial_thoracic_8^Thoracic_aortic_aneurysm_and_aortic_dissection': 1,\n",
       " 'Dentin_dysplasia_type_I': 3,\n",
       " 'Combined_oxidative_phosphorylation_deficiency^Combined_oxidative_phosphorylation_deficiency_23': 1,\n",
       " 'Alazami-Yuan_syndrome': 2,\n",
       " 'Acute_myeloid_leukemia^Budd-Chiari_syndrome^Familial_erythrocytosis^Familial_erythrocytosis_1^Myelofibrosis^Polycythemia_vera^Thrombocythemia_3': 1,\n",
       " 'Glycogen_storage_disease^Glycogen_storage_disease_XV^Polyglucosan_body_myopathy^Polyglucosan_body_myopathy_2': 1,\n",
       " 'Anterior_segment_dysgenesis_3^Axenfeld-Rieger_syndrome_type_3^Rieger_syndrome': 3,\n",
       " 'Hereditary_hearing_loss_and_deafness^Peroxisome_biogenesis_disorder_11A^Peroxisome_biogenesis_disorder_11B^Peroxisome_biogenesis_disorders^Peroxisome_biogenesis_disorders_Zellweger_syndrome_spectrum': 4,\n",
       " 'POLYDACTYLY_POSTAXIAL_TYPE_A10': 4,\n",
       " 'Tyrosinemia_type_2': 2,\n",
       " 'Diabetes_mellitus_type_2^Familial_hypoplastic_glomerulocystic_kidney^Renal_cell_carcinoma_nonpapillary': 2,\n",
       " 'Spinocerebellar_ataxia_autosomal_recessive_4': 5,\n",
       " 'Amyloidosis_primary_localized_cutaneous_2': 1,\n",
       " 'Cardiomyopathy^Congenital_muscular_dystrophy^Congenital_muscular_dystrophy-dystroglycanopathy_without_mental_retardation_type_B4^Dilated_cardiomyopathy_1X^Familial_dilated_cardiomyopathy^Fukuyama_congenital_muscular_dystrophy^Limb-girdle_muscular_dystrophy^Limb-girdle_muscular_dystrophy_autosomal_recessive^Limb-girdle_muscular_dystrophy-dystroglycanopathy_type_C4^Muscular_dystrophy-dystroglycanopathy_(congenital_with_brain_and_eye_anomalies)_type_A_1^Primary_dilated_cardiomyopathy^alpha-Dystroglycan_related_dystrophy': 1,\n",
       " 'Persistent_hyperplastic_primary_vitreous_autosomal_recessive': 2,\n",
       " 'Mitochondrial_complex_III_deficiency_nuclear_type_7': 2,\n",
       " 'Pancreatic_cancer_1': 5,\n",
       " 'HEMOLYTIC_ANEMIA_CONGENITAL_X-LINKED': 1,\n",
       " 'Diamond-Blackfan_anemia^Diamond-Blackfan_anemia_12': 4,\n",
       " 'PAGANINI-MIOZZO_SYNDROME': 2,\n",
       " 'Immunodeficiency_42': 2,\n",
       " 'NEURODEVELOPMENTAL_DISORDER_WITH_EPILEPSY_AND_HYPOPLASIA_OF_THE_CORPUS_CALLOSUM': 2,\n",
       " 'Hereditary_hearing_loss_and_deafness^Leigh_syndrome^Mitochondrial_diseases^Pyruvate_dehydrogenase_E3-binding_protein_deficiency^Pyruvate_dehydrogenase_complex_deficiency': 7,\n",
       " 'Microcephaly_with_spastic_quadriplegia': 2,\n",
       " 'CILIARY_DYSKINESIA_PRIMARY_40': 24,\n",
       " 'Giant_axonal_neuropathy_2_autosomal_dominant:.': 1,\n",
       " 'Epileptic_encephalopathy_early_infantile_31': 1,\n",
       " 'Macular_corneal_dystrophy_Type_I': 1,\n",
       " 'Xia-Gibbs_syndrome': 4,\n",
       " 'Ulnar-mammary_syndrome': 2,\n",
       " 'Renal_dysplasia_retinal_pigmentary_dystrophy_cerebellar_ataxia_and_skeletal_dysplasia^Retinitis_pigmentosa_80': 22,\n",
       " 'Lipid_proteinosis': 1,\n",
       " 'Retinitis_pigmentosa^Retinitis_pigmentosa_57': 3,\n",
       " 'Immunodeficiency_18': 2,\n",
       " 'Platelet-type_bleeding_disorder_18': 1,\n",
       " 'Benign_familial_neonatal_seizures^Benign_familial_neonatal_seizures_1^Early_infantile_epileptic_encephalopathy_7': 1,\n",
       " 'Baraitser-Winter_syndrome_1^Juvenile-onset_dystonia': 4,\n",
       " 'Hereditary_spherocytosis^Spherocytosis_type_1': 5,\n",
       " 'Deafness_autosomal_recessive_30^Hereditary_hearing_loss_and_deafness^Nonsyndromic_Hearing_Loss_and_Deafness_Autosomal_Recessive^Nonsyndromic_hearing_loss_and_deafness': 2,\n",
       " 'gamma-Glutamyltransferase_deficiency': 10,\n",
       " 'Bietti_crystalline_corneoretinal_dystrophy': 2,\n",
       " 'Hepatocellular_carcinoma': 4,\n",
       " 'Myasthenic_syndrome_congenital_8': 9,\n",
       " 'Antenatal_Bartter_Syndrome^Bartter_syndrome^Bartter_syndrome_type_1_antenatal': 3,\n",
       " 'Inclusion_body_myopathy_3': 18,\n",
       " 'Afibrinogenemia_congenital^Dysfibrinogenemia_congenital^Familial_visceral_amyloidosis_Ostertag_type^Hereditary_factor_I_deficiency_disease': 12,\n",
       " 'Primary_autosomal_recessive_microcephaly^Primary_autosomal_recessive_microcephaly_2': 5,\n",
       " 'Fleck_retina_familial_benign': 1,\n",
       " 'Peeling_skin_syndrome_3': 3,\n",
       " 'KNOPS_BLOOD_GROUP_SYSTEM^Susceptibility_to_malaria': 4,\n",
       " 'Siderius_X-linked_mental_retardation_syndrome': 35,\n",
       " 'Mitral_valve_prolapse_2': 7,\n",
       " 'POLYDACTYLY_POSTAXIAL_TYPE_A8^Polydactyly_preaxial_I': 1,\n",
       " 'Primary_hypertrophic_osteoarthropathy_autosomal_recessive_2': 2,\n",
       " 'Dizygotic_twins^Ovarian_dysgenesis_1^Ovarian_hyperstimulation_syndrome': 2,\n",
       " 'Carcinoma_of_colon': 10,\n",
       " 'Dysferlinopathy^Limb-girdle_muscular_dystrophy^Limb-girdle_muscular_dystrophy_autosomal_recessive^Limb-girdle_muscular_dystrophy_type_2B^Miyoshi_muscular_dystrophy_1^Myopathy_distal_with_anterior_tibial_onset': 7,\n",
       " 'Familial_renal_hypouricemia': 3,\n",
       " 'Cardiomyopathy_familial_hypertrophic_26^Myofibrillar_myopathy^Myofibrillar_myopathy_filamin_C-related^Myopathy_distal_4': 9,\n",
       " 'Rheumatoid_arthritis_systemic_juvenile': 2,\n",
       " 'Diabetes_mellitus_type_2^Familial_hyperinsulinism^Leucine-induced_hypoglycemia^Neonatal_diabetes_mellitus^Permanent_neonatal_diabetes_mellitus^Persistent_hyperinsulinemic_hypoglycemia_of_infancy^Transient_neonatal_diabetes_mellitus_2^Transitory_neonatal_diabetes_mellitus': 2,\n",
       " 'Gonadotropin-independent_familial_sexual_precocity^Leydig_cell_agenesis^Precocious_puberty': 1,\n",
       " 'Holoprosencephaly_5^Holoprosencephaly_sequence': 1,\n",
       " 'Desbuquois_dysplasia_2^Pseudoxanthoma_elasticum': 3,\n",
       " 'Advanced_sleep_phase_syndrome_familial_2': 5,\n",
       " \"O''''DONNELL-LURIA-RODAN_SYNDROME\": 4,\n",
       " 'Epileptic_encephalopathy_early_infantile_1^Lissencephaly^Lissencephaly_2_X-linked^Malformation_of_cortical_development^Mental_retardation_with_or_without_seizures_ARX-related_X-linked^Partington_X-linked_mental_retardation_syndrome^Proud_Levine_Carpenter_syndrome': 3,\n",
       " 'IMMUNODEFICIENCY_62': 9,\n",
       " 'Biotin-Responsive_Multiple_Carboxylase_Deficiencies^Holocarboxylase_synthetase_deficiency': 1,\n",
       " 'Acheiropodia^Mirror_image_polydactyly^Polydactyly_preaxial_II^Syndactyly_type_IV^Tibia_hypoplasia_or_aplasia_of_with_polydactyly': 2,\n",
       " 'Autosomal_dominant_non-syndromic_sensorineural_deafness_type_DFNA^DFNA_3_Nonsyndromic_Hearing_Loss_and_Deafness^Deafness_X-linked_2^Deafness_autosomal_dominant_3a^Deafness_autosomal_dominant_nonsyndromic_sensorineural_3^Deafness_autosomal_recessive_1A^Deafness_digenic_GJB2/GJB3^Hereditary_hearing_loss_and_deafness^Hystrix-like_ichthyosis_with_deafness^Keratitis-ichthyosis-deafness_syndrome_autosomal_dominant^Keratoderma_palmoplantar_deafness^Knuckle_pads_deafness_AND_leukonychia_syndrome^Mutilating_keratoderma^Nonsyndromic_Hearing_Loss_and_Deafness_Autosomal_Recessive^Nonsyndromic_hearing_loss_and_deafness': 2,\n",
       " 'PROTEASOME-ASSOCIATED_AUTOINFLAMMATORY_SYNDROME_3': 5,\n",
       " 'Hereditary_xanthinuria^Xanthinuria_type_2': 2,\n",
       " 'Epilepsy_familial_temporal_lobe_5^Febrile_seizures_familial_11': 3,\n",
       " 'DEAFNESS_AUTOSOMAL_RECESSIVE_108': 1,\n",
       " 'Classic_congenital_adrenal_hyperplasia_due_to_21-hydroxylase_deficiency^Congenital_adrenal_hyperplasia^Disorder_of_endocrine_system': 6,\n",
       " 'Glanzmann_thrombasthenia^Myocardial_infarction_1^Platelet-type_bleeding_disorder_16': 1,\n",
       " 'Platelet-type_bleeding_disorder_11': 3,\n",
       " 'Left_ventricular_noncompaction_1': 1,\n",
       " 'Myd88_deficiency^Waldenstrom_macroglobulinemia': 1,\n",
       " 'Myoglobinuria_acute_recurrent_autosomal_recessive': 1,\n",
       " 'Acute_myeloid_leukemia^Thrombocytopenia_5': 3,\n",
       " 'Multiple_sulfatase_deficiency': 2,\n",
       " 'Megalencephaly-polymicrogyria-polydactyly-hydrocephalus_syndrome_3': 1,\n",
       " 'Warburg_micro_syndrome^Warburg_micro_syndrome_4': 2,\n",
       " 'Acrokeratosis_verruciformis_of_Hopf^Keratosis_follicularis': 3,\n",
       " 'Leprosy_5': 3,\n",
       " 'Bardet-Biedl_syndrome^Bardet-Biedl_syndrome_1:.': 1,\n",
       " 'Griscelli_syndrome_type_3': 3,\n",
       " 'Nemaline_myopathy^Nemaline_myopathy_6': 3,\n",
       " 'Glioma^Glioma_susceptibility_1^Lung_cancer^Neoplasm_of_lung^Neoplasm_of_stomach^Pertuzumab_response^Trastuzumab_response': 7,\n",
       " 'MICROCEPHALY_20_PRIMARY_AUTOSOMAL_RECESSIVE^Meckel_syndrome_12': 4,\n",
       " 'Hermansky-Pudlak_syndrome^Hermansky-Pudlak_syndrome_3': 1,\n",
       " 'Beta-thalassemia_dominant_inclusion_body_type^Erythrocytosis_6_familial^Fetal_hemoglobin_quantitative_trait_locus_1^Hb_SS_disease^Heinz_body_anemia^Hemoglobinopathy^METHEMOGLOBINEMIA_BETA_TYPE^Susceptibility_to_malaria^Thalassemia^alpha_Thalassemia^beta_Thalassemia': 26,\n",
       " 'Bethlem_myopathy_2^Ullrich_congenital_muscular_dystrophy_2': 8,\n",
       " 'Fructose-biphosphatase_deficiency': 2,\n",
       " 'Immunodeficiency_31C^Immunodeficiency_31a^Mycobacterial_and_viral_infections_susceptibility_to_autosomal_recessive': 2,\n",
       " 'EPILEPSY_FAMILIAL_ADULT_MYOCLONIC_7': 5,\n",
       " 'Deafness_autosomal_recessive_18b': 7,\n",
       " 'Glycogen_storage_disease^Glycogen_storage_disease_type_II^Metabolic_myopathy': 15,\n",
       " 'Combined_cellular_and_humoral_immune_defects_with_granulomas^Histiocytic_medullary_reticulosis^Severe_immunodeficiency_autosomal_recessive_T-cell_negative_B-cell_negative_NK_cell-positive': 3,\n",
       " 'Pituitary_dependent_hypercortisolism': 11,\n",
       " 'Ciliary_dyskinesia_primary_14^Primary_ciliary_dyskinesia': 1,\n",
       " 'Choreoacanthocytosis': 6,\n",
       " 'Immunodeficiency_32a^Immunodeficiency_32b': 3,\n",
       " 'Heimler_syndrome_2^Hereditary_hearing_loss_and_deafness^Peroxisome_biogenesis_disorder_4B^Peroxisome_biogenesis_disorder_4a_(zellweger)^Peroxisome_biogenesis_disorders^Peroxisome_biogenesis_disorders_Zellweger_syndrome_spectrum': 1,\n",
       " 'Cerebrooculofacioskeletal_syndrome_3^Xeroderma_pigmentosum^Xeroderma_pigmentosum_group_G': 1,\n",
       " 'MITOCHONDRIAL_COMPLEX_I_DEFICIENCY_NUCLEAR_TYPE_13': 2,\n",
       " 'Hypermethioninemia_with_s-adenosylhomocysteine_hydrolase_deficiency': 13,\n",
       " 'MENTAL_RETARDATION_AUTOSOMAL_DOMINANT_45': 7,\n",
       " 'Verheij_syndrome': 10,\n",
       " 'Familial_cancer_of_breast^Fanconi_anemia^Fanconi_anemia_complementation_group_N^Hereditary_cancer-predisposing_syndrome^Neoplasm_of_the_breast^Pancreatic_cancer_3': 2,\n",
       " 'EPILEPTIC_ENCEPHALOPATHY_EARLY_INFANTILE_77': 4,\n",
       " 'Cone-rod_dystrophy_9': 4,\n",
       " 'Aortic_aneurysm_familial_thoracic_7^Thoracic_aortic_aneurysm_and_aortic_dissection': 3,\n",
       " 'Osteogenesis_imperfecta_type_xiv': 1,\n",
       " '.:Apolipoprotein_C2_deficiency^Familial_chylomicronemia_syndrome': 1,\n",
       " 'Congenital_muscular_dystrophy^Laminin_alpha_2-related_dystrophy^MUSCULAR_DYSTROPHY_LIMB-GIRDLE_AUTOSOMAL_RECESSIVE_23^Merosin_deficient_congenital_muscular_dystrophy': 4,\n",
       " 'Glutathione_peroxidase_deficiency': 4,\n",
       " 'Cardiomyopathy_dilated_1NN^LEOPARD_syndrome_2^Noonan_syndrome^Noonan_syndrome_5^Noonan_syndrome_with_multiple_lentigines^Rasopathy': 2,\n",
       " ...}"
      ]
     },
     "execution_count": 434,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mt.aggregate_rows(hl.agg.counter(hl.delimit(mt.info.ISEQ_CLINVAR_DISEASES)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 426,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(11, 151)"
      ]
     },
     "execution_count": 426,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mt.filter_rows(mt.info.ISEQ_CLINVAR_DISEASES.contains('Schizophrenia'), keep = True).count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 445,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table><thead style=\"font-weight: bold;\"><tr><td>locus</td><td>alleles</td><td>rsid</td></tr>\n",
       "<tr><td>locus&lt;GRCh38&gt;</td><td>array&lt;str&gt;</td><td>str</td></tr>\n",
       "</thead><tbody><tr><td>chr11:70473268</td><td>[&quot;C&quot;,&quot;T&quot;]</td><td>&quot;rs140134890&quot;</td></tr>\n",
       "<tr><td>chr11:70490374</td><td>[&quot;C&quot;,&quot;T&quot;]</td><td>&quot;rs117843717&quot;</td></tr>\n",
       "<tr><td>chr11:70698758</td><td>[&quot;G&quot;,&quot;A&quot;]</td><td>NA</td></tr>\n",
       "<tr><td>chr11:70820627</td><td>[&quot;C&quot;,&quot;T&quot;]</td><td>NA</td></tr>\n",
       "<tr><td>chr11:70820628</td><td>[&quot;G&quot;,&quot;A&quot;]</td><td>NA</td></tr>\n",
       "<tr><td>chr11:70952780</td><td>[&quot;A&quot;,&quot;AACTCGCC&quot;]</td><td>NA</td></tr>\n",
       "</tbody></table>"
      ],
      "text/plain": [
       "+----------------+------------------+---------------+\n",
       "| locus          | alleles          | rsid          |\n",
       "+----------------+------------------+---------------+\n",
       "| locus<GRCh38>  | array<str>       | str           |\n",
       "+----------------+------------------+---------------+\n",
       "| chr11:70473268 | [\"C\",\"T\"]        | \"rs140134890\" |\n",
       "| chr11:70490374 | [\"C\",\"T\"]        | \"rs117843717\" |\n",
       "| chr11:70698758 | [\"G\",\"A\"]        | NA            |\n",
       "| chr11:70820627 | [\"C\",\"T\"]        | NA            |\n",
       "| chr11:70820628 | [\"G\",\"A\"]        | NA            |\n",
       "| chr11:70952780 | [\"A\",\"AACTCGCC\"] | NA            |\n",
       "+----------------+------------------+---------------+"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mt.filter_rows(mt.info.ISEQ_CLINVAR_DISEASES.contains('Autism_17^Autism_spectrum_disorder'), keep = True).rsid.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 468,
   "metadata": {},
   "outputs": [],
   "source": [
    "a = 'Cowden_syndrome_1^Familial_cancer_of_breast^Glioma_susceptibility_2^Macrocephaly/autism_syndrome^Malignant_tumor_of_prostate^Meningioma_familial^Neoplasm_of_the_breast^Neoplasm_of_the_genitourinary_tract^PTEN_hamartoma_tumor_syndrome'\n",
    "\n",
    "voi = mt.filter_rows(mt.info.ISEQ_CLINVAR_DISEASES.contains(a), keep = True) # VariantOfInterest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 470,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table><thead style=\"font-weight: bold;\"><tr><td>locus</td><td>alleles</td><td>S_136.GT</td><td>S_170c.GT</td><td>S_170d.GT</td><td>S_6981.GT</td></tr>\n",
       "<tr><td>locus&lt;GRCh38&gt;</td><td>array&lt;str&gt;</td><td>call</td><td>call</td><td>call</td><td>call</td></tr>\n",
       "</thead><tbody><tr><td>chr10:87931071</td><td>[&quot;G&quot;,&quot;A&quot;]</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td></tr>\n",
       "</tbody></table><p style=\"background: #fdd; padding: 0.4em;\">showing the first 4 of 151 columns</p>\n"
      ],
      "text/plain": [
       "+----------------+------------+----------+-----------+-----------+-----------+\n",
       "| locus          | alleles    | S_136.GT | S_170c.GT | S_170d.GT | S_6981.GT |\n",
       "+----------------+------------+----------+-----------+-----------+-----------+\n",
       "| locus<GRCh38>  | array<str> | call     | call      | call      | call      |\n",
       "+----------------+------------+----------+-----------+-----------+-----------+\n",
       "| chr10:87931071 | [\"G\",\"A\"]  | 0/0      | 0/0       | 0/0       | 0/0       |\n",
       "+----------------+------------+----------+-----------+-----------+-----------+\n",
       "showing the first 4 of 151 columns"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "voi.GT.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 471,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: 150, 1: 1}"
      ]
     },
     "execution_count": 471,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "voi.aggregate_entries(hl.agg.counter(voi.GT.n_alt_alleles()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='5.1'></a> \n",
    "\n",
    "\n",
    "### 5.1 Are there known pathogenic variants in > 1 sample?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 494,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(114, 151)"
      ]
     },
     "execution_count": 494,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mt.filter_rows(mt.info.ISEQ_AGGREGATED_CLINVAR_SIGNIFICANCE.contains('pathogenic'), keep = True).count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 496,
   "metadata": {},
   "outputs": [],
   "source": [
    "pathogenic = mt.filter_rows(mt.info.ISEQ_AGGREGATED_CLINVAR_SIGNIFICANCE.contains('pathogenic'), keep = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 499,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'12': 1,\n",
       " '8': 1,\n",
       " '4': 6,\n",
       " '9': 1,\n",
       " '5': 4,\n",
       " '6': 2,\n",
       " '1': 61,\n",
       " '2': 21,\n",
       " '7': 3,\n",
       " '3': 14}"
      ]
     },
     "execution_count": 499,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathogenic.aggregate_rows(hl.agg.counter(hl.delimit(pathogenic.info.AC)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 503,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table><thead style=\"font-weight: bold;\"><tr><td>locus</td><td>alleles</td><td>rsid</td></tr>\n",
       "<tr><td>locus&lt;GRCh38&gt;</td><td>array&lt;str&gt;</td><td>str</td></tr>\n",
       "</thead><tbody><tr><td>chr7:142750600</td><td>[&quot;A&quot;,&quot;C&quot;]</td><td>NA</td></tr>\n",
       "</tbody></table>"
      ],
      "text/plain": [
       "+----------------+------------+------+\n",
       "| locus          | alleles    | rsid |\n",
       "+----------------+------------+------+\n",
       "| locus<GRCh38>  | array<str> | str  |\n",
       "+----------------+------------+------+\n",
       "| chr7:142750600 | [\"A\",\"C\"]  | NA   |\n",
       "+----------------+------------+------+"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pathogenic.filter_rows(pathogenic.info.AC == [12]).rsid.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://www.ncbi.nlm.nih.gov/clinvar/RCV000031923/#clinical-assertions this is the link to this variant "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='6'></a> \n",
    "\n",
    "### 6. Annotate with CADD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 486,
   "metadata": {},
   "outputs": [],
   "source": [
    "db = hl.experimental.DB()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 487,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<hail.experimental.db.DB at 0x7f938d3b4650>"
      ]
     },
     "execution_count": 487,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 488,
   "metadata": {},
   "outputs": [
    {
     "ename": "FatalError",
     "evalue": "IOException: No FileSystem for scheme: gs\n\nJava stack trace:\njava.io.IOException: No FileSystem for scheme: gs\n\tat org.apache.hadoop.fs.FileSystem.getFileSystemClass(FileSystem.java:2660)\n\tat org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:2667)\n\tat org.apache.hadoop.fs.FileSystem.access$200(FileSystem.java:94)\n\tat org.apache.hadoop.fs.FileSystem$Cache.getInternal(FileSystem.java:2703)\n\tat org.apache.hadoop.fs.FileSystem$Cache.get(FileSystem.java:2685)\n\tat org.apache.hadoop.fs.FileSystem.get(FileSystem.java:373)\n\tat org.apache.hadoop.fs.Path.getFileSystem(Path.java:295)\n\tat is.hail.io.fs.HadoopFS.is$hail$io$fs$HadoopFS$$_fileSystem(HadoopFS.scala:156)\n\tat is.hail.io.fs.HadoopFS.isDir(HadoopFS.scala:173)\n\tat is.hail.variant.RelationalSpec$.readMetadata(MatrixTable.scala:39)\n\tat is.hail.variant.RelationalSpec$.readReferences(MatrixTable.scala:71)\n\tat is.hail.variant.ReferenceGenome$.fromHailDataset(ReferenceGenome.scala:586)\n\tat is.hail.variant.ReferenceGenome.fromHailDataset(ReferenceGenome.scala)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\n\tat sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.lang.reflect.Method.invoke(Method.java:498)\n\tat py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)\n\tat py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)\n\tat py4j.Gateway.invoke(Gateway.java:282)\n\tat py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)\n\tat py4j.commands.CallCommand.execute(CallCommand.java:79)\n\tat py4j.GatewayConnection.run(GatewayConnection.java:238)\n\tat java.lang.Thread.run(Thread.java:745)\n\n\n\nHail version: 0.2.26-2dcc3d963867\nError summary: IOException: No FileSystem for scheme: gs",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFatalError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-488-2739b8c14f82>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mmt\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mannotate_rows_db\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmt\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'CADD'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m</usr/local/lib/python3.7/site-packages/decorator.py:decorator-gen-1605>\u001b[0m in \u001b[0;36mannotate_rows_db\u001b[0;34m(self, rel, *names)\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.7/site-packages/hail/typecheck/check.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(__original_func, *args, **kwargs)\u001b[0m\n\u001b[1;32m    583\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m__original_func\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    584\u001b[0m         \u001b[0margs_\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwargs_\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcheck_all\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m__original_func\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwargs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcheckers\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mis_method\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mis_method\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 585\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0m__original_func\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs_\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs_\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    586\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    587\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.7/site-packages/hail/experimental/db.py\u001b[0m in \u001b[0;36mannotate_rows_db\u001b[0;34m(self, rel, *names)\u001b[0m\n\u001b[1;32m    204\u001b[0m                 \u001b[0mrel\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mannotate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0mdataset\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mgenes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdataset\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    205\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 206\u001b[0;31m                 \u001b[0mindexed_value\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdataset\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex_compatible_version\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    207\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexed_value\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhl\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtstruct\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexed_value\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    208\u001b[0m                     \u001b[0mindexed_value\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mhl\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_defined\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexed_value\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.7/site-packages/hail/experimental/db.py\u001b[0m in \u001b[0;36mindex_compatible_version\u001b[0;34m(self, key_expr)\u001b[0m\n\u001b[1;32m     60\u001b[0m             \u001b[0mindex\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     61\u001b[0m             for index in (version.maybe_index(key_expr, all_matches)\n\u001b[0;32m---> 62\u001b[0;31m                           for version in self.versions)\n\u001b[0m\u001b[1;32m     63\u001b[0m             if index is not None]\n\u001b[1;32m     64\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcompatible_indexed_values\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.7/site-packages/hail/experimental/db.py\u001b[0m in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m     58\u001b[0m         \u001b[0mall_matches\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'unique'\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkey_properties\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     59\u001b[0m         compatible_indexed_values = [\n\u001b[0;32m---> 60\u001b[0;31m             \u001b[0mindex\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     61\u001b[0m             for index in (version.maybe_index(key_expr, all_matches)\n\u001b[1;32m     62\u001b[0m                           for version in self.versions)\n",
      "\u001b[0;32m/usr/local/lib/python3.7/site-packages/hail/experimental/db.py\u001b[0m in \u001b[0;36m<genexpr>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m     60\u001b[0m             \u001b[0mindex\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     61\u001b[0m             for index in (version.maybe_index(key_expr, all_matches)\n\u001b[0;32m---> 62\u001b[0;31m                           for version in self.versions)\n\u001b[0m\u001b[1;32m     63\u001b[0m             if index is not None]\n\u001b[1;32m     64\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcompatible_indexed_values\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.7/site-packages/hail/experimental/db.py\u001b[0m in \u001b[0;36mmaybe_index\u001b[0;34m(self, indexer_key_expr, all_matches)\u001b[0m\n\u001b[1;32m     25\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     26\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mmaybe_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer_key_expr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mall_matches\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 27\u001b[0;31m         return hl.read_table(self.url)._maybe_flexindex_table_by_expr(\n\u001b[0m\u001b[1;32m     28\u001b[0m             indexer_key_expr, all_matches=all_matches)\n\u001b[1;32m     29\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m</usr/local/lib/python3.7/site-packages/decorator.py:decorator-gen-1357>\u001b[0m in \u001b[0;36mread_table\u001b[0;34m(path, _intervals, _filter_intervals)\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.7/site-packages/hail/typecheck/check.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(__original_func, *args, **kwargs)\u001b[0m\n\u001b[1;32m    583\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m__original_func\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    584\u001b[0m         \u001b[0margs_\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwargs_\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcheck_all\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m__original_func\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwargs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcheckers\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mis_method\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mis_method\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 585\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0m__original_func\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs_\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs_\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    586\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    587\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.7/site-packages/hail/methods/impex.py\u001b[0m in \u001b[0;36mread_table\u001b[0;34m(path, _intervals, _filter_intervals)\u001b[0m\n\u001b[1;32m   2249\u001b[0m     \u001b[0;34m:\u001b[0m\u001b[0;32mclass\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;31m`\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTable\u001b[0m\u001b[0;31m`\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2250\u001b[0m     \"\"\"\n\u001b[0;32m-> 2251\u001b[0;31m     \u001b[0;32mfor\u001b[0m \u001b[0mrg_config\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mEnv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbackend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mload_references_from_dataset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2252\u001b[0m         \u001b[0mhl\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mReferenceGenome\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_from_config\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrg_config\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2253\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.7/site-packages/hail/backend/backend.py\u001b[0m in \u001b[0;36mload_references_from_dataset\u001b[0;34m(self, path)\u001b[0m\n\u001b[1;32m    160\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    161\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mload_references_from_dataset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpath\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 162\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mjson\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloads\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mEnv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhail\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvariant\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mReferenceGenome\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfromHailDataset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    163\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    164\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mfrom_fasta_file\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfasta_file\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex_file\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mx_contigs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_contigs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmt_contigs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpar\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.7/site-packages/py4j/java_gateway.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args)\u001b[0m\n\u001b[1;32m   1255\u001b[0m         \u001b[0manswer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgateway_client\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend_command\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcommand\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1256\u001b[0m         return_value = get_return_value(\n\u001b[0;32m-> 1257\u001b[0;31m             answer, self.gateway_client, self.target_id, self.name)\n\u001b[0m\u001b[1;32m   1258\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1259\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mtemp_arg\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mtemp_args\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.7/site-packages/hail/utils/java.py\u001b[0m in \u001b[0;36mdeco\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    223\u001b[0m             raise FatalError('%s\\n\\nJava stack trace:\\n%s\\n'\n\u001b[1;32m    224\u001b[0m                              \u001b[0;34m'Hail version: %s\\n'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 225\u001b[0;31m                              'Error summary: %s' % (deepest, full, hail.__version__, deepest)) from None\n\u001b[0m\u001b[1;32m    226\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mpyspark\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msql\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mutils\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mCapturedException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    227\u001b[0m             raise FatalError('%s\\n\\nJava stack trace:\\n%s\\n'\n",
      "\u001b[0;31mFatalError\u001b[0m: IOException: No FileSystem for scheme: gs\n\nJava stack trace:\njava.io.IOException: No FileSystem for scheme: gs\n\tat org.apache.hadoop.fs.FileSystem.getFileSystemClass(FileSystem.java:2660)\n\tat org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:2667)\n\tat org.apache.hadoop.fs.FileSystem.access$200(FileSystem.java:94)\n\tat org.apache.hadoop.fs.FileSystem$Cache.getInternal(FileSystem.java:2703)\n\tat org.apache.hadoop.fs.FileSystem$Cache.get(FileSystem.java:2685)\n\tat org.apache.hadoop.fs.FileSystem.get(FileSystem.java:373)\n\tat org.apache.hadoop.fs.Path.getFileSystem(Path.java:295)\n\tat is.hail.io.fs.HadoopFS.is$hail$io$fs$HadoopFS$$_fileSystem(HadoopFS.scala:156)\n\tat is.hail.io.fs.HadoopFS.isDir(HadoopFS.scala:173)\n\tat is.hail.variant.RelationalSpec$.readMetadata(MatrixTable.scala:39)\n\tat is.hail.variant.RelationalSpec$.readReferences(MatrixTable.scala:71)\n\tat is.hail.variant.ReferenceGenome$.fromHailDataset(ReferenceGenome.scala:586)\n\tat is.hail.variant.ReferenceGenome.fromHailDataset(ReferenceGenome.scala)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\n\tat sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.lang.reflect.Method.invoke(Method.java:498)\n\tat py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)\n\tat py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)\n\tat py4j.Gateway.invoke(Gateway.java:282)\n\tat py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)\n\tat py4j.commands.CallCommand.execute(CallCommand.java:79)\n\tat py4j.GatewayConnection.run(GatewayConnection.java:238)\n\tat java.lang.Thread.run(Thread.java:745)\n\n\n\nHail version: 0.2.26-2dcc3d963867\nError summary: IOException: No FileSystem for scheme: gs"
     ]
    }
   ],
   "source": [
    "mt = db.annotate_rows_db(mt, 'CADD') # this is problem with google sdk in the container - to be fixed"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
